{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Questions for evaluation\n",
    "## Evolutionary rates. \n",
    "Explain why the ratio of non-synonymous to synonymous rates, Ka/Ks, is regarded as an indicator of positive selection. How do you interpret that Ka/Ks is significantly larger in one gene than in another one although it is smaller than one in both genes?\n",
    "### Response 1.\n",
    "\n",
    "The non synonymous substitution rate $K_a$ is usually smaller than the synonymous rate $K_s$ because a change in an amino acid entails more risks for the correct operation of the gene protein product and can cause the organism to lose fitness (deleterious mutation)\n",
    "\n",
    "Why $K_a/K_s$ is significantly larger in one gene than in another one although it is smaller than one in both genes?\n",
    "\n",
    "This ratio has variations from gene to gene (and in sites of the same gene) depending on the influence of protein function in the organism fitness, and so tends to be lower in genes that code more fundamental proteins. In the extreme opposite case, the ratio could be larger in genes that no code protein at all (gene duplications). \n",
    "\n",
    "Why $K_a/K_s$ is regarded as an indicator of positive selection?\n",
    "\n",
    "If ratio is more than one, it implies that the amino acid substitution far from causing a loss of fitness in the organisms affected, has given greater fitness advantage compared to others, and the gene is being conserved into population."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "lang": "en"
   },
   "source": [
    "## Population size\n",
    "Why do we say that effective population size has the role of the inverse of an evolutionary temperature? Out of two different populations, which one do you expect to evolve faster, the smaller one of the larger one? Why?\n",
    "### Response 2\n",
    "\n",
    "As temperature rises more microstates at a specific energy level are available to the system according to the Boltzamnn distribution of statistical mechanics, that has the form:\n",
    "\n",
    "$$f(E) = \\dfrac{1}{g(E/T)}$$\n",
    "\n",
    "Where $g$ is a increasing function depending on $E/T$ (in fact an exponential) and $f$ the probability that a particle has energy $E$ (proportional to number of microstates), because if $T$ increases, the ratio $E/T$ decreases, $g$ decreases and $1/g$ increases.\n",
    "\n",
    "In detailed form:\n",
    "\n",
    "$$f(E) = \\dfrac{1}{exp(E/k*T)}$$\n",
    "\n",
    "We can construct an analogy, between the probability of finding a particle of an energy level E, with the probability of finding an individual that carries a certain allele A within the population.\n",
    "\n",
    "This probability is inversely proportional to the size of the population, since the larger the population, the more organisms with alleles of other types can be found. Let's say that the bigger the population is, the lower is the probability that an organism acquires a new allele, that is, it behaves as an inverse to the temperature with respect to energy.\n",
    "\n",
    "Therefore, and answering the second question, small populations tend to evolve faster because they allow a greater number of alleles to coexist in the population."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "lang": "en"
   },
   "source": [
    "## Non-negligible mutation rate.\n",
    "Which phenomena occur when there is more than one mutation in the same genome, and there are two or more different mutants in the same population?\n",
    "### Response 3\n",
    "\n",
    "If there is more than one mutation in the same genome, and one of the mutations is advantageous enough, the other mutations, neutral of also disadvantageous could be fixed in the population, because the overall fitness of the organism was increased.\n",
    "\n",
    "The frequency of this mechanism depends on the location of the mutations in the genome. The closer they are, the more difficult it is for them to be separated by recombination events.\n",
    "\n",
    "Answering the second question, these meiotic recombinations can serve to bring mutations within the chromosome, so that deleterious mutations carried by one of the progenitors are located near advantageous mutations carried by the other parent and thus pass to the offspring. Thus we would find ourselves in the previous situation, and a neutral or even non-advantageous mutation would have possibilities to be fixed in the population."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Phylogenetic trees. \n",
    "Consider the phylogenetic tree of n protein sequences. How many branches are there? How many pairwise distances? How many internal nodes? How many possible rooted and unrooted trees?\n",
    "### Response 4.\n",
    "#### Rooted trees.\n",
    "We consider only binary trees.\n",
    "The rooted tree of $n$ protein sequences ($n$ leaf nodes) $T_n$ can be defined recursively as the set formed by its internal nodes $R_i$ and leaf nodes $L_j$ as follows:\n",
    "\n",
    "$$T_2 = L_2 \\cup L_1 \\cup R_1$$\n",
    "$$T_{n} = L_n \\cup T_{n-1} \\cup R_{n-1}$$\n",
    "\n",
    "*Note: $T_2$ is the first case: the minimal tree has two final nodes and one internal node*\n",
    "\n",
    "From  this definition we can guess some properties easily:\n",
    "\n",
    "**Number of internal nodes (cardinalR)**\n",
    "\n",
    "It seems by construction that it could be *$n-1$*. We will prove it by structural induction over $T$ sets:\n",
    "\n",
    "$cardinalR(T_2) = 1$, as $T_2$ only has the internal node $R_1$\n",
    "\n",
    "We suppose now, that $cardinalR(T_{n-1}) = n-2$, what about $T_{n}$?\n",
    " \n",
    "$cardinalR(T_{n}) = cardinalR(L_n \\cup T_{n-1} \\cup R_{n-1}) = cardinalR(L_n) + cardinalR(T_{n-1}) + cardinalR(R_{n-1}) = 0 + (n - 2) + 1 = n - 1$\n",
    "\n",
    "So, it's demonstrated.\n",
    "\n",
    "**Number of branches (countBranches)**\n",
    "\n",
    "From each of the internal nodes start two branches. So, the number of branches is:\n",
    "$$countBranches(n) = 2n-2$$\n",
    "\n",
    "**Number of trees (countTrees)**\n",
    "\n",
    "Think now on how many ways we have to obtain recursively a tree of $n$ nodes from a tree of $n-1$ nodes.\n",
    "\n",
    "We can insert a new branch pairing over an existing branch. Taken into account that is irrelevant insert the branch to the right or on the left of any existing branch, we have $countBranches(n-1)$ ways to do that operation. Also we have another possibility that is generate a new branch past the root of the tree. \n",
    "\n",
    "So we have $countBranches(n-1) + 1$ possibilities of generate a $n$ tree given a $n-1$ tree and then:\n",
    "\n",
    "$countTrees(n) = (countBranches(n-1) + 1) * countTrees(n-1) = ( (2(n-1) - 2)+ 1 ) * countTrees(n-1)$\n",
    "\n",
    "so\n",
    "\n",
    "$countTrees(n) = (2n-3) * countTrees(n-1)$, for $n > 2$\n",
    "\n",
    "and\n",
    "\n",
    "$countTrees(2) = 1$, trivially, because it's irrelevant the order of the two leaf nodes.\n",
    "\n",
    "and recursively applying the formula:\n",
    "\n",
    "$$ countTrees(n) = 1*3*5*...*(2n-3) = (2n-3)!!$$\n",
    "\n",
    "\n",
    "**Number of pairwise distances**\n",
    "\n",
    "It's equal to the number of different pairs we can form with the leaf nodes.\n",
    "\n",
    "$$\\binom{n}{2} = n(n-1)/2$$\n",
    "\n",
    "#### Unrooted trees.\n",
    "We consider only unrooted trees as the derived from a star topology like the target topology used in the neighbor join algorithm, that is, n leaf nodes and one internal node.\n",
    "So each of the internal nodes have three branches starting from it.\n",
    "\n",
    "The construction of such unrooted three is as follows. We take a pair of leaf nodes out from the star topology creating one internal node with three branches, one for every member of the pair of extracted nodes and other ending on the internal node of the star topology. This internal node is used as a leaf node in the remaining star topology, now a star with $n-1$ leaf nodes and the same internal node. \n",
    "\n",
    "This process can be done recursively until finishing with a tree with one internal node, three leaf nodes and three branches, that is the minimal possible unrooted tree. The case with two branches is a rooted tree.\n",
    "\n",
    "**Number of branches**\n",
    "\n",
    "By construction, for a tree with n leaf nodes, we have\n",
    "\n",
    "$countBranches(n) = 2 + countBranches(n-1)$\n",
    "\n",
    "and\n",
    "\n",
    "$countBranches(3) = 3$\n",
    "\n",
    "$countBranches(4) = 2 + countBranches(3) = 5$\n",
    "\n",
    "$countBranches(5) = 2 + countBranches(4) = 7$\n",
    "\n",
    "and so on...\n",
    "\n",
    "It seems that this should be the generic formulation:\n",
    "$$countBranches(n) = 2n -3$$\n",
    "\n",
    "Effectively, by induction over n:\n",
    "\n",
    "$countBranches(n) = 2 + countBranches(n-1) = 2 + (2*(n-1) - 3) = 2 + (2n - 2 - 3) = 2n -3$\n",
    "\n",
    "And for the first case:\n",
    "\n",
    "$countBranches(3) = 2*3 -3 = 6 - 3 = 3$\n",
    "\n",
    "**Number of internal nodes**\n",
    "\n",
    "The approach is very similar:\n",
    "By construction, for a tree with n leaf nodes, we have\n",
    "\n",
    "$countNodes(n) = 1 + countNodes(n-1)$\n",
    "\n",
    "and\n",
    "\n",
    "$countNodes(3) = 1$\n",
    "\n",
    "so\n",
    "\n",
    "$countNodes(4) = 1 + countNodes(3) = 2$\n",
    "\n",
    "$countNodes(5) = 1 + countNodes(4) = 3$\n",
    "\n",
    "ans so on...\n",
    "\n",
    "It seems that this should be the generic formulation:\n",
    "$$countNodes(n) = n - 2$$\n",
    "\n",
    "Effectively, by induction over n:\n",
    "$countNodes(n) = 1 + countNodes(n-1) = 1 + ((n-1) - 2) = 1 + (n - 3) = n -2$\n",
    "\n",
    "And for the first case\n",
    "$countNodes(3) = 3 - 2 = 1$\n",
    "\n",
    "**Number of trees (countTrees)**\n",
    "\n",
    "Think now on how many ways we have to obtain recursively a tree of $n$ nodes from a tree of $n-1$ nodes.\n",
    "\n",
    "We can insert a new branch pairing over an existing branch. Taken into account that is irrelevant insert the branch to the right or on the left of any existing branch, we have $countBranches(n-1)$ ways to do that operation. \n",
    "\n",
    "So we have $countBranches(n-1)$ possibilities of generate a $n$ tree given a $n-1$ tree and then:\n",
    "\n",
    "$countTrees(n) = (countBranches(n-1)) * countTrees(n-1) = (2(n-1) - 3) * countTrees(n-1)$\n",
    "\n",
    "so\n",
    "\n",
    "$countTrees(n) = (2n-5) * countTrees(n-1)$, for $n > 3$\n",
    "\n",
    "and\n",
    "\n",
    "$countTrees(3) = 1$, trivially, because it's irrelevant the order of the three leaf nodes.\n",
    "\n",
    "and recursively applying the formula:\n",
    "\n",
    "$$ countTrees(n) = 1*3*5*...*(2n-5) = (2n-5)!!$$\n",
    "\n",
    "\n",
    "**Number of pairwise distances**\n",
    "It's equal to the number of different pairs we can form with the leaf nodes.\n",
    "\n",
    "$$\\binom{n}{2} = n(n-1)/2$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Additive distances. \n",
    "Sequence distances are said to be additive when the distance between two sequences is the sum of the length of the branches that connect them. If there are n species, how many equations express the additivity conditions? Are the free parameters (branch lengths) underdetermined or overdetermined? What is the number of sequences for which the number of equations is the same as the number of parameters? For this number of sequences, which condition expresses the molecular clock hypothesis?\n",
    "### Response 5.\n",
    "\n",
    "The number of additive conditions for n species equals the number of pairwise distances $\\binom{n}{2} = n(n-1)/2$. \n",
    "\n",
    "The number of branches in rooted trees (see previous exercise), is $2n-2$.\n",
    "\n",
    "The number of parameters (branch lengths) is underdetermined if is greater than the number of equations over distances(additive conditions):\n",
    "\n",
    "$$2n-2 > \\dfrac{n(n-1)}{2}$$\n",
    "\n",
    "For instance, with $n = 2$ we have a distance equation and two branches: $d_{12} = b_1 + b_2$ where $d_{12}$ is the pairwise distance between the sequences at two leafs nodes $L_1$ and $L_2$, and $b_i$ are the branch lengths.\n",
    "\n",
    "Conversely, the branch lengths are overdetermined if the number of equations (pairwise distances) are greater than the number of parameters:\n",
    "\n",
    "$$2n-2 < \\dfrac{n(n-1)}{2}$$\n",
    "\n",
    "And it is determined if we have equality. Equality is reached for rooted trees at $n = 4$. For $n > 4$ the branch lengths are overdetermined.\n",
    "\n",
    "If molecular clock hypothesis is true for this four sequences, all the distances from the leaves to the root must be equal.\n",
    "\n",
    "This can be written as a set of three equations. Be $i = 1,2,3,4$ the labels of the leaf nodes and $0$ the label of the root node and d_{ij} the distance between nodes:\n",
    "\n",
    "$$ \\begin{array}{lcl} d_{01} = d_{02}\\\\ d_{01} = d_{03} \\\\ d_{02} = d_{03} \\end{array}$$\n",
    "\n",
    "Because the six branch lengths are determined uniquely by the six equations, there is nothing we can do to force ultrametricity. But molecular clock conditions also could be satisfied depending on the values of the distances, i.e, the three additional conditions being linearly proportional to the other six equations. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Distance methods versus model-based methods for reconstructing phylogenetic trees. \n",
    "Which are the advantages and disadvantages of these two types of methods?\n",
    "### Response 6\n",
    "\n",
    "*Advantages Model-based methods vs distance methods:*\n",
    "\n",
    "1) Don't use distances as the basis of branch length calculations. For this reason are more accurate than distance method. The use of distances has a lot of well know problems: \n",
    "\n",
    "- Construction of the tree heuristics. Main methods do strong biological assumptions as additive distances (NJ) or ultrametricity (UPGMA). Also the trees are prone to errors derived for the errors in pairwise alignments, that in CLUSTAL remains unchanged all over the computation. More sofisticated methods (T-COFFEE or MUSCLE), try to avoid this problems, allow some changes in the initial alignments (T-COFFEE) and in the tree topology, that is explored iteratively (MUSCLE).\n",
    "\n",
    "2) Model methods are based on a consistent and compact model of evolution, and have a more solid biological basis. Distance methods are a eclectic mixture of biological concepts.\n",
    "\n",
    "*Advantages Distance method vs Model-based methods*\n",
    "\n",
    "1) Less computational time. Model based methods have NP complexity. Distance method give a relatively fast tree and somehow accurate.\n",
    "\n",
    "2) ML works on samples from tree population, because it's impossible to explore the whole population of  trees: this population is huge and grows O(n!) with the number of sequences (see exercise 4). There are local optimization methods but is not guarantee that the tree selected was not in a local maximum. \n",
    "\n",
    "\n",
    "Both types of methods share also problems related to the alignment strategies:\n",
    "\n",
    "- Correct selection of true orthologs as leaf nodes.\n",
    "\n",
    "- Dependence of alignment method: parametrization and scoring (substitution matrices, gap penalties, ...). A subtle difference in the choice of parameters could originate a different tree. In particular the scoring of gaps, gap open penalty and gap extend are crucial and with no clear biological basis.  \n",
    "\n",
    "- Alignment method as is. Heuristics. Multialignment is NP complete and is not possible to do a n-dimensional \"multiwise\" precise alignment. Commonly the starting point is the whole set of pairwise comparisons between every pair of sequences. The pairwise alignment has to be done by heuristic approaches also, if the number of sequences is huge.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## AIC BIC scores.\n",
    "Why are the AIC or BIC scores needed in order to test alternative models of molecular evolution?\n",
    "### Response 7\n",
    "\n",
    "Models with more parameters yield higher likelihood than models with less parameters (overfiting)\n",
    "\n",
    "So, it's necessary a strategy to score the testing tree penalizing the number of parameters. \n",
    "\n",
    "AIC and BIC are statistical estimators of the relative quality of the models. The more generic and also simpler quality control formula is such this:\n",
    "\n",
    "$$q(p, L) = k_1*q_1(p) - k_2*q_2(L)$$\n",
    "\n",
    "where p is the number of parameters, L the likelihood of the tree, k_1 and k_2 two positive constants and q_1 and q_2 two increasing funcions. The model with lower Q is the preferred.\n",
    "\n",
    "AIC and BIC are in accordance with this general formula."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Substitution rates\n",
    "Which features determine the among-site variation of the substitution rate? How is this variation implemented in empirical models of the substitution process?\n",
    "### Response 8\n",
    "\n",
    "The  functional of the gene or protein determine the among-site variation of the substitution rate: \n",
    "\n",
    "1. Proteins performing fundamentally functions tend to evolve more slowly than other proteins.\n",
    "\n",
    "2. Some protein domains need to be evolutionary conserved more than other domains of the same protein because they participate more strongly on the protein function (active sites). In that domains, substitution rate is lower than other domains.\n",
    "\n",
    "3. At gene level, third and second codon positions are less conserved than the first codon position, because the first position determines more strongly the amino acid at translation step than the second position, and the second codon position more strongly than third codon position, so substitution rates ($r$), verify $r_1<r_2<r_3$.\n",
    "\n",
    "\n",
    "**Empirical models**\n",
    "\n",
    "They are based on the hypothesis that each site has a rate that is determined by its position in the molecule, regardless the residue that occupies the site. \n",
    "\n",
    "We can typify two types of approaches.\n",
    "\n",
    "1. Step wise. Dividing the gene/protein in two or more regions of different substitution rates. The simplest model take into account invariant sites, at which the sequence is considered fully conserved. \n",
    "\n",
    "2. Continuous wise. Gamma distributions. The transition from active sites with low or null substitution rate to sites with high rate are modeled with a function depending on a shape parameter. The larger the shape parameter $\\alpha$, the smaller the substitution rate."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Phylogenetic networks.\n",
    "When is it suitable to model evolutionary relationships with a network rather than with a tree, despite the increase of complication?\n",
    "### Response 9\n",
    "Generally the genetic information flows not only in the direction from one evolutionary parent. If we consider that these alternative gene flows or alternative genetic sources are relevant, we need to improve the topology of genetic relationship to a more flexible model that take into account all possible paths to exchange genetic information. In this case we need to adopt a network topology, a hierarchical binary tree is not enough.\n",
    "\n",
    "Alternative genetic information paths are:\n",
    "\n",
    "1. Recombination between parents. If a species has two parents, both parent are evolutionary related to it, but they could be unrelated among them.\n",
    "\n",
    "2. Hybridation.\n",
    "\n",
    "3. Horizontal gene transfer.\n",
    "\n",
    "4. Incorporation of genomes (Margulis hypothesis).\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Influence of protein stability on fitness \n",
    "How do we model the influence of protein stability on fitness? Explain why we say that the evolution of protein stability is largely a neutral process, and in which range of parameters this is true.\n",
    "### Response 10\n",
    "\n",
    "The fitness are modeled as the probability to find the protein in the native state, in function of free Gibbs energy of the transition folded state. The more negative $\\Delta G_{folded}$ the larger the fitness, because the transition to folder state is more likely to occur spontaneously, i.e., the reaction ${protein}_{unfolded} \\rightleftharpoons {protein}_{folded}$ is more biased from left to right. \n",
    "\n",
    "$$ \\Delta {fitness} = {fitness}_{folded} - {fitness}_{unfolded}$$\n",
    "$$ \\Delta {fitness} \\cong \\dfrac{1}{1 + e^{\\Delta G_{folded}/k_B T}}$$\n",
    "\n",
    "This way of thinking assumes that the protein develops all its functionality at a folded native state, and thus this the more folded versions of the protein, the the best fitness to organism. But it's not always true, there are natively unstructured proteins and the protein function could depend on the linkage to another proteins or compounds of the system.\n",
    "\n",
    "There are a lot of experimental studies showing that the mutations often have no relevant effect over  protein's structure or function. It seems that protein evolution can described by Kimura's neutral theory of evolution: most genetic change is due to the stochastic fixation of neutral mutations, that not change the organism fitness.\n",
    "\n",
    "From the previous equation, if temperature is low or the protein is enough large $|\\Delta G|$ high, that is $|\\Delta G| >> KT$ denominator goes to a very high value, and the variation in fitness tend to zero.\n",
    "\n",
    "As the effect of a mutation on stability is $\\Delta\\Delta G= \\Delta G_{mutant} − \\Delta G_{wild−type}$, we have \n",
    "\n",
    "$\\Delta G_{mutant} = \\Delta G_{wild−type} + \\Delta\\Delta G$\n",
    "\n",
    "So the limit condition that requires $|\\Delta G| = |\\Delta G_{mutant}|$ be large, imposes also that $|\\Delta G_{wild−type}| >> \\Delta\\Delta G$\n",
    "\n",
    "So the more stable the wild type of the protein is, more neutrality is reached, and this favors evolvability.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "lang": "en"
   },
   "source": [
    "## Chaperones.\n",
    "Describe the role of chaperones in evolution\n",
    "### Response 11\n",
    "\n",
    "Chaperones help the folding of unfolded and wrongly folded proteins, preventing aggregation.\n",
    "\n",
    "So chaperones confer to the system: protein, substrate and the same chaperones a $|\\Delta G_{folded}|$ higher than they would have if they were not easing the folding.\n",
    "\n",
    "Therefore confer more stability to the protein and then, according to what was discussed at the end of the previous exercise, favor their evolution, increasing their ability to store non-deleterious changes, or what is the same,  accumulating possible new functions.\n",
    "\n",
    "Also, chaperones buffer deleterious mutations, by increasing their expression levels."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "lang": "en"
   },
   "source": [
    "## Protein stability\n",
    "Describe the trade-offs and synergies between stability against unfolding and stability against misfolding.\n",
    "### Response 12.\n",
    "A protein primary structure needs, to achieve proper functioning, to follow two structural pathways to the 3D structure:\n",
    "\n",
    "1. To ease the folding to a stable three-dimensional shape. Then to have a suitable unfolding stability.\n",
    "\n",
    "2. To provide appropriate residues for binding with other proteins or ligands. So to have a adequate misfolding stability.\n",
    "\n",
    "This originates a trade-off: the protein folding process needs to deal with two constraints at the same time: that the fold is stable but not stable enough to block the functional residues needed to the proper protein function.\n",
    "\n",
    "I think on two types of synergies to face these difficulties:\n",
    "\n",
    "1. Temporal synergies, based on the thermodynamic oscillation between less compact (misfolded)and more compact (folded) states, so that the necessary interactions can be favored in short time.\n",
    "\n",
    "2. Spacial synergies. Only those residues whose properties do not contribute to function must be the basis for the folding, and the segments containing functional residues needed for binding must to be the hardest to fold."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "lang": "en"
   },
   "source": [
    "## Protein classification\n",
    "How do protein proteins structures change when their sequences diverge? Is it justified to assume that homologous proteins (same superfamily) always belongs to the same structural class (fold)?\n",
    "### Response 13.\n",
    "\n",
    "The RMSD of protein sequences increases exponentially with identity percentage so that the structures are very tolerant to changes in sequences, i. e. structure divergence is evolutionary slower than sequence divergence: selective pressure is stronger over protein structure than over protein sequence, structure is strongly conserved by evolution.\n",
    "\n",
    "This scenario encourages us to classify proteins into families based on the homology between their sequences. They can be aligned following one of the known algorithms of alignment.\n",
    "\n",
    "However, as we commented about exercise 6, there is some aspects of those algorithms that reside in the correct score of the gaps, with little or no biological basis. These gaps can separate or close areas of the proteins that interact with each other during the bending process, significantly modifying their native state.\n",
    "\n",
    "Therefore, a good alignment, with a high percentage of identity, but that includes sequences of gaps in relevant sites, may be overestimating the three-dimensional analogy of the proteins that are being compared.\n",
    "\n",
    "Another aspect of the alignment of sequences that can be misleading, is that they do not take into account possible internal coupling (that could be even co-evolutionary), between certain residues that interact to achieve the native conformation. These residues should be anchor characters with a higher algorithmic score to favor their 1-D alignment."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Disordered protein\n",
    "Which are the evolutionary properties of disordered proteins that differentiate them most from structured proteins?\n",
    "### Response 14."
   ]
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "nbTranslate": {
   "displayLangs": [
    "en",
    "sp"
   ],
   "hotkey": "alt-t",
   "langInMainMenu": true,
   "sourceLang": "es",
   "targetLang": "en",
   "useGoogleTranslate": true
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "213px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
