{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Questions for evaluation\n",
    "## Evolutionary rates. \n",
    "Explain why the ratio of non-synonymous to synonymous rates, Ka/Ks, is regarded as an indicator of positive selection. How do you interpret that Ka/Ks is significantly larger in one gene than in another one although it is smaller than one in both genes?\n",
    "### Response 1.\n",
    "## Population size. \n",
    "Why do we say that effective population size has the role of the inverse of an evolutionary temperature? Out of two different populations, which one do you expect to evolve faster, the smaller of the larger one? Why?\n",
    "### Response 2.\n",
    "## Non-negligible mutation rate. \n",
    "Which phenomena may occur when there is more than one mutation in the same genome, and there are two or more different mutant individuals in the same population? \n",
    "### Response 3."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Phylogenetic trees. \n",
    "Consider the phylogenetic tree of n protein sequences. How many branches are there? How many pairwise distances? How many internal nodes? How many possible rooted and unrooted trees?\n",
    "### Response 4.\n",
    "#### Rooted trees.\n",
    "We consider only binary trees.\n",
    "The rooted tree of $n$ protein sequences ($n$ leaf nodes) $T_n$ can be defined recursively as the set formed by its internal nodes $R_i$ and leaf nodes $L_j$ as follows:\n",
    "\n",
    "$$T_2 = L_2 \\cup L_1 \\cup R_1$$\n",
    "$$T_{n} = L_n \\cup T_{n-1} \\cup R_{n-1}$$\n",
    "\n",
    "*Note: $T_2$ is the first case: the minimal tree has two final nodes and one internal node*\n",
    "\n",
    "From  this definition we can guess some properties easily:\n",
    "\n",
    "**Number of internal nodes (cardinalR)**\n",
    "\n",
    "It seems by construction that it could be *$n-1$*. We will prove it by structural induction over $T$ sets:\n",
    "\n",
    "$cardinalR(T_2) = 1$, as $T_2$ only has the internal node $R_1$\n",
    "\n",
    "We suppose now, that $cardinalR(T_{n-1}) = n-2$, what about $T_{n}$?\n",
    " \n",
    "$cardinalR(T_{n}) = cardinalR(L_n \\cup T_{n-1} \\cup R_{n-1}) = cardinalR(L_n) + cardinalR(T_{n-1}) + cardinalR(R_{n-1}) = 0 + (n - 2) + 1 = n - 1$\n",
    "\n",
    "So, it's demonstrated.\n",
    "\n",
    "**Number of branches (countBranches)**\n",
    "\n",
    "From each of the internal nodes start two branches. So, the number of branches is:\n",
    "$$countBranches(n) = 2n-2$$\n",
    "\n",
    "**Number of trees (countTrees)**\n",
    "\n",
    "Think now on how many ways we have to obtain recursively a tree of $n$ nodes from a tree of $n-1$ nodes.\n",
    "\n",
    "We can insert a new branch pairing over an existing branch. Taken into account that is irrelevant insert the branch to the right or on the left of any existing branch, we have $countBranches(n-1)$ ways to do that operation. Also we have another possibility that is generate a new branch past the root of the tree. \n",
    "\n",
    "So we have $countBranches(n-1) + 1$ possibilities of generate a $n$ tree given a $n-1$ tree and then:\n",
    "\n",
    "$countTrees(n) = (countBranches(n-1) + 1) * countTrees(n-1) = ( (2(n-1) - 2)+ 1 ) * countTrees(n-1)$\n",
    "\n",
    "so\n",
    "\n",
    "$countTrees(n) = (2n-3) * countTrees(n-1)$, for $n > 2$\n",
    "\n",
    "and\n",
    "\n",
    "$countTrees(2) = 1$, trivially, because it's irrelevant the order of the two leaf nodes.\n",
    "\n",
    "and recursively applying the formula:\n",
    "\n",
    "$$ countTrees(n) = 1*3*5*...*(2n-3) = (2n-3)!!$$\n",
    "\n",
    "\n",
    "**Number of pairwise distances**\n",
    "\n",
    "It's equal to the number of different pairs we can form with the leaf nodes.\n",
    "\n",
    "$$\\binom{n}{2} = n(n-1)/2$$\n",
    "\n",
    "#### Unrooted trees.\n",
    "We consider only unrooted trees as the derived from a star topology like the target topology used in the neighbor join algorithm, that is, n leaf nodes and one internal node.\n",
    "So each of the internal nodes have three branches starting from it.\n",
    "\n",
    "The construction of such unrooted three is as follows. We take a pair of leaf nodes out from the star topology creating one internal node with three branches, one for every member of the pair of extracted nodes and other ending on the internal node of the star topology. This internal node is used as a leaf node in the remaining star topology, now a star with $n-1$ leaf nodes and the same internal node. \n",
    "\n",
    "This process can be done recursively until finishing with a tree with one internal node, three leaf nodes and three branches, that is the minimal possible unrooted tree. The case with two branches is a rooted tree.\n",
    "\n",
    "**Number of branches**\n",
    "\n",
    "By construction, for a tree with n leaf nodes, we have\n",
    "\n",
    "$countBranches(n) = 2 + countBranches(n-1)$\n",
    "\n",
    "and\n",
    "\n",
    "$countBranches(3) = 3$\n",
    "\n",
    "$countBranches(4) = 2 + countBranches(3) = 5$\n",
    "\n",
    "$countBranches(5) = 2 + countBranches(4) = 7$\n",
    "\n",
    "and so on...\n",
    "\n",
    "It seems that this should be the generic formulation:\n",
    "$$countBranches(n) = 2n -3$$\n",
    "\n",
    "Effectively, by induction over n:\n",
    "\n",
    "$countBranches(n) = 2 + countBranches(n-1) = 2 + (2*(n-1) - 3) = 2 + (2n - 2 - 3) = 2n -3$\n",
    "\n",
    "And for the first case:\n",
    "\n",
    "$countBranches(3) = 2*3 -3 = 6 - 3 = 3$\n",
    "\n",
    "**Number of internal nodes**\n",
    "\n",
    "The approach is very similar:\n",
    "By construction, for a tree with n leaf nodes, we have\n",
    "\n",
    "$countNodes(n) = 1 + countNodes(n-1)$\n",
    "\n",
    "and\n",
    "\n",
    "$countNodes(3) = 1$\n",
    "\n",
    "so\n",
    "\n",
    "$countNodes(4) = 1 + countNodes(3) = 2$\n",
    "\n",
    "$countNodes(5) = 1 + countNodes(4) = 3$\n",
    "\n",
    "ans so on...\n",
    "\n",
    "It seems that this should be the generic formulation:\n",
    "$$countNodes(n) = n - 2$$\n",
    "\n",
    "Effectively, by induction over n:\n",
    "$countNodes(n) = 1 + countNodes(n-1) = 1 + ((n-1) - 2) = 1 + (n - 3) = n -2$\n",
    "\n",
    "And for the first case\n",
    "$countNodes(3) = 3 - 2 = 1$\n",
    "\n",
    "**Number of trees (countTrees)**\n",
    "\n",
    "Think now on how many ways we have to obtain recursively a tree of $n$ nodes from a tree of $n-1$ nodes.\n",
    "\n",
    "We can insert a new branch pairing over an existing branch. Taken into account that is irrelevant insert the branch to the right or on the left of any existing branch, we have $countBranches(n-1)$ ways to do that operation. \n",
    "\n",
    "So we have $countBranches(n-1)$ possibilities of generate a $n$ tree given a $n-1$ tree and then:\n",
    "\n",
    "$countTrees(n) = (countBranches(n-1)) * countTrees(n-1) = (2(n-1) - 3) * countTrees(n-1)$\n",
    "\n",
    "so\n",
    "\n",
    "$countTrees(n) = (2n-5) * countTrees(n-1)$, for $n > 3$\n",
    "\n",
    "and\n",
    "\n",
    "$countTrees(3) = 1$, trivially, because it's irrelevant the order of the three leaf nodes.\n",
    "\n",
    "and recursively applying the formula:\n",
    "\n",
    "$$ countTrees(n) = 1*3*5*...*(2n-5) = (2n-5)!!$$\n",
    "\n",
    "\n",
    "**Number of pairwise distances**\n",
    "It's equal to the number of different pairs we can form with the leaf nodes.\n",
    "\n",
    "$$\\binom{n}{2} = n(n-1)/2$$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Additive distances. \n",
    "Sequence distances are said to be additive when the distance between two sequences is the sum of the length of the branches that connect them. If there are n species, how many equations express the additivity conditions? Are the free parameters (branch lengths) underdetermined or overdetermined? What is the number of sequences for which the number of equations is the same as the number of parameters? For this number of sequences, which condition expresses the molecular clock hypothesis?\n",
    "### Response 5.\n",
    "\n",
    "The number of additive conditions for n species equals the number of pairwise distances $\\binom{n}{2} = n(n-1)/2$. \n",
    "\n",
    "The number of branches in rooted trees (see previous exercise), is $2n-2$.\n",
    "\n",
    "The number of parameters (branch lengths) is underdetermined if is greater than the number of equations over distances(additive conditions):\n",
    "\n",
    "$$2n-2 > \\dfrac{n(n-1)}{2}$$\n",
    "\n",
    "For instance, with $n = 2$ we have a distance equation and two branches: $d_{12} = b_1 + b_2$ where $d_{12}$ is the pairwise distance between the sequences at two leafs nodes $L_1$ and $L_2$, and $b_i$ are the branch lengths.\n",
    "\n",
    "Conversely, the branch lengths are overdetermined if the number of equations (pairwise distances) are greater than the number of parameters:\n",
    "\n",
    "$$2n-2 < \\dfrac{n(n-1)}{2}$$\n",
    "\n",
    "And it is determined if we have equality. Equality is reached for rooted trees at $n = 4$. For $n > 4$ the branch lengths are overdetermined.\n",
    "\n",
    "If molecular clock hypothesis is true for this four sequences, all the distances from the leaves to the root must be equal.\n",
    "\n",
    "This can be written as a set of three equations. Be $i = 1,2,3,4$ the labels of the leaf nodes and $0$ the label of the root node and d_{ij} the distance between nodes:\n",
    "\n",
    "$$ \\begin{array}{lcl} d_{01} = d_{02}\\\\ d_{01} = d_{03} \\\\ d_{02} = d_{03} \\end{array}$$\n",
    "\n",
    "Because the six branch lengths are determined uniquely by the six equations, there is nothing we can do to force ultrametricity. But molecular clock conditions also could be satisfied depending on the values of the distances, i.e, the three additional conditions being linearly proportional to the other six equations. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Distance methods versus model-based methods for reconstructing phylogenetic trees. \n",
    "Which are the advantages and disadvantages of these two types of methods?\n",
    "### Response 6\n",
    "\n",
    "*Advantages Model-based methods vs distance methods:*\n",
    "\n",
    "1) Don't use distances as the basis of branch length calculations. For this reason are more accurate than distance method. The use of distances has a lot of well know problems: \n",
    "\n",
    "- Construction of the tree heuristics. Main methods do strong biological assumptions as additive distances (NJ) or ultrametricity (UPGMA). Also the trees are prone to errors derived for the errors in pairwise alignments, that in CLUSTAL remains unchanged all over the computation. More sofisticated methods (T-COFFEE or MUSCLE), try to avoid this problems, allow some changes in the initial alignments (T-COFFEE) and in the tree topology, that is explored iteratively (MUSCLE).\n",
    "\n",
    "2) Model methods are based on a consistent and compact model of evolution, and have a more solid biological basis. Distance methods are a eclectic mixture of biological concepts.\n",
    "\n",
    "*Advantages Distance method vs Model-based methods*\n",
    "\n",
    "1) Less computational time. Model based methods have NP complexity. Distance method give a relatively fast tree and somehow accurate.\n",
    "\n",
    "2) ML works on samples from tree population, because it's impossible to explore the whole population of  trees: this population is huge and grows O(n!) with the number of sequences (see exercise 4). There are local optimization methods but is not guarantee that the tree selected was not in a local maximum. \n",
    "\n",
    "\n",
    "Both types of methods share also problems related to the alignment strategies:\n",
    "\n",
    "- Correct selection of true orthologs as leaf nodes.\n",
    "\n",
    "- Dependence of alignment method: parametrization and scoring (substitution matrices, gap penalties, ...). A subtle difference in the choice of parameters could originate a different tree. In particular the scoring of gaps, gap open penalty and gap extend are crucial and with no clear biological basis.  \n",
    "\n",
    "- Alignment method as is. Heuristics. Multialignment is NP complete and is not possible to do a n-dimensional \"multiwise\" precise alignment. Commonly the starting point is the whole set of pairwise comparisons between every pair of sequences. The pairwise alignment has to be done by heuristic approaches also, if the number of sequences is huge.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## AIC BIC scores.\n",
    "Why are the AIC or BIC scores needed in order to test alternative models of molecular evolution?\n",
    "### Response 7\n",
    "\n",
    "Models with more parameters yield higher likelihood than models with less parameters (overfiting)\n",
    "\n",
    "So, it's necessary a strategy to score the testing tree penalizing the number of parameters. \n",
    "\n",
    "AIC and BIC are statistical estimators of the relative quality of the models. The more generic and also simpler quality control formula is such this:\n",
    "\n",
    "$$q(p, L) = k_1*q_1(p) - k_2*q_2(L)$$\n",
    "\n",
    "where p is the number of parameters, L the likelihood of the tree, k_1 and k_2 two positive constants and q_1 and q_2 two increasing funcions. The model with lower Q is the preferred.\n",
    "\n",
    "AIC and BIC are in accordance with this general formula."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Substitution rates\n",
    "Which features determine the among-site variation of the substitution rate? How is this variation implemented in empirical models of the substitution process?\n",
    "### Response 8"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Phylogenetic networks.\n",
    "When is it suitable to model evolutionary relationships with a network rather than with a tree, despite the increase of complication?\n",
    "### Response 9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "nbTranslate": {
   "displayLangs": [
    "en",
    "sp"
   ],
   "hotkey": "alt-t",
   "langInMainMenu": true,
   "sourceLang": "sp",
   "targetLang": "en",
   "useGoogleTranslate": false
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
