{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Protein Structure Prediction Practical\n",
    "\n",
    "I choose the target 1 and target 4 proteins to find a structure prediction:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "lang": "en"
   },
   "source": [
    "## Target 1\n",
    "\n",
    "The target protein structure:\n",
    "\n",
    "```text\n",
    "> Target 1, 202 residues\n",
    "QNYVFTSGNGIGGNGFTARVMAKVGTSYITFPCAYDAKLISFNGTVQGNITIKVRSRDIS\n",
    "GIPLDQQTTNPTFTFWLCRDTTNSSLKNFSNCSNNQFTPNKTTVISKTNPNGLNAWVTST\n",
    "DQNWYLGNTSGDTIYGVPNGDGKLSGTYLYSVVFSVVQFYGTSMPRLAFTVNLTPTGANS\n",
    "NHLYVPDNNAAVTIGPVKFWY\n",
    "```\n",
    "\n",
    "I have done the following steps:\n",
    "\n",
    "1. **Identify structural templates**. For that I perform a search on **HHPRED** with this target. The best alignments are shown in the top two line on the figure \\ref{fig:fig1}. These alignments show that there are bad similarity, in the domain of the twilight zone.\n",
    "\n",
    " So, the templates proposed by HHPRED are the proteins \"2VTW\" and \"2IUM\", both with a very good E-value and probability, but bad identity.\n",
    "The respective alignments are shown in figures \\ref{fig:fig2} and \\ref{fig:fig3}.\n",
    "\n",
    " A more laborious search in BLAST (PSI-BLAST and so on) reveal the results of figure \\ref{fig:fig11}, where the 2VTW protein is also found according to the HHPRED result. Again the identity is low: 20% for BLAST and 21% for HHPRED.\n",
    " \n",
    " Can we really rely on these proteins as structural templates? Notice that the coverage is low, in the case of 2VTW: 39%. This may seem a disadvantage, but if we take a look at the 3D structure of this protein we see that it consists of six chains, and our alignment takes place in only one of them. That is, our structural templates actually participate in only one sixth of the sequence.\n",
    " \n",
    " According to HHPRED documentation, the estimated probability of the template is the most relevant criterion to decide whether a template is actually homologous or just a high-scoring chance hit. When it is larger than 95%, say, the homology is nearly certain. This is the case.\n",
    " \n",
    " Also both possible templates are very close in their secondary structure, leaving aside that 2VTW consists of 6 chains and IUM of 3, so we definitely rule out that the alignments are not due to chance.\n",
    " \n",
    "\n",
    "2. But could we model the target protein completely or we have regions that could be modeled independently or even not modeled? The search for Signal Peptide and transmembrane regions produces null results at **SignalP** (figure \\ref{fig:fig5}) and **TMHMM** websites (figure \\ref{fig:fig4}). **Phobius** website (figure \\ref{fig:fig6}) is aligned with these predictions. This implies that we can trust the two previous templates.\n",
    "\n",
    "\n",
    "3. As the template proteins seem adequate to produce model, I direct **HHPRED** top accomplish this task, one model based on 2VTW protein and another combining 2VTW and 2IUM proteins.\n",
    "\n",
    "\n",
    "4. I verify the alignments on **Pymol** and they seem accurate. Figure \\ref{fig:fig7} shows how similar both model are. Figures \\ref{fig:fig8} and \\ref{fig:fig9} show the extremely close alignment between the models and their respective templates.\n",
    "\n",
    "\n",
    "5. For a last investigation I launch the prediction on a generalist server **Robetta** (figure \\ref{fig:fig12}) , that confirms the choose of templates, the absence of transmembrane or signal peptides and low disorder.\n",
    "\n",
    "\n",
    "6. **Quality of the model** (2VTW).  A query in **PROQ** (figure \\ref{fig:fig13}), revealed a good model quality."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Target 4\n",
    "\n",
    "The target protein structure:\n",
    "\n",
    "```text\n",
    ">Target 4, 154 residues\n",
    "MKKVLLLVFLSLTWLSASAQALVPITWTAYGLTFEAPKGILVEEDTEETFLLNNSRFYIT\n",
    "IQSLDSDGMTKSDLKSVLKDYANDDGVKDQSAVQEFELPQFFGTYLKGSCETDHCLYACL\n",
    "MTKAAGSGFYISIIYSKENENIAEKILKSFTMEE\n",
    "```\n",
    "\n",
    "I have done the following steps:\n",
    "\n",
    "1. First I perform a search on **HHPRED** with this target. The best alignments are shown in the two top line on the figure \\ref{fig:fig20}. These alignments are on the frontier of the twilight zone, but the HHPRED probability is 100%, so as I commented before, we can have confidence on the results.\n",
    "\n",
    " So, I choose as template the protein \"3HLZ\", with a very good E-value and probability.\n",
    "The alignment is shown in figure \\ref{fig:fig21}.\n",
    "\n",
    "\n",
    "2. The search for Signal Peptide and transmembrane regions produces signal peptide detection at **SignalP** website (figure \\ref{fig:fig22}) and **Phobius** website (figure \\ref{fig:fig23}). Both agree in that we have probably a Signal Peptide. To obtain a good model we needed to cut off out protein at a cleavage site at amino acid 22. No transmembrane regions, the protein is predicted as non cytoplasmic. \n",
    "\n",
    "\n",
    "3. As a additional check, we obtain the secondary structure prediction from **Robetta** website: \\ref{fig:fig24} \n",
    "\n",
    "\n",
    "4. Another check is to take in account the function expected of the target protein and the functions known, if any, for the templates. In this case I don't know the function expected for our protein.\n",
    "\n",
    "\n",
    "5. I verify the alignments on **Pymol** and they seem accurate. Figure \\ref{fig:fig25}. There are some regions that could be refined by hand.\n",
    "\n",
    "\n",
    "6. Finally, I check the model quality at **PROQ**. Figure \\ref{fig:fig26}.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figures"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\\begin{figure}\n",
    "  \\centerline{\\includegraphics{prot1/1_hhpred_prot1.png}}\n",
    "  \\caption{\\label{fig:fig1} HHPRED output}\n",
    "\\end{figure}\n",
    "\n",
    "\\begin{figure}\n",
    "  \\centerline{\\includegraphics{prot1/2_hhpred_ium_alignment.jpg}}\n",
    "  \\caption{\\label{fig:fig2} HHPRED alignment with IUM protein}\n",
    "\\end{figure}\n",
    "\n",
    "\\begin{figure}\n",
    "  \\centerline{\\includegraphics{prot1/2_hhpred_vtw_alignment.png}}\n",
    "  \\caption{\\label{fig:fig3} HHPRED alignment with VTW protein}\n",
    "\\end{figure}\n",
    "\n",
    "\\begin{figure}\n",
    "  \\centerline{\\includegraphics{prot1/3_tmhmm_transmembrane_prot1.png}}\n",
    "  \\caption{\\label{fig:fig4} TMHMM Transmembrane prediction}\n",
    "\\end{figure}\n",
    "\n",
    "\\begin{figure}\n",
    "  \\centerline{\\includegraphics{prot1/3_signalip_signal_peptide_prot1.png}}\n",
    "  \\caption{\\label{fig:fig5} SignalP. Signal Peptide prediction}\n",
    "\\end{figure}\n",
    "\n",
    "\\begin{figure}\n",
    "  \\centerline{\\includegraphics{prot1/3_phobius_transmembrane_prot1.png}}\n",
    "  \\caption{\\label{fig:fig6} Phobius transmembrane and signal peptide prediction}\n",
    "\\end{figure}\n",
    "\n",
    "\\begin{figure}\n",
    "  \\centerline{\\includegraphics{prot1/5_pymol_align_models_ium_green_2_vtw_prot1.png}}\n",
    "  \\caption{\\label{fig:fig7} Pymol. Both models aligned}\n",
    "\\end{figure}\n",
    "\n",
    "\\begin{figure}\n",
    "  \\centerline{\\includegraphics{prot1/5_pymol_align_template_model_ium_prot1.png}}\n",
    "  \\caption{\\label{fig:fig8} Pymol. Model (2VTW based) aligned with template}\n",
    "\\end{figure}\n",
    "\n",
    "\\begin{figure}\n",
    "  \\centerline{\\includegraphics{prot1/5_pymol_align_template_model_vtw_prot1.png}}\n",
    "  \\caption{\\label{fig:fig9} Pymol. Combined model(VTW and 2IUM) aligned with template 2IUM}\n",
    "\\end{figure}\n",
    "\n",
    "\\begin{figure}\n",
    "  \\centerline{\\includegraphics{prot1/3_psipred_secondary_prediction_prot1.pdf}}\n",
    "  \\caption{\\label{fig:fig10} PSIPRED. Secondary prediction}\n",
    "\\end{figure}\n",
    "\n",
    "\\begin{figure}\n",
    "  \\centerline{\\includegraphics{prot1/2_blast_prot1.png}}\n",
    "  \\caption{\\label{fig:fig11} BLAST computed similar sequences.}\n",
    "\\end{figure}\n",
    "\n",
    "\\begin{figure}\n",
    "  \\centerline{\\includegraphics{prot1/2_robetta_prot1.png}}\n",
    "  \\caption{\\label{fig:fig12} Robetta global prediction.}\n",
    "\\end{figure}\n",
    "\n",
    "\\begin{figure}\n",
    "  \\centerline{\\includegraphics{prot1/ProQ_vtw_model_prot1.png}}\n",
    "  \\caption{\\label{fig:fig13} ProQ quality estimation.}\n",
    "\\end{figure}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\\begin{figure}\n",
    "  \\centerline{\\includegraphics{prot4/HHPRED_global.png}}\n",
    "  \\caption{\\label{fig:fig20} HHPRED output}\n",
    "\\end{figure}\n",
    "\n",
    "\\begin{figure}\n",
    "  \\centerline{\\includegraphics{prot4/HHPRED_3HLZ.png}}\n",
    "  \\caption{\\label{fig:fig21} HHPRED alignment with 3HLZ protein}\n",
    "\\end{figure}\n",
    "\n",
    "\\begin{figure}\n",
    "  \\centerline{\\includegraphics{prot4/Phobius_prot4.png}}\n",
    "  \\caption{\\label{fig:fig22} Phobius prediction}\n",
    "\\end{figure}\n",
    "\n",
    "\\begin{figure}\n",
    "  \\centerline{\\includegraphics{prot4/SignalP_prot4.png}}\n",
    "  \\caption{\\label{fig:fig23} SignalP output}\n",
    "\\end{figure}\n",
    "\n",
    "\\begin{figure}\n",
    "  \\centerline{\\includegraphics{prot4/02_robetta_prot4.png}}\n",
    "  \\caption{\\label{fig:fig24} Robetta output}\n",
    "\\end{figure}\n",
    "\n",
    "\\begin{figure}\n",
    "  \\centerline{\\includegraphics{prot4/PYMOL_model2_3hlz.png}}\n",
    "  \\caption{\\label{fig:fig25} PYMOL 3D alignment with 3HLZ protein}\n",
    "\\end{figure}\n",
    "\n",
    "\\begin{figure}\n",
    "  \\centerline{\\includegraphics{prot4/ProQ_model_prot4.png}}\n",
    "  \\caption{\\label{fig:fig26} PROQ 3HLZ template based model quality}\n",
    "\\end{figure}\n",
    "\n",
    "\\begin{figure}\n",
    "  \\centerline{\\includegraphics{prot4/I-TASSER results.pdf}}\n",
    "  \\caption{\\label{fig:fig27} I-TASSER predictions}\n",
    "\\end{figure}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {
    "latex_metadata": {
     "hidden": "true",
     "lexer": "bash"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is pdfTeX, Version 3.14159265-2.6-1.40.19 (TeX Live 2018) (preloaded format=pdflatex)\n",
      " \\write18 enabled.\n",
      "entering extended mode\n",
      "(./structure_prediction.tex\n",
      "LaTeX2e <2018-04-01> patch level 2\n",
      "Babel <3.18> and hyphenation patterns for 84 language(s) loaded.\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/base/article.cls\n",
      "Document Class: article 2014/09/29 v1.4h Standard LaTeX document class\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/base/size11.clo))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/placeins/placeins.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/amsfonts/amssymb.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/amsfonts/amsfonts.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/amsmath/amsmath.sty\n",
      "For additional information on amsmath, use the `?' option.\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/amsmath/amstext.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/amsmath/amsgen.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/amsmath/amsbsy.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/amsmath/amsopn.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/amscls/amsthm.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/mathtools/mathtools.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/graphics/keyval.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/tools/calc.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/mathtools/mhsetup.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/base/fontenc.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/base/t1enc.def))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/psnfss/mathpazo.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/graphics/graphicx.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/graphics/graphics.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/graphics/trig.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/graphics-cfg/graphics.cfg)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/graphics-def/pdftex.def)))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/caption/caption.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/caption/caption3.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/adjustbox/adjustbox.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/xkeyval/xkeyval.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/xkeyval/xkeyval.tex\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/xkeyval/xkvutils.tex)))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/adjustbox/adjcalc.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/adjustbox/trimclip.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/collectbox/collectbox.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/adjustbox/tc-pdftex.def))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/ifoddpage/ifoddpage.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/varwidth/varwidth.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/xcolor/xcolor.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/graphics-cfg/color.cfg))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/tools/enumerate.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/geometry/geometry.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/ifpdf.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/ifvtex.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/ifxetex/ifxetex.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/base/textcomp.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/base/ts1enc.def\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/base/ts1enc.dfu)))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/upquote/upquote.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/eurosym/eurosym.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/ucs/ucs.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/ucs/data/uni-global.def))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/base/inputenc.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/ucs/utf8x.def))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/fancyvrb/fancyvrb.sty\n",
      "Style option: `fancyvrb' v2.7a, with DG/SPQR fixes, and firstline=lastline fix \n",
      "<2008/02/07> (tvz))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/oberdiek/grffile.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/oberdiek/kvoptions.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/ltxcmds.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/kvsetkeys.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/infwarerr.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/etexcmds.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/ifluatex.sty))))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/pdftexcmds.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/hyperref/hyperref.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/hobsub-hyperref.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/hobsub-generic.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/oberdiek/auxhook.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/hyperref/pd1enc.def)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/latexconfig/hyperref.cfg)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/url/url.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/hyperref/hpdftex.def\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/oberdiek/rerunfilecheck.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/tools/longtable.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/booktabs/booktabs.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/enumitem/enumitem.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/ulem/ulem.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/minted/minted.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/fvextra/fvextra.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/base/ifthen.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/etoolbox/etoolbox.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/lineno/lineno.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/tools/shellesc.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/ifplatform/ifplatform.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/catchfile.sty)\n",
      "(./structure_prediction.w18))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/xstring/xstring.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/xstring/xstring.tex))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/framed/framed.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/float/float.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/tcolorbox/tcolorbox.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/pgf/basiclayer/pgf.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/pgf/utilities/pgfrcs.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/utilities/pgfutil-common.te\n",
      "x\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/utilities/pgfutil-common-li\n",
      "sts.tex))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/utilities/pgfutil-latex.def\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/ms/everyshi.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/utilities/pgfrcs.code.tex))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/pgf/basiclayer/pgfcore.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/pgf/systemlayer/pgfsys.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/systemlayer/pgfsys.code.tex\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/utilities/pgfkeys.code.tex\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/utilities/pgfkeysfiltered.c\n",
      "ode.tex))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/systemlayer/pgf.cfg)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/systemlayer/pgfsys-pdftex.d\n",
      "ef\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/systemlayer/pgfsys-common-p\n",
      "df.def)))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/systemlayer/pgfsyssoftpath.\n",
      "code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/systemlayer/pgfsysprotocol.\n",
      "code.tex))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcore.code.tex\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmath.code.tex\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathcalc.code.tex\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathutil.code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathparser.code.tex\n",
      ")\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathfunctions.code.\n",
      "tex\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathfunctions.basic\n",
      ".code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathfunctions.trigo\n",
      "nometric.code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathfunctions.rando\n",
      "m.code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathfunctions.compa\n",
      "rison.code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathfunctions.base.\n",
      "code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathfunctions.round\n",
      ".code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathfunctions.misc.\n",
      "code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathfunctions.integ\n",
      "erarithmetics.code.tex)))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathfloat.code.tex)\n",
      ")\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcorepoints.co\n",
      "de.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcorepathconst\n",
      "ruct.code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcorepathusage\n",
      ".code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcorescopes.co\n",
      "de.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcoregraphicst\n",
      "ate.code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcoretransform\n",
      "ations.code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcorequick.cod\n",
      "e.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcoreobjects.c\n",
      "ode.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcorepathproce\n",
      "ssing.code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcorearrows.co\n",
      "de.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcoreshade.cod\n",
      "e.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcoreimage.cod\n",
      "e.tex\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcoreexternal.\n",
      "code.tex))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcorelayers.co\n",
      "de.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcoretranspare\n",
      "ncy.code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcorepatterns.\n",
      "code.tex)))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/modules/pgfmoduleshapes.cod\n",
      "e.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/modules/pgfmoduleplot.code.\n",
      "tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/pgf/compatibility/pgfcomp-version\n",
      "-0-65.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/pgf/compatibility/pgfcomp-version\n",
      "-1-18.sty)) (/usr/local/texlive/2018/texmf-dist/tex/latex/tools/verbatim.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/environ/environ.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/trimspaces/trimspaces.sty)))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/tcolorbox/tcbminted.code.tex\n",
      "Library (tcolorbox): 'tcbminted.code.tex' version '4.13'\n",
      "\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/tcolorbox/tcblistingscore.code.te\n",
      "x\n",
      "Library (tcolorbox): 'tcblistingscore.code.tex' version '4.13'\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/tcolorbox/tcbprocessing.code.tex\n",
      "Library (tcolorbox): 'tcbprocessing.code.tex' version '4.13'\n",
      ")))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/tcolorbox/tcbbreakable.code.tex\n",
      "Library (tcolorbox): 'tcbbreakable.code.tex' version '4.13'\n",
      ") (/usr/local/texlive/2018/texmf-dist/tex/latex/tcolorbox/tcbskins.code.tex\n",
      "Library (tcolorbox): 'tcbskins.code.tex' version '4.13'\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/pgf/frontendlayer/tikz.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/pgf/utilities/pgffor.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/pgf/utilities/pgfkeys.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/utilities/pgfkeys.code.tex)\n",
      ") (/usr/local/texlive/2018/texmf-dist/tex/latex/pgf/math/pgfmath.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmath.code.tex))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/utilities/pgffor.code.tex\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmath.code.tex)))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/frontendlayer/tikz/tikz.cod\n",
      "e.tex\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/libraries/pgflibraryplothan\n",
      "dlers.code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/modules/pgfmodulematrix.cod\n",
      "e.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/frontendlayer/tikz/librarie\n",
      "s/tikzlibrarytopaths.code.tex)))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/tcolorbox/tcbskinsjigsaw.code.tex\n",
      "Library (tcolorbox): 'tcbskinsjigsaw.code.tex' version '4.13'\n",
      ")) (/usr/local/texlive/2018/texmf-dist/tex/latex/mdframed/mdframed.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/l3packages/xparse/xparse.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/l3kernel/expl3.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/l3kernel/expl3-code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/l3kernel/l3pdfmode.def)))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/oberdiek/zref-abspage.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/oberdiek/zref-base.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/needspace/needspace.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/mdframed/md-frame-1.mdf))/anaconda2/envs/python3/bin/pygmentize\n",
      "\n",
      "(./structure_prediction.aux)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/base/ts1cmr.fd)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/psnfss/t1ppl.fd)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/context/base/mkii/supp-pdf.mkii\n",
      "[Loading MPS to PDF converter (version 2006.09.02).]\n",
      ") (/usr/local/texlive/2018/texmf-dist/tex/latex/oberdiek/epstopdf-base.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/oberdiek/grfext.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/latexconfig/epstopdf-sys.cfg))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/caption/ltcaption.sty)\n",
      "*geometry* driver: auto-detecting\n",
      "*geometry* detected driver: pdftex\n",
      "\n",
      "Package geometry Warning: The marginal notes overrun the paper.\n",
      "     Add 25.63757pt and more to the right margin.\n",
      "\n",
      "*geometry* verbose mode - [ preamble ] result:\n",
      "* driver: pdftex\n",
      "* paper: <default>\n",
      "* layout: <same size as paper>\n",
      "* layoutoffset:(h,v)=(0.0pt,0.0pt)\n",
      "* modes: \n",
      "* h-part:(L,W,R)=(43.36243pt, 527.57014pt, 43.36243pt)\n",
      "* v-part:(T,H,B)=(72.26999pt, 650.43001pt, 72.26999pt)\n",
      "* \\paperwidth=614.295pt\n",
      "* \\paperheight=794.96999pt\n",
      "* \\textwidth=527.57014pt\n",
      "* \\textheight=650.43001pt\n",
      "* \\oddsidemargin=-28.90756pt\n",
      "* \\evensidemargin=-28.90756pt\n",
      "* \\topmargin=-37.0pt\n",
      "* \\headheight=12.0pt\n",
      "* \\headsep=25.0pt\n",
      "* \\topskip=11.0pt\n",
      "* \\footskip=30.0pt\n",
      "* \\marginparwidth=59.0pt\n",
      "* \\marginparsep=10.0pt\n",
      "* \\columnsep=10.0pt\n",
      "* \\skip\\footins=10.0pt plus 4.0pt minus 2.0pt\n",
      "* \\hoffset=0.0pt\n",
      "* \\voffset=0.0pt\n",
      "* \\mag=1000\n",
      "* \\@twocolumnfalse\n",
      "* \\@twosidefalse\n",
      "* \\@mparswitchfalse\n",
      "* \\@reversemarginfalse\n",
      "* (1in=72.27pt=25.4mm, 1cm=28.453pt)\n",
      "\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/ucs/ucsencs.def)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/hyperref/nameref.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/gettitlestring.sty))\n",
      "(./structure_prediction.out) (./structure_prediction.out)\n",
      "ABD: EveryShipout initializing macros\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/psnfss/ot1ppl.fd)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/psnfss/omlzplm.fd)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/psnfss/omszplm.fd)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/psnfss/omxzplm.fd)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/psnfss/ot1zplm.fd)\n",
      "(./structure_prediction.toc) [1{/usr/local/texlive/2018/texmf-var/fonts/map/pdf\n",
      "tex/updmap/pdftex.map}]\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/base/t1cmtt.fd) [2] [3] [4 <./pro\n",
      "t1/1_hhpred_prot1.png>] [5 <./prot1/2_hhpred_ium_alignment.jpg>] [6 <./prot1/2_\n",
      "hhpred_vtw_alignment.png> <./prot1/3_tmhmm_transmembrane_prot1.png>] [7 <./prot\n",
      "1/3_signalip_signal_peptide_prot1.png>] [8 <./prot1/3_phobius_transmembrane_pro\n",
      "t1.png> <./prot1/5_pymol_align_models_ium_green_2_vtw_prot1.png>] [9 <./prot1/5\n",
      "_pymol_align_template_model_ium_prot1.png> <./prot1/5_pymol_align_template_mode\n",
      "l_vtw_prot1.png>] [10 <./prot1/3_psipred_secondary_prediction_prot1.pdf>]\n",
      "[11 <./prot1/2_blast_prot1.png> <./prot1/2_robetta_prot1.png>] [12 <./prot1/Pro\n",
      "Q_vtw_model_prot1.png> <./prot4/HHPRED_global.png>] [13 <./prot4/HHPRED_3HLZ.pn\n",
      "g> <./prot4/Phobius_prot4.png>] [14 <./prot4/SignalP_prot4.png>] [15 <./prot4/0\n",
      "2_robetta_prot4.png>] [16 <./prot4/PYMOL_model2_3hlz.png> <./prot4/ProQ_model_p\n",
      "rot4.png>] [17 <./prot4/I-TASSER results.pdf\n",
      "\n",
      "pdfTeX warning: pdflatex (file ./prot4/I-TASSER results.pdf): PDF inclusion: in\n",
      "valid other resource which is no dict (key 'ProcSets', type <array>); ignored.\n",
      ">] (./structure_prediction.aux)\n",
      "\n",
      "LaTeX Warning: Label(s) may have changed. Rerun to get cross-references right.\n",
      "\n",
      " ){/usr/local/texlive/2018/texmf-dist/fonts/enc/dvips/cm-super/cm-super-t1.enc}\n",
      "{/usr/local/texlive/2018/texmf-dist/fonts/enc/dvips/base/8r.enc}</usr/local/tex\n",
      "live/2018/texmf-dist/fonts/type1/public/cm-super/sftt1095.pfb></usr/local/texli\n",
      "ve/2018/texmf-dist/fonts/type1/urw/palatino/uplb8a.pfb></usr/local/texlive/2018\n",
      "/texmf-dist/fonts/type1/urw/palatino/uplr8a.pfb>\n",
      "Output written on structure_prediction.pdf (17 pages, 3926917 bytes).\n",
      "Transcript written on structure_prediction.log.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[NbConvertApp] Converting notebook structure_prediction.ipynb to latex\n",
      "[NbConvertApp] Writing 34461 bytes to structure_prediction.tex\n",
      "[NbConvertApp] Converting notebook structure_prediction.ipynb to html_with_toclenvs\n",
      "[NbConvertApp] Writing 306094 bytes to structure_prediction.html\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd /Users/nandoide/Desktop/uni/STRBI.practical\n",
    "jupyter nbconvert --to=latex --template=~/report.tplx structure_prediction.ipynb\n",
    "pdflatex -shell-escape structure_prediction\n",
    "jupyter nbconvert --to html_with_toclenvs structure_prediction.ipynb"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "latex_metadata": {
   "author": "Fernando Freire",
   "title": "Structure Prediction. Practical"
  },
  "nbTranslate": {
   "displayLangs": [
    "en",
    "sp"
   ],
   "hotkey": "alt-t",
   "langInMainMenu": true,
   "sourceLang": "es",
   "targetLang": "en",
   "useGoogleTranslate": true
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
