{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Worksheet PCALG"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Class AlignmentSequences\n",
    "This class implements recursively three alignment algorithms:\n",
    "1. Global alignment (Needleman-Wunsch inspired).\n",
    "2. Local alignment (Smith-Waterman inspired) \n",
    "3. Longest common substring.(the search for the longest common sequence can also be considered a type of alignment)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 443,
   "metadata": {
    "code_folding": [
     125,
     181,
     215
    ]
   },
   "outputs": [],
   "source": [
    "\"\"\"This module shows alternative recursive implementations of global sequence alignments:\n",
    "    Global alignment (Needleman-Wunsch based)\n",
    "    Local (Smith-Waterman based)\n",
    "    Finding of the longest common substring. \n",
    "Todo:\n",
    "    * Return all the solutions of the alignments. Now it only returns one solution\n",
    "    * Control of errors\n",
    "    * Implement multi-alignments\n",
    "    * Implement heuristic algorithms\n",
    "\"\"\"\n",
    "import time\n",
    "import sys\n",
    "\n",
    "MIN = -sys.maxsize - 1\n",
    "COMPAC = 100000\n",
    "\"\"\"int: Constant to compact max score.\"\"\"\n",
    "SCORE_MATCH = 2\n",
    "\"\"\"int: Default match score.\"\"\"\n",
    "SCORE_NO_MATCH = -3\n",
    "\"\"\"int: Default no match score.\"\"\"\n",
    "SCORE_GAP_INI = -10\n",
    "\"\"\"int: Default gap ini in affine gap penalty.\"\"\"\n",
    "SCORE_GAP_CONT = -2\n",
    "\"\"\"int: Default gap continuation in affine gap penalty\"\"\"\n",
    "DEFAULT_SUBST_MATRIX = {('A', 'A'): 0, ('A', 'C'): 1, ('A', 'G'): 1, ('A', 'T'): 1, ('C', 'A'): 1, ('C', 'C'): 0, ('C', 'G'): 1, ('C', 'T'): 1, ('G', 'A'): 1, ('G', 'C'): 1, ('G', 'G'): 0, ('G', 'T'): 1, ('T', 'A'): 1, ('T', 'C'): 1, ('T', 'G'): 1, ('T', 'T'): 0}\n",
    "\"\"\"dict: Default substitution matrix (for \"ACGT\" common nucleotide alphabet)\"\"\"\n",
    "\n",
    "sys.setrecursionlimit(5000)\n",
    "\n",
    "class AlignSequences:\n",
    "    \"\"\"Recursive implementation of global, local and long substring alignments methods.\n",
    "\n",
    "    Attributes:\n",
    "        sequences (list of str): Contains the wo sequences to align. The first\n",
    "            one (index 0) is the query sequence (BLAST concept) or bottom sequence on alignment prints\n",
    "            or vertical sequence in the common graphical representation of score matrix.\n",
    "        len_seq0 (int): Sequence 0 length.\n",
    "        len_seq1 (int): Sequence 1 length.\n",
    "        mode (str): Computation mode:\n",
    "            'GLOBAL'           Global Alignment\n",
    "            'LOCAL'            Local Alignment\n",
    "            'LONG_SUBSTRING'   Obtain long common substring\n",
    "        score_match (int): Score of match characters.\n",
    "        score_no_match (int): Score of no match characters.\n",
    "        score_gap_ini (int): Score of gap init.\n",
    "        score_gap_cont (int): Score of gap continuation.\n",
    "        score (int): Score of last computed alignment.\n",
    "        gaps (int): Number of gaps of the last computed alignment.\n",
    "        matches (int): Number of matches of the last computed alignment.\n",
    "        unmatches (int): Number of unmatches of the last computed alignment.\n",
    "        align_seq0 (str): Sequence 0 with the gaps necessary for the alignment.\n",
    "        align_seq1 (str): Sequence 1 with the gaps necessary for the alignment.\n",
    "        matching (str): Printable line with the align relations ('|, '.', ' ') between\n",
    "            both align_seq, necessary for printing the alignment.\n",
    "        ini_time (int): Initial time of computation, for profiling purposes\n",
    "        finish_time (int): Final time of computation, for profiling purposes\n",
    "        score_store (dict of tuple int): Store of scores, for each calculated cell with tuple (i,j,g) \n",
    "            where i is the coordinate of the bottom sequence, j the coordinate of the top sequence \n",
    "            and g has the value 1 if the cell is a gap init cell and 0 if it's a gap continuation.\n",
    "            For a explanation of calculared cell see align method.\n",
    "        matches_store (dict of tuple int): Store of the number of matches in the calculated cell\n",
    "        gaps_store (dict of tuple int): Store of the number of gaps in the calculated cell\n",
    "        max_score_index (tuple of int): Cell coordinate tuple of the cell with the maximun score\n",
    "        max_score (int): maximum computed score\n",
    "        forward_arrow (dict of str): Store of the optimal displacements accomplished at a cell \n",
    "            to guarantee an optimal score: 'v' vertical (down), 'h' horizontal (rigth), \n",
    "            'd' diagonal.\n",
    "        stacks (list of lost of str): Stacks of sequences related to principal sequences in a msa\n",
    "        \n",
    "    \"\"\"\n",
    "   \n",
    "    def __init__(self, sequences, mode=\"ALIGN\", score_match=SCORE_MATCH, score_no_match=SCORE_NO_MATCH,\\\n",
    "                 score_gap_ini=SCORE_GAP_INI, score_gap_cont=SCORE_GAP_CONT, subst_matrix={}, stacks=[[],[]]):\n",
    "        \"\"\"Init parameters of alignment\"\"\"\n",
    "        self.set_sequences(sequences)\n",
    "        self.set_stacks(stacks)\n",
    "        self.len_seq0 = len(self.sequences[0])\n",
    "        self.len_seq1 = len(self.sequences[1])\n",
    "        self.init_stores()\n",
    "        self.set_scores(score_match, score_no_match, score_gap_ini, score_gap_cont)\n",
    "        self.set_mode(mode)\n",
    "        self.score = 0\n",
    "        self.matches = 0\n",
    "        self.unmatches = 0\n",
    "        self.gaps = 0\n",
    "        self.align_seq0 = \"\"\n",
    "        self.align_seq1 = \"\"\n",
    "        self.matching = \"\"\n",
    "        self.ini_time = 0\n",
    "        self.finish_time = 0\n",
    "        self.set_subst_matrix(subst_matrix)\n",
    "       \n",
    "    def init_stores(self):\n",
    "        \"\"\"Init dictionary that store temp data of the alignment\"\"\"\n",
    "        self.score_store = {}\n",
    "        self.matches_store = {}\n",
    "        self.gaps_store = {}\n",
    "        self.max_score_index = (0, 0, 0)\n",
    "        self.max_score = 0\n",
    "        self.forward_arrow = {}\n",
    "    \n",
    "    def set_sequences(self, sequences):\n",
    "        \"\"\"Update the target sequences of the alignment\"\"\"\n",
    "        self.sequences = sequences\n",
    "        \n",
    "    def set_stacks(self, stacks=[[],[]]):\n",
    "        \"\"\"Update the stacks for msa\"\"\"\n",
    "        self.stacks = stacks\n",
    "        \n",
    "    def set_subst_matrix(self, subst_matrix={}):\n",
    "        \"\"\"Update the score matrix\"\"\"\n",
    "        self.subst_matrix = subst_matrix\n",
    "    \n",
    "    def set_scores(self, score_match=SCORE_MATCH, score_no_match=SCORE_NO_MATCH,\\\n",
    "                   score_gap_ini=SCORE_GAP_INI, score_gap_cont=SCORE_GAP_CONT):\n",
    "        \"\"\"Update the weigth scores of the alignment\"\"\"\n",
    "        self.score_match = score_match\n",
    "        self.score_no_match = score_no_match\n",
    "        self.score_gap_ini = score_gap_ini\n",
    "        self.score_gap_cont = score_gap_cont\n",
    "    \n",
    "    def set_mode(self, mode=\"ALIGN\"):\n",
    "        \"\"\"Set computation mode\"\"\"\n",
    "        self.mode = mode\n",
    "        \n",
    "    def forward_track(self, index):\n",
    "        \"\"\"Calc alignments in forward direction.\n",
    "        \n",
    "            The alignment strings are calculated from init cell (0,0) in global\n",
    "            alignments or maximum score cell in local alignments.\n",
    "            \n",
    "            In local mode it's necessary to extend the alignments (local) to the total length of \n",
    "            the sequences to show the location of the alignment, and in order to compare with \n",
    "            BioPython outputs.\n",
    "            \n",
    "            Args:\n",
    "                index (tuple of int): Cell coordinates of the starting cell\n",
    "\n",
    "            Returns:\n",
    "                string: align sequence 0 (bottom) for printing purposes\n",
    "                string: align sequence 1 (top) for printing purposes\n",
    "                tuple of int: Coordinates of the last cell\n",
    "        \"\"\"\n",
    "        ret_align_seq0, ret_align_seq1 = \"\", \"\"\n",
    "        (i, j, gap_ini) = index\n",
    "        ret_final_pos = (self.len_seq0, self.len_seq1)\n",
    "        while i < self.len_seq0 or j < self.len_seq1:\n",
    "            if self.mode == \"LOCAL\" and self.score_store[(i, j, gap_ini)] == 0: \n",
    "                ret_final_pos = (i, j)\n",
    "                break\n",
    "            arrow = self.forward_arrow[(i, j, gap_ini)]\n",
    "            if arrow == \"d\":\n",
    "                ret_align_seq0 += self.sequences[0][i]\n",
    "                ret_align_seq1 += self.sequences[1][j]\n",
    "                i, j, gap_ini = i + 1, j + 1, 1\n",
    "            elif arrow == \"h\":\n",
    "                ret_align_seq0 += \"-\"\n",
    "                ret_align_seq1 += self.sequences[1][j]\n",
    "                i, j, gap_ini = i , j + 1, 0\n",
    "            elif arrow == \"v\":\n",
    "                ret_align_seq0 += self.sequences[0][i]\n",
    "                ret_align_seq1 += \"-\"\n",
    "                i, j, gap_ini = i + 1, j, 0\n",
    "        #compute the complete align in local mode\n",
    "        if self.mode == \"LOCAL\":\n",
    "            ret_align_seq0 = self.sequences[0][0:index[0]] +\\\n",
    "                             ret_align_seq0 + self.sequences[0][ret_final_pos[0]:]\n",
    "            ret_align_seq1 = self.sequences[1][0:index[1]] +\\\n",
    "                             ret_align_seq1 + self.sequences[1][ret_final_pos[1]:]\n",
    "            diff_pos_ini = index[1] - index[0]\n",
    "            if diff_pos_ini > 0:\n",
    "                ret_align_seq0 = '-' * diff_pos_ini + ret_align_seq0\n",
    "            else:\n",
    "                ret_align_seq1 = '-' * -diff_pos_ini + ret_align_seq1\n",
    "            diff_len = len(ret_align_seq1) - len(ret_align_seq0)\n",
    "            if diff_len > 0:\n",
    "                ret_align_seq0 += '-' * diff_len\n",
    "            else:\n",
    "                ret_align_seq1 += '-' * -diff_len\n",
    "        return ret_align_seq0, ret_align_seq1, ret_final_pos\n",
    "    \n",
    "    def calc_matching(self, align_seq0, align_seq1, ini_pos=(), final_pos=()):\n",
    "        \"\"\"Calc matching string\n",
    "        \n",
    "            The matching string is the string line to print between the top and\n",
    "            bottom alignment strings. It contains the match (|), no match (.) and\n",
    "            gap ( ) indicators.\n",
    "            \n",
    "            Args:\n",
    "                align_seq0 (string): Bottom sequence\n",
    "                align_seq1 (string): Top sequence\n",
    "                ini_pos (tuple of int): Initial cell coordinates\n",
    "                final_pos (tuple of int): Final cell coordinates\n",
    "\n",
    "            Returns:\n",
    "                string: Matching string\n",
    "                \n",
    "        \"\"\"\n",
    "        count = 0\n",
    "        ret_matching = \"\"\n",
    "        diff_pos_ini = ini_pos[1] - ini_pos[0]\n",
    "        if diff_pos_ini > 0:\n",
    "            delta_pos = diff_pos_ini\n",
    "        else:\n",
    "            delta_pos = 0\n",
    "        for n, (i, j) in enumerate(zip(align_seq0, align_seq1)):\n",
    "            if self.mode == \"LOCAL\" and not (n >= ini_pos[0] + delta_pos and n < final_pos[0] + delta_pos):\n",
    "                ret_matching += ' '\n",
    "            else:\n",
    "                if i == j: ret_matching += '|'\n",
    "                elif i != j and i != '-' and j != '-': ret_matching += '.'\n",
    "                else: ret_matching += ' '\n",
    "            count += 1\n",
    "        return ret_matching\n",
    "        \n",
    "    def store(self, key, score, matches, gaps):\n",
    "        \"\"\"Store info related to a computed cell\n",
    "        The maximum score is computed having into account the number of matches, if there are\n",
    "        most than one solution. If the score are equal, the path with more matches is selected.\n",
    "             Args:\n",
    "                key (tuple of int): Cell coordinates\n",
    "                score (int): Cell score\n",
    "                matches (int): Cell matches\n",
    "                gaps (int): Cell gaps\n",
    "        \"\"\"\n",
    "        self.score_store[key] = score\n",
    "        super_score = score * COMPAC + 10 * matches\n",
    "        if super_score > self.max_score: \n",
    "            self.max_score_index = key\n",
    "            self.max_score = super_score\n",
    "        self.matches_store[key] = matches\n",
    "        self.gaps_store[key] = gaps\n",
    "    \n",
    "    def calc_score_binary(self, seq_0, seq_1, i, j):\n",
    "        \"\"\"Compute alignment scores for two sequences\n",
    "        If there are a substitution matrix (actually dictionary) defined, \n",
    "        the scores are computed from the dictionary.\n",
    "            Args:\n",
    "                seq_0 (int): Sequence 0\n",
    "                seq_1 (int): Sequence 1\n",
    "                i (int): Sequence 0 index\n",
    "                j (int): Sequence 1 index\n",
    "        \"\"\"\n",
    "        if self.subst_matrix:\n",
    "            subst_matrix_index = (seq_0[i], seq_1[j])\n",
    "            subst_matrix_index_swap = (seq_1[j], seq_0[i])\n",
    "            if subst_matrix_index in self.subst_matrix:\n",
    "                matrix_score = self.subst_matrix[subst_matrix_index]\n",
    "            elif subst_matrix_index_swap in self.subst_matrix:\n",
    "                matrix_score = self.subst_matrix[subst_matrix_index_swap]\n",
    "        # There could be gaps in msa, in this case we penalize as minimun value \n",
    "        # of score matrix (TODO)\n",
    "        # if not subst matrix we penalize it as initial gap position\n",
    "        # it's not a CLUSTALW like implementation\n",
    "        if seq_0[i] == \"-\" or  seq_1[j] == '-':\n",
    "            inc_matches = 0\n",
    "            if self.subst_matrix:\n",
    "                inc_score = self.score_gap_ini + self.score_gap_cont\n",
    "            else:\n",
    "                inc_score = self.score_gap_ini + self.score_gap_cont\n",
    "        else:\n",
    "            if seq_0[i] == seq_1[j]:\n",
    "                if self.subst_matrix:\n",
    "                    inc_score = matrix_score\n",
    "                else:\n",
    "                    inc_score = self.score_match\n",
    "                inc_matches = 1\n",
    "            else:\n",
    "                if self.subst_matrix:\n",
    "                    inc_score = matrix_score\n",
    "                else:\n",
    "                    inc_score = self.score_no_match\n",
    "                inc_matches = 0\n",
    "        return inc_score, inc_matches\n",
    "    \n",
    "    def calc_score(self, i, j):\n",
    "        \"\"\"Compute alignment scores.\n",
    "        If there are stacks associated with the sequence, we compute the score weigthing the\n",
    "        scores of the stacks. Stacks contains also the guiding sequences.\n",
    "            Args:\n",
    "                i (int): Sequence 0 index\n",
    "                j (int): Sequence 1 index\n",
    "        \"\"\"\n",
    "        if self.stacks == [[],[]]:\n",
    "            return self.calc_score_binary(self.sequences[0], self.sequences[1], i, j)\n",
    "        else:\n",
    "            computed_score = 0\n",
    "            computed_matches = 0\n",
    "            nvalues = 0\n",
    "            for seq_0 in self.stacks[0]:\n",
    "                for seq_1 in self.stacks[1]:\n",
    "                    score, matches = self.calc_score_binary(seq_0, seq_1, i, j)\n",
    "                    computed_score += score\n",
    "                    computed_matches += matches\n",
    "                    nvalues += 1\n",
    "            ret_score = computed_score /  nvalues\n",
    "            ret_matches = computed_matches / nvalues\n",
    "            return ret_score, ret_matches\n",
    "            \n",
    "    def align(self, i=0, j=0, ini_gap=1):\n",
    "        \"\"\"Recursive align of sequences\n",
    "        For each cell, which coordinates are (i, j, ini_gap), calc the maximum score path from\n",
    "        three alternative displacements:\n",
    "        \n",
    "        1) To (i + 1, j + 1, 1), that is, matching or no matching the seq0(i) and seq1(i) characters.\n",
    "        This is a diagonal displacement. \n",
    "        2) To (i, j + 1, 0), that is, seting a gap in seq0 and advance seq1. Horizontal displacement.\n",
    "        3) To (i + 1, j, 0), that is, seting a gap in seq1 and advance seq0. Vertical displacement.\n",
    "        \n",
    "        The scores of these displacements are calculated adding the score ot the target cells \n",
    "        (that are computed recursively) and the matrix, default of gap scores in each case.\n",
    "        \n",
    "        The score, matches, gaps and forward_arrow are stored at related dictionary entry based on \n",
    "        coordinates (i, j, ini_gap), all of them asociated to the maximum score of \n",
    "        the three possible paths starting from the cell, avoiding recomputation \n",
    "        of the cell if it's called from another recuersive path.\n",
    "        \n",
    "        Each cell has a third score coordinate, because a cell could be called from a cell with yet \n",
    "        has a gap (only from horizontal or vertical prior displacement) or from a cell with has a match/no match.\n",
    "        Then we need to store two scores, matches, gaps and forward_arrows related to the two possible \n",
    "        cell incarnations at coordinates (i, j, 0) and (i, j, 1).\n",
    "        \n",
    "        We store matches and gaps in order to have one aditional criterion to tiebreaker \n",
    "        if some of the scores are equal. We are using this aproach in local alignment computation. If two\n",
    "        scores are equal we choose the solucion with the greatest number of matches.\n",
    "        \n",
    "        We store the displacement directions in forward_arrow dict to compute the alignment. \n",
    "        It's posible to avoid this, using only the score information, but we have let this aproach \n",
    "        as proof of concept and for clarity in the algorithm.\n",
    "        \n",
    "        In this scenario we observe that the differences between the global, \n",
    "        local and long substring algorithms are minimal.\n",
    "        \n",
    "        Local algorithm:\n",
    "\n",
    "            Starting from the global algorithm, which would be the most general, \n",
    "            the local algorithm only changes two aspects:\n",
    "                1. Rejection of the roads with negative values of the score, equaling these values \n",
    "                to 0, that is, not letting previous alignments of poor quality affect the final result.\n",
    "                \n",
    "                2. Use as cell of beginning of the alignment the one with the highest scores. \n",
    "                In our implementation we also take into account the number of matches, \n",
    "                as we have already mentioned.\n",
    "\n",
    "        Finally, but outside the algorithm of alignment itself (at forward_track and matching methods)\n",
    "        it only remains to extend the alignment obtained to show its location within the chains to be aligned.\n",
    "\n",
    "        Search algorithm of the long common substring:\n",
    "\n",
    "            Modify the global algorithm in the following aspects:\n",
    "                1. Only computes matches between characters or gaps in one or another initial sequence.\n",
    "        \n",
    "         Args:\n",
    "                i (int): Sequence 0 index\n",
    "                j (int): Sequence 1 index\n",
    "                ini_gap (int): 1 if gap initiation, 0 if gap continuation\n",
    "        \"\"\"\n",
    "        score_diag, score_hor, score_ver = MIN, MIN, MIN\n",
    "        matches_diag, matches_hor, matches_ver = MIN, MIN, MIN\n",
    "        gaps_diag, gaps_hor, gaps_ver = MIN, MIN, MIN\n",
    "        #align and advance seq0 and seq1\n",
    "        #in long_substring mode only matches are processed\n",
    "        if i < self.len_seq0 and j < self.len_seq1 and\\\n",
    "        (self.mode != \"LONG_SUBSTRING\" or self.sequences[0][i] == self.sequences[1][j]):\n",
    "            inc_score, inc_matches = self.calc_score(i, j)\n",
    "            key = (i + 1, j + 1, 1)\n",
    "            if key in self.score_store:\n",
    "                score_diag, matches_diag, gaps_diag = \\\n",
    "                self.score_store[key] + inc_score, self.matches_store[key] + inc_matches, self.gaps_store[key]\n",
    "            else:\n",
    "                score, matches, gaps = self.align(i + 1, j + 1, 1)\n",
    "                self.store(key, score, matches, gaps)\n",
    "                score_diag, matches_diag, gaps_diag = score + inc_score, matches + inc_matches, gaps\n",
    "        #don't align and gap in seq0 (advance seq1)\n",
    "        if j < self.len_seq1:\n",
    "            gap_score = self.score_gap_cont + ini_gap * self.score_gap_ini\n",
    "            key = (i, j + 1, 0)\n",
    "            if key in self.score_store:\n",
    "                score_hor, matches_hor, gaps_hor = self.score_store[key] + gap_score,\\\n",
    "                self.matches_store[key], self.gaps_store[key] + 1\n",
    "            else:\n",
    "                score, matches, gaps = self.align(i, j + 1, 0)\n",
    "                self.store(key, score, matches, gaps)\n",
    "                score_hor, matches_hor, gaps_hor = score + gap_score, matches, gaps + 1\n",
    "        #don't align and gap in seq1 (advance seq0)\n",
    "        if i < self.len_seq0:\n",
    "            gap_score = self.score_gap_cont + ini_gap * self.score_gap_ini\n",
    "            key = (i + 1, j, 0)\n",
    "            if key in self.score_store:\n",
    "                score_ver, matches_ver, gaps_ver =\\\n",
    "                self.score_store[key] + gap_score, self.matches_store[key], self.gaps_store[key] + 1\n",
    "            else:\n",
    "                score, matches, gaps = self.align(i + 1, j, 0)\n",
    "                self.store(key, score, matches, gaps)\n",
    "                score_ver, matches_ver, gaps_ver = score + gap_score, matches, gaps + 1\n",
    "        #choose the high score path\n",
    "        matcher_diag, matcher_hor, matcher_ver = score_diag, score_hor, score_ver\n",
    "        if i < self.len_seq0 or j < self.len_seq1:\n",
    "            if self.mode == \"LOCAL\" and matcher_diag < 0 and matcher_hor < 0 and matcher_ver < 0:\n",
    "                score_diag, score_hor, score_ver = 0, 0, 0\n",
    "                #matcher_diag, matcher_hor, matcher_ver  = 0, 0, 0\n",
    "            if matcher_diag > matcher_hor and matcher_diag > matcher_ver:\n",
    "                ret_score, ret_matches, ret_gaps, ret_arrow =\\\n",
    "                score_diag, matches_diag, gaps_diag, \"d\"\n",
    "            elif matcher_hor > matcher_ver:\n",
    "                ret_score, ret_matches, ret_gaps, ret_arrow =\\\n",
    "                score_hor, matches_hor, gaps_hor, \"h\"\n",
    "            else:\n",
    "                ret_score, ret_matches, ret_gaps, ret_arrow =\\\n",
    "                score_ver, matches_ver, gaps_ver, \"v\"\n",
    "        else:\n",
    "            ret_score, ret_matches, ret_gaps, ret_arrow =\\\n",
    "            0, 0, 0, \"\"\n",
    "        self.forward_arrow[(i, j, ini_gap)] = ret_arrow\n",
    "        if i == 0 and j == 0:\n",
    "            self.store((0, 0, 1), ret_score, ret_matches, ret_gaps)\n",
    "            if self.mode in [\"GLOBAL\", \"LONG_SUBSTRING\"]: self.max_score_index = (0, 0, 1)\n",
    "            else: ret_score = self.max_score // COMPAC\n",
    "            ret_matches = self.matches_store[self.max_score_index]\n",
    "            ret_gaps = self.gaps_store[self.max_score_index]\n",
    "            \n",
    "        return ret_score, ret_matches, ret_gaps\n",
    "\n",
    "    def compute(self, mode=\"LOCAL\", silent=False):\n",
    "        \"\"\"Calc alignment\n",
    "            Args:\n",
    "                mode (str): Type of algorithm (local, global or long substring)\n",
    "                silent (bool): If true don't show alignment output\n",
    "        \"\"\"\n",
    "        self.ini_time = time.time()\n",
    "        self.init_stores()\n",
    "        self.set_mode(mode)\n",
    "        self.score, self.matches, self.gaps = self.align()\n",
    "        self.align_seq0, self.align_seq1, final_pos = self.forward_track(self.max_score_index)\n",
    "        self.matching = self.calc_matching(self.align_seq0, self.align_seq1, self.max_score_index, final_pos)\n",
    "        self.unmatches = self.matching.count('.')\n",
    "        self.gaps = self.matching.count(' ')\n",
    "        self.finish_time = time.time()\n",
    "        if not silent:\n",
    "            self.view()\n",
    "    \n",
    "    def get_len_long_common_substring(self):\n",
    "        \"\"\"Getter for the len of the common substring\n",
    "        That is equal to the number of matches of the alignment        \n",
    "        \"\"\"\n",
    "        return self.matches\n",
    "    \n",
    "    def get_long_common_substring(self):\n",
    "        \"\"\"Returns the longest common substring\n",
    "        whitout alignment (positional) information\n",
    "        \"\"\"\n",
    "        long_common_substring = \"\"\n",
    "        for (char, match_char) in zip(self.align_seq1, self.matching):\n",
    "            if match_char == '|':\n",
    "                long_common_substring += char\n",
    "        return long_common_substring\n",
    "        \n",
    "    def view(self):\n",
    "        \"\"\"Prints the alignment data\"\"\"\n",
    "        #unmatches = self.matching.count('.')\n",
    "        #gaps = self.matching.count(' ')\n",
    "        if self.matching:\n",
    "            gap_groups = self.matching.count('| ') + self.matching.count('. ') + self.matching[0].count(' ')\n",
    "        else:\n",
    "            gap_groups = 0\n",
    "        print(\" \")\n",
    "        if self.mode == \"LOCAL\":\n",
    "            print(\"### AlignSequences. Local alignment (Smith-Waterman)\")\n",
    "        elif self.mode == \"LONG_SUBSTRING\":\n",
    "            print(\"### AlignSequences. Long substring finder\")\n",
    "        else:\n",
    "            print(\"### AlignSequences. Global alignment (Needleman-Wunsch)\")\n",
    "        if self.subst_matrix:\n",
    "            print(\"\\tUsing score matrix\")\n",
    "        print(self.align_seq1)\n",
    "        print(self.matching)\n",
    "        print(self.align_seq0)\n",
    "        print(\"\\tScore:\", self.score)\n",
    "        print(\"\\tSimilarity (wo gaps):\", self.matches / (self.matches + self.unmatches))\n",
    "        print(\"\\tDistance (wo gaps):\", self.unmatches / (self.matches + self.unmatches))\n",
    "        print(\"\\tDistance:\", self.unmatches / (self.matches + self.unmatches + self.gaps))\n",
    "        print(\"\\tInit index:\", self.max_score_index)\n",
    "        print(\"\\tMatches:\", self.matches, \" Unmatches:\", self.unmatches, \" Gaps:\", self.gaps, \" Gap groups:\", gap_groups)\n",
    "        #simple scoring verification todo: apply to matrix\n",
    "        if not self.subst_matrix:\n",
    "            print(\"\\tScore verified:\", self.matches * self.score_match + unmatches * self.score_no_match \\\n",
    "                  + gaps * self.score_gap_cont + gap_groups * self.score_gap_ini)\n",
    "        print(\"\\tFinish. Execution milliseconds:\", round((self.finish_time - self.ini_time) * 1000))\n",
    "        print(\"\\tScore Dictionary Size\", len(list(self.score_store.keys())))\n",
    "        \n",
    "    def edit_distance(self, score_match=0, score_no_match=-1, score_gap_ini=0, score_gap_cont=-1):\n",
    "        \"\"\"Calculates an edit distance as requested in questions 1 and 3\n",
    "        It's the same computation as a global alignment with -1 penalities applied to\n",
    "        score_gap_cont and score_nomatch and 0 in score_match and score_gap_ini\n",
    "            \n",
    "            Args:\n",
    "                score_match (int): Score of match characters.\n",
    "                score_no_match (int): Score of no match characters.\n",
    "                score_gap_ini (int): Score of gap init.\n",
    "                score_gap_cont (int): Score of gap continuation.\n",
    "        \"\"\"\n",
    "        self.set_scores(score_match, score_no_match, score_gap_ini, score_gap_cont)\n",
    "        self.compute(\"GLOBAL\", True)\n",
    "        return abs(self.score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 1\n",
    "In this exercise you will test the Needleman-Wunsch algorithm on a short\n",
    "sequence parts of hemoglobin (PDB code 1AOW) and myoglobin 1 (PDB code 1AZI). Here\n",
    "you will align the sequence **HGSAQVKGHG** to the sequence **KTEAEMKASEDLKKHGT**.\n",
    "The two sequences are arranged in a matrix in Table 1. \n",
    "\n",
    "The sequences start at the upper right corner, the initial gap penalties are listed at each offset starting position. The gap penalty is considered as -8.\n",
    " \n",
    "\n",
    "The similarity scores Si,j from looking up the matches comes from the **BLOSUM40** table. The\n",
    "table is shown here labeled with 1-letter amino acid codes:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Response\n",
    "As I have developed my version of global alignment, very similar to Needleman-Wunsch algorithm, I'm going to use this.\n",
    "The algorithm is implemented in the class `AlignSequences` that I include on top."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 406,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " \n",
      "### AlignSequences. Global alignment (Needleman-Wunsch)\n",
      "\tUsing score matrix\n",
      "KTEAEMKASEDLKKHGT\n",
      "  ..  .| . .|.|| \n",
      "--HG--SA-Q-VKGHG-\n",
      "\tScore: -21\n",
      "\tSimilarity (wo gaps): 0.4\n",
      "\tDistance (wo gaps): 0.6\n",
      "\tDistance: 0.35294117647058826\n",
      "\tInit index: (0, 0, 1)\n",
      "\tMatches: 4  Unmatches: 6  Gaps: 7  Gap groups: 5\n",
      "\tFinish. Execution milliseconds: 4\n",
      "\tScore Dictionary Size 368\n"
     ]
    }
   ],
   "source": [
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio.SubsMat import MatrixInfo\n",
    "\n",
    "s1 = \"HGSAQVKGHG\"\n",
    "s2 = \"KTEAEMKASEDLKKHGT\"\n",
    "mode = \"GLOBAL\"\n",
    "matrix = MatrixInfo.blosum40\n",
    "\n",
    "align = AlignSequences([s1, s2])\n",
    "align.set_scores(score_match = 0, score_no_match = 0, score_gap_ini = 0, score_gap_cont = -8)\n",
    "align.set_subst_matrix(matrix)\n",
    "align.compute(mode.upper(), silent = False)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It's OK?\n",
    "I'm going to compare against **biopython**. By means of the function `test_alignment`. This function compare `AlignSequences` results with the alignments of Biopython. This software returns all the solutions of maximun score. My software not yet and so we test that my result is into the set of Biopython ones.\n",
    "I do the test with several BLOSUM matrices."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 586,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Passed test 1: scores are equal '-31'\n",
      "Passed test 1: alignments are equal '--HGS--A-Q-VKGHG-'\n",
      "Passed test 2: scores are equal '-32'\n",
      "Passed test 2: alignments are equal '--HG--SA-Q-VKGHG-'\n",
      " \n",
      "### AlignSequences. Global alignment (Needleman-Wunsch)\n",
      "\tUsing score matrix\n",
      "KTEAEMKASEDLKKHGT\n",
      "  ..  .| . .|.|| \n",
      "--HG--SA-Q-VKGHG-\n",
      "\tScore: -21\n",
      "\tSimilarity (wo gaps): 0.4\n",
      "\tDistance (wo gaps): 0.6\n",
      "\tDistance: 0.35294117647058826\n",
      "\tInit index: (0, 0, 1)\n",
      "\tMatches: 4  Unmatches: 6  Gaps: 7  Gap groups: 5\n",
      "\tFinish. Execution milliseconds: 2\n",
      "\tScore Dictionary Size 368\n",
      "\n",
      "BioPython alignment:\n",
      "KTEAEMKASEDLKKHGT\n",
      "  ..  .| . .|.|| \n",
      "--HG--SA-Q-VKGHG-\n",
      "  Score=-21\n",
      "\n",
      "Passed test 3: scores are equal '-21'\n",
      "Passed test 3: alignments are equal '--HG--SA-Q-VKGHG-'\n",
      " \n",
      "Passed All Test\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "failed = 0\n",
    "passed = 0\n",
    "launched = 0\n",
    "\n",
    "def test_alignment(number, s1, s2, verbose=False, tipus=\"local\", matrix={},\\\n",
    "                          score_match=2, score_no_match=-3, score_gap_ini=-5, score_gap_cont=-2):\n",
    "    \"\"\"Comparisons of global and local alignments between Biopython and AlignSequences implementation.\n",
    "    \n",
    "    Args:\n",
    "        number (int): The number(identifier) of the test.\n",
    "        s1 (str): Query string to align.\n",
    "        s2 (str): Subject string to align\n",
    "        verbose (bool): If True print outputs, default False\n",
    "        tipus (str) : If 'local' the alignment is local (Smith), if 'global' Waterman.\n",
    "        matrix (dict of int) : Substitution matrix\n",
    "        score_match (int): Score of character match\n",
    "        score_no_match (int): Score of character no match\n",
    "        score_gap_ini (int): Score of gap initiation\n",
    "        score_gap_cont (int): Score of gap continuation\n",
    "        \n",
    "    \"\"\"\n",
    "    global failed, passed, launched\n",
    "    try:\n",
    "        launched += 1\n",
    "        align = AlignSequences([s1, s2])\n",
    "        align.set_scores(score_match, score_no_match, score_gap_ini, score_gap_cont)\n",
    "        if matrix != {}:\n",
    "            method = getattr(pairwise2.align, tipus + 'ds')\n",
    "            alignments = method(s2, s1, matrix,\\\n",
    "                                             score_gap_ini + score_gap_cont, score_gap_cont)\n",
    "            align.set_subst_matrix(matrix)\n",
    "        else:\n",
    "            method = getattr(pairwise2.align, tipus + 'ms')\n",
    "            alignments = method(s2, s1, score_match, score_no_match,\\\n",
    "                                              score_gap_ini + score_gap_cont, score_gap_cont)\n",
    "        \n",
    "        align.compute(tipus.upper(), silent=not verbose)\n",
    "        m = re.match(r\".*Score=([-1234567890]*)\", format_alignment(*alignments[0]).replace(\"\\n\", \"\"))\n",
    "        score = int(m.group(1))\n",
    "        \n",
    "        #search AlignSequences alignment in all possibles alignments fron Biopython\n",
    "        found = False\n",
    "        for a in alignments:\n",
    "            if verbose: \n",
    "                print()\n",
    "                print(\"BioPython alignment:\")\n",
    "                print(format_alignment(*a))\n",
    "            if (align.align_seq0 == a[1]):\n",
    "                    found = True\n",
    "                    break             \n",
    "        assert(align.score == score)\n",
    "        print (\"Passed test %s: scores are equal '%s'\" % (number, align.score ))\n",
    "        assert(found)\n",
    "        print (\"Passed test %s: alignments are equal '%s'\" % (number, align.align_seq0 ))\n",
    "        passed += 1\n",
    "    \n",
    "    except AssertionError:\n",
    "        print (\"Failed test %s: alignments differ: \\nBiopython:\\n'%s'\\nScore = %s \\\n",
    "        \\nAlignSequences\\n'%s'\\nScore = %s\"\\\n",
    "               % (number, alignments[0][1], score, align.align_seq0, align.score ))\n",
    "        failed += 1\n",
    "        exit(1)\n",
    "\n",
    "prot1 = \"HGSAQVKGHG\"\n",
    "prot2 = \"KTEAEMKASEDLKKHGT\"\n",
    "\n",
    "test_alignment(1, prot1, prot2, False, \"global\", MatrixInfo.blosum80, 0, 0, 0, -8)\n",
    "test_alignment(2, prot1, prot2, False, \"global\", MatrixInfo.blosum62, 0, 0, 0, -8)\n",
    "test_alignment(3, prot1, prot2, True, \"global\", MatrixInfo.blosum40, 0, 0, 0, -8)\n",
    "\n",
    "print(\" \")\n",
    "if launched == passed: print('Passed All Test')\n",
    "else: print(\"ERROR: There are failed tests\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 2 \n",
    "Given the multiple sequence set:\n",
    "- S1: PPGVKSDCAS\n",
    "- S2: PADGVKDCAS\n",
    "- S3: PPDGKSDS\n",
    "- S4: GADGKDCCS\n",
    "- S5: GADGKDCAS\n",
    "\n",
    "Use the popular progressive alignment method to globally align the previous set of sequences.\n",
    "Generate the guide tree by neighbord-joining. Compare your outcome (alignment) with the\n",
    "one from Clustal-W."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "### Fasta files for the exercise"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 487,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting exercise2_sequences.fasta\n"
     ]
    }
   ],
   "source": [
    "%%writefile exercise2_sequences.fasta\n",
    ">s0\n",
    "PPGVKSDCAS\n",
    ">s1\n",
    "PADGVKDCAS\n",
    ">s2\n",
    "PPDGKSDS\n",
    ">s3\n",
    "GADGKDCCS\n",
    ">s4\n",
    "GADGKDCAS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "This is another file, a fasta file with globin sequences, for a further test."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 548,
   "metadata": {
    "hidden": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting globin_sequences.fasta\n"
     ]
    }
   ],
   "source": [
    "%%writefile globin_sequences.fasta\n",
    ">myoglobine_human\n",
    "MGLSDGEWQLVLNVWGKVEADIPGHGQEVLIRLFKGHPETLEKFDKFKHLKSEDEMKASEDLKKHGATVL\n",
    "TALGGILKKKGHHEAEIKPLAQSHATKHKIPVKYLEFISECIIQVLQSKHPGDFGADAQGAMNKALELFR\n",
    "KDMASNYKELGFQG\n",
    ">myoglobine_chimp\n",
    "GLSDGEWQLVLNVWGKVEADIPGHGQEVLIRLFKGHPETLEKFDKFKHLKSEDEMKASEDLKKHGATVLT\n",
    "ALGGILKKKGHHEAEIKPLAQSHATKHKIPVKYLEFISECIIQVLHSKHPGDFGADAQGAMNKALELFRK\n",
    "DMASNYKELGFQG\n",
    ">myoglobine_gorilla\n",
    "GLSDGEWQLVLNVWGKVEADISGHGQEVLIRLFKGHPETLEKFDKFKHLKSEDEMKASEDLKKHGATVLT\n",
    "ALGGILKKKGHHEAEIKPLAQSHATKHKIPVKYLEFISECIIQVLQSKHPGDFGADAQGAMNKALELFRK\n",
    "DMASNYKELGFQG\n",
    ">myoglobine_orang-oetang\n",
    "GLSDGEWQLVLNVWGKVEADIPSHGQEVLIRLFKGHPETLEKFDKFKHLKSEDEMKASEDLKKHGATVLT\n",
    "ALGGILKKKGHHEAEIKPLAQSHATKHKIPVKYLEFISESIIQVLQSKHPGDFGADAQGAMNKALELFRK\n",
    "DMASNYKELGFQG\n",
    ">myoglobine_babboon\n",
    "GLSDGEWQLVLNVWGKVEADIPSHGQEVLIRLFKGHPETLEKFDKFKHLKSEDEMKASEDLKKHGATVLT\n",
    "ALGGILKKKGHHEAEIKPLAQSHATKHKIPVKYLELISESIIQVLQSKHPGDFGADAQGAMNKALELFRN\n",
    "DMAAKYKELGFQG\n",
    ">myoglobine_pig\n",
    "MGLSDGEWQLVLNVWGKVEADVAGHGQEVLIRLFKGHPETLEKFDKFKHLKSEDEMKASEDLKKHGNTVL\n",
    "TALGGILKKKGHHEAELTPLAQSHATKHKIPVKYLEFISEAIIQVLQSKHPGDFGADAQGAMSKALELFR\n",
    "NDMAAKYKELGFQG\n",
    ">myoglobine_beaver\n",
    "GLSDGEWQLVLHVWGKVEADLAGHGQEVLIRLFKGHPETLEKFNKFKHIKSEDEMKASEDLKKHGVTVLT\n",
    "ALGGVLKKKGHHEAEIKPLAQSHATKHKIPIKYLEFISEAIIHVLQSKHPGBFGADABGAMNKALELFRK\n",
    "DIAAKYKELGFQG\n",
    ">myoglobine_rabbit\n",
    "GLSDAEWQLVLNVWGKVEADLAGHGQEVLIRLFHTHPETLEKFDKFKHLKSEDEMKASEDLKKHGNTVLT\n",
    "ALGAILKKKGHHEAEIKPLAQSHATKHKIPVKYLEFISEAIIHVLHSKHPGDFGADAQAAMSKALELFRN\n",
    "DIAAQYKELGFQG\n",
    ">myoglobine_horse\n",
    "GLSDGEWQQVLNVWGKVEADIAGHGQEVLIRLFTGHPETLEKFDKFKHLKTEAEMKASEDLKKHGTVVLT\n",
    "ALGGILKKKGHHEAELKPLAQSHATKHKIPIKYLEFISDAIIHVLHSKHPGDFGADAQGAMTKALELFRN\n",
    "DIAAKYKELGFQG\n",
    ">myoglobine_chicken\n",
    "MGLSDQEWQQVLTIWGKVEADIAGHGHEVLMRLFHDHPETLDRFDKFKGLKTPDQMKGSEDLKKHGATVL\n",
    "TQLGKILKQKGNHESELKPLAQTHATKHKIPVKYLEFISEVIIKVIAEKHAADFGADSQAAMKKALELFR\n",
    "NDMASKYKEFGFQG\n",
    ">myoglobine_alligator\n",
    "MELSDQEWKHVLDIWTKVESKLPEHGHEVIIRLLQEHPETQERFEKFKHMKTADEMKSSEKMKQHGNTVF\n",
    "TALGNILKQKGNHAEVLKPLAKSHALEHKIPVKYLEFISEIIVKVIAEKYPADFGADSQAAMRKALELFR\n",
    "NDMASKYKEFGYQG\n",
    ">myoglobine_lizard\n",
    "MGLSDEEWKKVVDIWGKVEPDLPSHGQEVIIRMFQNHPETQDRFAKFKNLKTLDEMKNSEDLKKHGTTVL\n",
    "TALGRILKQKGHHEAEIAPLAQTHANTHKIPIKYLEFICEVIVGVIAEKHSADFGADSQEAMRKALELFR\n",
    "NDMASRYKELGFQG\n",
    ">myoglobine_carp\n",
    "MHDAELVLKCWGGVEADFEGTGGEVLTRLFKQHPETQKLFPKFVGIASNELAGNAAVKAHGATVLKKLGE\n",
    "LLKARGDHAAILKPLATTHANTHKIALNNFRLITEVLVKVMAEKAGLDAGGQSALRRVMDVVIGDIDTYY\n",
    "KEIGFAG\n",
    ">myoglobine_tuna\n",
    "MADFDAVLKCWGPVEADYTTIGGLVLTRLFKEHPDTQKLFPKFAGIAQADLAGNAAISAHGATVLKKLGE\n",
    "LLKAKGSHASILKPMANSHATKHKIPINNFKLISEVLVKVMQEKAGLDAGGQTALRNVMGIIIADLEANY\n",
    "KELGFTG"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "lang": "en"
   },
   "source": [
    "### MSA Clustal progressive alignment. Methods.\n",
    "\n",
    "In these functions, what is necessary to perform a multiple alignment of sequences is developed.\n",
    "\n",
    "The method of progressive alignment based on a guide tree is used.\n",
    "\n",
    "The guide tree can be obtained in two alternative ways: **Unweighted Pair Group Method with Arithmetic Mean (UPGMA)** and **Neighbor Join (NJ)**, the same options present in _CLUSTAL_ software.\n",
    "\n",
    "The alignment has three known phases. These are the particularities of my implementation on each phase:\n",
    "\n",
    "1. Perform pairwise alignments between all the sequences involved and assign them a score.\n",
    "\n",
    " In the case of **UPGMA** we use the proportion (in percentages) between matches and matches plus no matches (without taking gaps into account). That is, we use a measure of the identity between the two sequences involved. You can also use the distance, which would be the complement to 100 of identity, but we wanted to do so to be able to compare with the information that _CLUSTAL_ throws at the beginning of his output. It does not affect the result, we simply have to look for maximum identities to build the guide tree, instead of minimum distances.\n",
    " \n",
    " In the case of **NJ**, we have chosen to use distances, computed also in percentages. Also not taking into account the number of gaps in the denominators.\n",
    " \n",
    "\n",
    "2. Build the guide tree. As I said, we can do it using UPGMA or NJ. The NJ method generates an unrooted tree. As we need a root, for purposes of the subsequent alignment, we have chosen to root it by clustering the two nodes that have no relation. We do not know if it is the method used by _CLUSTAL_ or similar programs, but it seems to work.\n",
    "\n",
    "\n",
    "3. Multiple alignment. It is, as we know, a progressive alignment following the order indicated by the guide tree. In each step we need to align two groups of sequences, of length $n >= 1$ and $m >= 1$. Each group is aligned as a whole, in the sense that the gaps entered in one of the sequences of the group must be introduced in the same positions in the rest of the sequences of their group.\n",
    "\n",
    " To assign a score to a position, the combined score of all the residuals of that position is used. To do this we produce the Cartesian product $n x m$ of all the characters of that position and calculate the average of scores:\n",
    " \n",
    " $$\\frac{\\sum_{\\begin{subarray}{l} 0\\leq i\\lt n\\\\0\\leq j\\lt m\\end{subarray}} matrix(i,j)}{n m}$$\n",
    " \n",
    "If in any of the positions we have a gap, we have chosen to penalize it as the sum of penalties assigned to the start of the gap plus gap continuation penalty. It is a criterion, _CLUSTAL_ we know that it uses another one.\n",
    "\n",
    "If we already have a pairwise development, as it was my case, it would be easy to extend it to address MSA?. The answer is affirmative. With slight modifications in the class **AlignSequences**, we have managed to address an MSA, in the following way:\n",
    "    \n",
    "1. Generalize the one-position scoring algorithm to take into account all the sequences of both groups, averaging the scores as indicated above.\n",
    "\n",
    "    \n",
    "2. Take a sequence from each group (the first) to perform a simple pairwise alignment (but with the scores calculated as indicated in 1).\n",
    "\n",
    "    \n",
    "3. Compare the sequences resulting from the pairwise alignment with their originals from each group, compute where the gap is introduced and introduce the gap at the same positions in the rest of the sequences of each group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 627,
   "metadata": {
    "code_folding": [
     380
    ]
   },
   "outputs": [],
   "source": [
    "\"\"\"This methods shows alternative implementations of multiple sequence alignments, CLUSTAL like.\n",
    "    \n",
    "TODO:\n",
    "    * Many more tests. Create a test battery.\n",
    "\n",
    "    * Achieve that the results obtained are more similar to those of CLUSTAL (if they have to be). \n",
    "    Given the lack of detailed information it will be necessary to resort \n",
    "    to the sources (in C++) of CLUSTAL.\n",
    "\n",
    "    * Allow to configure the initial alignment and the final multialignments with different parameters.\n",
    "\n",
    "    * Draw the alignments in a more standard way.\n",
    "\n",
    "    * Draw the phylogenetic trees.\n",
    "\n",
    "    * Include all new methods in AlignSequences or in another class.\n",
    "\"\"\"\n",
    "MIN_SCORE = 0\n",
    "def readFasta(file):\n",
    "    \"\"\" \n",
    "    Reads all sequences of a FASTA file \n",
    "    Args:\n",
    "        file (str): name of the imput FASTA file\n",
    "    Returns:\n",
    "        dict of str, str: sequences readed\n",
    "    \"\"\"  \n",
    "    ret_seqs = {}\n",
    "    seq = \"\"\n",
    "    key_found = False\n",
    "    with open(file, 'r') as f:\n",
    "        key = \"\"\n",
    "        for line in f:\n",
    "            line = line.replace('\\n', '')\n",
    "            if len(line) > 0:\n",
    "                if line[0] == \">\":\n",
    "                    if key_found:\n",
    "                        ret_seqs[key] = seq\n",
    "                    key_found = True\n",
    "                    key = line[1:].split(\" \")[0]\n",
    "                    seq = \"\"\n",
    "                elif key_found:\n",
    "                    seq += line\n",
    "    if key_found:\n",
    "        ret_seqs[key] = seq\n",
    "    return ret_seqs\n",
    "\n",
    "def pairwise_align(s1, s2, matrix, mode, score_gap_ini=0, score_gap_cont=-8,\\\n",
    "                   score_match=3, score_no_match=-2):\n",
    "    \"\"\"\n",
    "    Performs initial pairwise alignments against the class AlignSequences \n",
    "    returning the %identity.\n",
    "    \n",
    "    Args:\n",
    "        s1 (str): First sequence to compare.\n",
    "        s2 (str): Second sequence to compare.\n",
    "        matrix (dict of tuples of int): Substitution matrix, Biopython format.\n",
    "        mode (str): Computation mode:\n",
    "            'GLOBAL'           Global Alignment\n",
    "            'LOCAL'            Local Alignment\n",
    "            'LONG_SUBSTRING'   Obtain long common substring\n",
    "        score_gap_ini (int): Score of gap init.\n",
    "        score_gap_cont (int): Score of gap continuation.\n",
    "        score_match (int): Score of match characters (used if no matrix informed)\n",
    "        score_no_match (int): Score of no match characters (used if no matrix informed)\n",
    "    \n",
    "    Returns:\n",
    "        (int): % identity between sequences\n",
    "    \"\"\"\n",
    "    align = AlignSequences([s1, s2])\n",
    "    align.set_scores(score_match, score_no_match, score_gap_ini, score_gap_cont)\n",
    "    align.set_subst_matrix(matrix)\n",
    "    align.compute(mode.upper(), silent = True)\n",
    "    return round((align.matches + 1) * 100 / (align.matches + align.unmatches + 2))\n",
    "\n",
    "def pairwise_align_distance(s1, s2, matrix, mode, score_gap_ini=0, score_gap_cont=-8):\n",
    "    \"\"\"\n",
    "    Performs initial pairwise alignments against the class AlignSequences \n",
    "    returning the distance between 0 and 100.\n",
    "    \n",
    "    Args:\n",
    "        s1 (str): First sequence to compare.\n",
    "        s2 (str): Second sequence to compare.\n",
    "        matrix (dict of tuples of int): Substitution matrix, Biopython format.\n",
    "        mode (str): Computation mode:\n",
    "            'GLOBAL'           Global Alignment\n",
    "            'LOCAL'            Local Alignment\n",
    "            'LONG_SUBSTRING'   Obtain long common substring\n",
    "        score_gap_ini (int): Score of gap init.\n",
    "        score_gap_cont (int): Score of gap continuation.\n",
    "        score_match (int): Score of match characters (used if no matrix informed)\n",
    "        score_no_match (int): Score of no match characters (used if no matrix informed)\n",
    "    \n",
    "    Returns:\n",
    "        (int): distance between sequences\n",
    "    \"\"\"\n",
    "    align = AlignSequences([s1, s2])\n",
    "    align.set_scores(0, 0, score_gap_ini, score_gap_cont)\n",
    "    align.set_subst_matrix(matrix)\n",
    "    align.compute(mode.upper(), silent = True)\n",
    "    identity = round((align.matches + 1) * 100 / (align.matches + align.unmatches + 2))\n",
    "    return 100 - identity\n",
    "\n",
    "def guide_tree_UPGMA(sequences, matrix={}, mode=\"GLOBAL\", score_gap_ini=0, score_gap_cont=-8,\\\n",
    "                    score_match=3, score_no_match=-2):\n",
    "    \"\"\"\n",
    "    Performs initial pairwise alignments against the class AlignSequences \n",
    "    returning the guide_tree derived from UPGMA method.\n",
    "    \n",
    "    Args:\n",
    "        s1 (str): First sequence to compare.\n",
    "        s2 (str): Second sequence to compare.\n",
    "        matrix (dict of tuples of int): Substitution matrix, Biopython format.\n",
    "        mode (str): Computation mode:\n",
    "            'GLOBAL'           Global Alignment\n",
    "            'LOCAL'            Local Alignment\n",
    "            'LONG_SUBSTRING'   Obtain long common substring\n",
    "        score_gap_ini (int): Score of gap init.\n",
    "        score_gap_cont (int): Score of gap continuation.\n",
    "        score_match (int): Score of match characters (used if no matrix informed)\n",
    "        score_no_match (int): Score of no match characters (used if no matrix informed)\n",
    "    \n",
    "    Returns:\n",
    "        (list of 3-tuples of int): guide three, the third position of the tuple contains the root\n",
    "                            of the other two nodes.\n",
    "        (dict of int, boolean = True): contains all nodes \n",
    "        \n",
    "    \"\"\"\n",
    "    tree = {} #initial tree\n",
    "    guide_tree = [] #guided tree, pairs to align in sequence\n",
    "    max_score = MIN_SCORE\n",
    "    max_score_position = ()\n",
    "    for i in range(0, len(sequences)):\n",
    "        for j in range(0 , i):\n",
    "            if (i,j) not in tree:\n",
    "                score = pairwise_align(sequences[i], sequences[j], matrix, mode,\\\n",
    "                                score_gap_ini, score_gap_cont, score_match, score_no_match)\n",
    "                tree[(i,j)] = score\n",
    "                if score >= max_score:\n",
    "                    max_score = score\n",
    "                    max_score_position = (i,j)\n",
    "    \n",
    "    print(tree)\n",
    "    len_tree = len(sequences)\n",
    "    guide_tree_nodes = {}\n",
    "    # Generate guide tree. At every step we compute another row averaging the\n",
    "    # most closer rows and removing all their row coordinates from the tree\n",
    "    while len(tree) > 0:\n",
    "        (imax, jmax) = max_score_position\n",
    "        guide_tree.append((imax, jmax, len_tree))\n",
    "        guide_tree_nodes[imax] = True\n",
    "        guide_tree_nodes[jmax] = True\n",
    "        guide_tree_nodes[len_tree] = False\n",
    "        \n",
    "        # Average scores from i,j rows into new row in new_row_pos\n",
    "        for j in range(0, len_tree):\n",
    "            if j in [imax, jmax]:\n",
    "                continue\n",
    "            nscores = 0.0;\n",
    "            for coordinate in [(imax, j), (j, imax), (jmax, j), (j, jmax)]:\n",
    "                if coordinate in tree:\n",
    "                    score = tree[coordinate]\n",
    "                    nscores += 1\n",
    "                    if (len_tree, j) not in tree:\n",
    "                        tree[(len_tree, j)] = score\n",
    "                    else:\n",
    "                        tree[(len_tree, j)] += score\n",
    "            if nscores > 0:\n",
    "                tree[(len_tree, j)] = tree[(len_tree, j)] / nscores\n",
    "            \n",
    "        # Tree cleaning and calc max score\n",
    "        max_score = MIN_SCORE\n",
    "        max_score_position = ()\n",
    "        for i in range(0, len_tree + 1):\n",
    "            for j in range(0, len_tree + 1):\n",
    "                if (i,j) in tree:\n",
    "                    if i == imax or i == jmax or j == imax or j == jmax:\n",
    "                        del(tree[(i,j)])\n",
    "                    else:\n",
    "                        if tree[(i,j)] >= max_score:\n",
    "                            max_score =  tree[(i,j)]\n",
    "                            max_score_position = (i,j)\n",
    "        \n",
    "        len_tree += 1\n",
    "\n",
    "    return guide_tree, guide_tree_nodes\n",
    "\n",
    "def guide_tree_UPGMA_from_fasta(file, matrix={}, mode=\"GLOBAL\", score_gap_ini=0, score_gap_cont=-8,\\\n",
    "                              score_match=3, score_no_match=-2):\n",
    "    \"\"\"\n",
    "    Performs initial pairwise alignments against the class AlignSequences \n",
    "    The secuences are read from a fasta file\n",
    "    returning the guide_tree derived from UPGMA method.\n",
    "    \n",
    "    Args:\n",
    "        file (str): Name of the FASTA file.\n",
    "        matrix (dict of tuples of int): Substitution matrix, Biopython format.\n",
    "        mode (str): Computation mode:\n",
    "            'GLOBAL'           Global Alignment\n",
    "            'LOCAL'            Local Alignment\n",
    "            'LONG_SUBSTRING'   Obtain long common substring\n",
    "        score_gap_ini (int): Score of gap init.\n",
    "        score_gap_cont (int): Score of gap continuation.\n",
    "        score_match (int): Score of match characters (used if no matrix informed)\n",
    "        score_no_match (int): Score of no match characters (used if no matrix informed)\n",
    "    \n",
    "    Returns:\n",
    "        list of 3-tuples of int: guide three, the third position of the tuple contains the root\n",
    "                            of the other two nodes.\n",
    "        dict of int, boolean = True: contains all nodes \n",
    "        list of str: sequences involved in the guide tree\n",
    "        \n",
    "    \"\"\"\n",
    "    sequences = list(readFasta(file).values())\n",
    "    sequence_names = list(readFasta(file).keys())\n",
    "    print(sequence_names)\n",
    "    guide_tree, guide_tree_nodes = guide_tree_UPGMA(sequences, matrix, mode,\\\n",
    "                                                   score_gap_ini, score_gap_cont,\\\n",
    "                                                   score_match, score_no_match)\n",
    "    return guide_tree, guide_tree_nodes, sequences\n",
    "\n",
    "def q(i, j, nseq, n, dmatrix):\n",
    "    \"\"\"\n",
    "    NJ method: calculate element of intermediate Q matrix.\n",
    "    \"\"\"\n",
    "    d = (nseq - 2) * dmatrix[(i,j)]\n",
    "    for k in range(0, n):\n",
    "        if (i,k) in dmatrix:\n",
    "            d -= dmatrix[(i,k)]\n",
    "        if (j,k) in dmatrix:\n",
    "            d -= dmatrix[(j,k)]\n",
    "    return d\n",
    "\n",
    "def calc_qmatrix(nseq, n, dmatrix):\n",
    "    \"\"\"\n",
    "    NJ method: calculate intermediate Q matrix.\n",
    "    \"\"\"\n",
    "    qmatrix = {}\n",
    "    for (i,j) in dmatrix:\n",
    "        qmatrix[(i,j)] = q(i, j, nseq, n, dmatrix)\n",
    "    return qmatrix\n",
    "\n",
    "def smallest_q(qmatrix):\n",
    "    \"\"\"\n",
    "    NJ method: returns the coordinates of the minimun score in intermediate Q matrix.\n",
    "    \"\"\"\n",
    "    sq = ()\n",
    "    min_sq = - MIN\n",
    "    for key in qmatrix.keys():\n",
    "        if qmatrix[key] <  min_sq:\n",
    "            min_sq = qmatrix[key]\n",
    "            sq = key\n",
    "    return sq\n",
    "\n",
    "def djoin(joined_pair, nseq, n, dmatrix):\n",
    "    \"\"\"\n",
    "    NJ method: returns distances of joined nodes to the rooted node, so, it returns the branch lengths\n",
    "    \"\"\"\n",
    "    (i, j) = joined_pair\n",
    "    d_i_1 = dmatrix[(i,j)] / 2.0\n",
    "    d_i_2 = 0\n",
    "    for k in range(0, n):\n",
    "        if (i,k) in dmatrix:\n",
    "            d_i_2 += dmatrix[(i,k)]\n",
    "        if (j,k) in dmatrix:\n",
    "            d_i_2 -= dmatrix[(j,k)]\n",
    "    d_i = d_i_1 - d_i_2 / (2*(nseq - 2))\n",
    "    d_j = dmatrix[(i,j)] - d_i\n",
    "    return d_i, d_j\n",
    "\n",
    "def dnjoin(k, joined_pair, dmatrix):\n",
    "    \"\"\"\n",
    "    NJ method: returns distance of sequence k to the new nodetht routes the joined_pair.\n",
    "    The distance is the mean of the distances ftom k to each of nodes joined.\n",
    "    \"\"\"\n",
    "    (i, j) = joined_pair\n",
    "    d_k = 0\n",
    "    if (i,k) in dmatrix:\n",
    "        d_k += dmatrix[(i,k)]\n",
    "    if (j,k) in dmatrix:\n",
    "        d_k += dmatrix[(j,k)]\n",
    "    d_k = (d_k - dmatrix[(i,j)]) / 2.0\n",
    "    return d_k\n",
    "\n",
    "def recalc_dmatrix(joined_pair, n, dmatrix):\n",
    "    \"\"\"\n",
    "    NJ method: recalc distance matrix taking into account the joined pair\n",
    "    \"\"\"\n",
    "    (i, j) = joined_pair\n",
    "    # Recalculate distances\n",
    "    for k in range(0, n):\n",
    "        if (i,k) in dmatrix and (j,k) in dmatrix:\n",
    "            dmatrix[(n + 1, k)] = dnjoin(k, joined_pair, dmatrix)\n",
    "            dmatrix[(k, n + 1)] = dmatrix[(n + 1, k)]\n",
    "    # Remove joined rows from dmatrix\n",
    "    for k in range(0, n + 1):\n",
    "        for l in range(0, n + 1):\n",
    "            if k == i or k == j or l == i or l== j:\n",
    "                if (k, l) in dmatrix:\n",
    "                    del(dmatrix[(k, l)])   \n",
    "    return\n",
    "\n",
    "def guide_tree_NJ(sequences, matrix={}, mode=\"GLOBAL\", score_gap_ini=0, score_gap_cont=-8,\\\n",
    "                              score_match=3, score_no_match=-2):\n",
    "    \"\"\"\n",
    "    Performs initial pairwise alignments against the class AlignSequences \n",
    "    returning the guide_tree derived from NJ method.\n",
    "    \n",
    "    Args:\n",
    "        s1 (str): First sequence to compare.\n",
    "        s2 (str): Second sequence to compare.\n",
    "        matrix (dict of tuples of int): Substitution matrix, Biopython format.\n",
    "        mode (str): Computation mode:\n",
    "            'GLOBAL'           Global Alignment\n",
    "            'LOCAL'            Local Alignment\n",
    "            'LONG_SUBSTRING'   Obtain long common substring\n",
    "        score_gap_ini (int): Score of gap init.\n",
    "        score_gap_cont (int): Score of gap continuation.\n",
    "        score_match (int): Score of match characters (used if no matrix informed)\n",
    "        score_no_match (int): Score of no match characters (used if no matrix informed)\n",
    "    \n",
    "    Returns:\n",
    "        (list of 3-tuples of int): guide three, the third position of the tuple contains the root\n",
    "                            of the other two nodes.\n",
    "        (dict of int, boolean = True): contains all nodes \n",
    "        \n",
    "    \"\"\"\n",
    "    dmatrix = {} #initial distance matrix\n",
    "    n = len(sequences)\n",
    "    for i in range(0, n):\n",
    "        for j in range(0 , i):\n",
    "            if (i,j) not in dmatrix:\n",
    "                distance = pairwise_align_distance(sequences[i], sequences[j], matrix, mode,\\\n",
    "                                score_gap_ini, score_gap_cont)\n",
    "                dmatrix[(i,j)] = distance\n",
    "                dmatrix[(j,i)] = distance\n",
    "    nseq = n\n",
    "    new_nodes = n - 2\n",
    "    guide_tree = [] #guided tree, pairs to align in sequence\n",
    "    guide_tree_nodes = {} #guided tree rooted nodes to complete\n",
    "    for i in range(0, n):\n",
    "        guide_tree_nodes[i] = False\n",
    "    while new_nodes > 0:\n",
    "        qmatrix = calc_qmatrix(nseq, n, dmatrix)\n",
    "        (joined_i, joined_j) = smallest_q(qmatrix)\n",
    "        #print(\"JOIN:\", (joined_i, joined_j))\n",
    "        guide_tree.append((joined_i, joined_j, n + 1))\n",
    "        guide_tree_nodes[joined_i] = True\n",
    "        guide_tree_nodes[joined_j] = True\n",
    "        guide_tree_nodes[n + 1] = False\n",
    "        recalc_dmatrix((joined_i, joined_j), n, dmatrix)\n",
    "        n += 1\n",
    "        nseq -= 1\n",
    "        new_nodes -= 1\n",
    "    # Root the tree\n",
    "    print(\"DMATRIX:\", dmatrix)\n",
    "    rooting_tuple = []\n",
    "    for node in guide_tree_nodes:\n",
    "        if not guide_tree_nodes[node]:\n",
    "            rooting_tuple.append(node)\n",
    "    rooting_tuple.append(n + 1)\n",
    "    guide_tree_nodes[n + 1] = True\n",
    "    print(\"Rooting tuple:\", rooting_tuple)\n",
    "    if len(rooting_tuple) == 3:\n",
    "        guide_tree.append(rooting_tuple)\n",
    "    assert len(rooting_tuple) == 3\n",
    "    return guide_tree, guide_tree_nodes\n",
    "\n",
    "def guide_tree_NJ_from_fasta(file, matrix={}, mode=\"GLOBAL\", score_gap_ini=0, score_gap_cont=-8,\\\n",
    "                              score_match=3, score_no_match=-2):\n",
    "    \"\"\"\n",
    "    Performs initial pairwise alignments against the class AlignSequences \n",
    "    The secuences are read from a fasta file\n",
    "    returning the guide_tree derived from NJ method.\n",
    "    \n",
    "    Args:\n",
    "        file (str): Name of the FASTA file.\n",
    "        matrix (dict of tuples of int): Substitution matrix, Biopython format.\n",
    "        mode (str): Computation mode:\n",
    "            'GLOBAL'           Global Alignment\n",
    "            'LOCAL'            Local Alignment\n",
    "            'LONG_SUBSTRING'   Obtain long common substring\n",
    "        score_gap_ini (int): Score of gap init.\n",
    "        score_gap_cont (int): Score of gap continuation.\n",
    "        score_match (int): Score of match characters (used if no matrix informed)\n",
    "        score_no_match (int): Score of no match characters (used if no matrix informed)\n",
    "    \n",
    "    Returns:\n",
    "        list of 3-tuples of int: guide three, the third position of the tuple contains the root\n",
    "                            of the other two nodes.\n",
    "        dict of int, boolean = True: contains all nodes \n",
    "        list of str: sequences involved in the guide tree\n",
    "        \n",
    "    \"\"\"\n",
    "    sequences = list(readFasta(file).values())\n",
    "    sequence_names = list(readFasta(file).keys())\n",
    "    print(sequence_names)\n",
    "    guide_tree, guide_tree_nodes = guide_tree_NJ(sequences, matrix, mode, score_gap_ini, score_gap_cont,\\\n",
    "                                               score_match, score_no_match)\n",
    "    return guide_tree, guide_tree_nodes, sequences\n",
    "\n",
    "def gapeator(a, a_gapped, b_stack):\n",
    "    \"\"\"\n",
    "    Introduces gaps in all the sequences of b_stack taking into account the positions\n",
    "    and the gaps introduced in sequence a to obtain sequence a_gapped\n",
    "    Args:\n",
    "        a (str): template sequence not gapped\n",
    "        a_gapped (str): template sequence gapped\n",
    "        b_stack (list of str): stack of b sequences ungapped\n",
    "    \n",
    "    Returns:\n",
    "        list of str: stack b gapped as a does\n",
    "    \"\"\"\n",
    "    ini_a_gapped = a_gapped\n",
    "    b_gapped_stack = []\n",
    "    for b in b_stack:\n",
    "        b_gapped = \"\"\n",
    "        a_gapped = ini_a_gapped\n",
    "        for k, (i, j) in enumerate(zip(a, b)):\n",
    "            index = a_gapped.index(i)\n",
    "            a_gapped = a_gapped[index + 1:]\n",
    "            #print(\"a_gapped\", a_gapped )\n",
    "            #print(i, j, index)\n",
    "            b_gapped += \"-\" * index + j\n",
    "            #print(\"b_gapped\", b_gapped )\n",
    "        b_gapped += b[k + 1:]\n",
    "        b_gapped_stack.append(b_gapped)\n",
    "    return b_gapped_stack\n",
    "\n",
    "def pairwise_align_msa_step(stack_0, stack_1, matrix, mode=\"GLOBAL\", score_match=0,\\\n",
    "                            score_no_match=0, score_gap_ini=0, score_gap_cont=-8):\n",
    "    \"\"\"\n",
    "    Performs msa alignment of sequence stack 0 and 1.\n",
    "    \n",
    "    Args:\n",
    "        stack_0 (list str): First stack of sequences to align.\n",
    "        stack_1 (list str): Secong stack of sequences to align.\n",
    "        matrix (dict of tuples of int): Substitution matrix, Biopython format.\n",
    "        mode (str): Computation mode:\n",
    "            'GLOBAL'           Global Alignment\n",
    "            'LOCAL'            Local Alignment\n",
    "            'LONG_SUBSTRING'   Obtain long common substring\n",
    "        score_match (int): Score of match characters (used if no matrix informed)\n",
    "        score_no_match (int): Score of no match characters (used if no matrix informed)\n",
    "        score_gap_ini (int): Score of gap init.\n",
    "        score_gap_cont (int): Score of gap continuation.\n",
    "       \n",
    "    Returns:\n",
    "        list of str: stack_0 gapped (with the gaps necessary for the alignment) \n",
    "        list of str: stack_0 gapped (with the gaps necessary for the alignment) \n",
    "    \"\"\"\n",
    "    align = AlignSequences([stack_0[0], stack_1[0]])\n",
    "    align.set_scores(score_match, score_no_match, score_gap_ini, score_gap_cont)\n",
    "    align.set_subst_matrix(matrix)\n",
    "    align.set_stacks([stack_0, stack_1])\n",
    "    align.compute(mode.upper(), silent = True)\n",
    "    # align_seq0 align_seq1 are the seq0 and seq1 alignments\n",
    "    # we need to deduce the rest of alignments. \n",
    "    # what we do is perform the same gap insertions, if any, as the first sequence of the stacks\n",
    "    stack_0_gapped = gapeator(stack_0[0], align.align_seq0, stack_0)\n",
    "    stack_1_gapped = gapeator(stack_1[0], align.align_seq1, stack_1)\n",
    "    return stack_0_gapped, stack_1_gapped\n",
    "\n",
    "def do_msa_from_fasta(file, method=\"NJ\", mode=\"GLOBAL\", matrix={}, score_gap_ini =-10,\\\n",
    "             score_gap_cont=-5, score_match=3, score_no_match=-2, verbose=False):\n",
    "    \"\"\"\n",
    "    Performs MSA alignments CLUSTAL like.\n",
    "    \n",
    "    Args:\n",
    "        file (str): Name of the FASTA file.\n",
    "        method (str): NJ neighbor join / UPGMA\n",
    "        matrix (dict of tuples of int): Substitution matrix, Biopython format.\n",
    "        mode (str): Computation mode:\n",
    "            'GLOBAL'           Global Alignment\n",
    "            'LOCAL'            Local Alignment\n",
    "            'LONG_SUBSTRING'   Obtain long common substring\n",
    "        score_gap_ini (int): Score of gap init.\n",
    "        score_gap_cont (int): Score of gap continuation.\n",
    "        score_match (int): Score of match characters (used if no matrix informed)\n",
    "        score_no_match (int): Score of no match characters (used if no matrix informed)\n",
    "        verbose (bool): If True prints verbose info\n",
    "    \n",
    "    Returns:\n",
    "        list of 3-tuples of int: guide three, the third position of the tuple contains the root\n",
    "                            of the other two nodes. \n",
    "        list of str: alignments\n",
    "        \n",
    "    \"\"\"\n",
    "    if method == \"NJ\":\n",
    "        guide_tree, guide_tree_nodes, sequences = guide_tree_NJ_from_fasta(file, matrix, \\\n",
    "                                                           mode, score_gap_ini, score_gap_cont,\\\n",
    "                                                           score_match, score_no_match) \n",
    "    else: #UPGMA\n",
    "        guide_tree, guide_tree_nodes, sequences = guide_tree_UPGMA_from_fasta(file, matrix, mode,\\\n",
    "                                                                              score_gap_ini, score_gap_cont,\\\n",
    "                                                                             score_match, score_no_match)\n",
    "    sequences_store = {}\n",
    "    verbose = False\n",
    "    for i in guide_tree_nodes.keys():\n",
    "        if i < len(sequences):\n",
    "            sequences_store[i] = [sequences[i]]\n",
    "            \n",
    "    if verbose: print(sequences_store)\n",
    "    for (i ,j ,k) in guide_tree:\n",
    "        stack_i = sequences_store[i]\n",
    "        stack_j = sequences_store[j]\n",
    "        if verbose: print(\"Stack i\", i, stack_i)\n",
    "        if verbose: print(\"Stack j\", j, stack_j)\n",
    "        stack_0, stack_1 = pairwise_align_msa_step(stack_i, stack_j, matrix)\n",
    "        sequences_store[k] = []\n",
    "        if verbose: print(\"==================\")\n",
    "        for s in stack_0:\n",
    "            if verbose: print(s)\n",
    "            sequences_store[k].append(s)\n",
    "        for s in stack_1:\n",
    "            if verbose: print(s)\n",
    "            sequences_store[k].append(s)\n",
    "        if verbose: print(\"==================\")\n",
    "        if verbose: print(\"New stack:\", k, sequences_store[k])\n",
    "    alignment = sequences_store[k]\n",
    "    return guide_tree, alignment\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Obtain MSA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 625,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['s0', 's1', 's2', 's3', 's4']\n",
      "{(1, 0): 50, (2, 0): 70, (2, 1): 60, (3, 0): 45, (3, 1): 73, (3, 2): 50, (4, 0): 55, (4, 1): 82, (4, 2): 50, (4, 3): 82}\n",
      "# Guide Tree UPGMA: [(4, 3, 5), (5, 1, 6), (2, 0, 7), (7, 6, 8)]\n",
      "# Alignment:\n",
      "PPDGKSD--S\n",
      "\n",
      "PPGVKSDCAS\n",
      "\n",
      "GADG-KDCAS\n",
      "\n",
      "GADG-KDCCS\n",
      "\n",
      "PADGVKDCAS\n",
      "\n",
      "\n",
      "['s0', 's1', 's2', 's3', 's4']\n",
      "DMATRIX: {(8, 3): 12.5, (3, 8): 12.5}\n",
      "Rooting tuple: [3, 8, 9]\n",
      "\n",
      "# Guide Tree NJ: [(2, 0, 6), (6, 1, 7), (7, 4, 8), [3, 8, 9]]\n",
      "# Alignment:\n",
      "GADGK-DCCS\n",
      "\n",
      "PPDGKSD--S\n",
      "\n",
      "PPGVKSDCAS\n",
      "\n",
      "PADGVKDCAS\n",
      "\n",
      "GADGK-DCAS\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "guide_tree, align = do_msa_from_fasta(\"exercise2_sequences.fasta\", method = \"UPGMA\", mode = \"GLOBAL\",\\\n",
    "                 matrix = MatrixInfo.blosum62, score_gap_ini = -10, \n",
    "                 score_gap_cont = -5, score_match = 3, score_no_match = -2, verbose = False)\n",
    "\n",
    "print(\"# Guide Tree UPGMA:\", guide_tree)\n",
    "print(\"# Alignment:\")\n",
    "for s in align:\n",
    "    print(s)\n",
    "    print()\n",
    "print()\n",
    "\n",
    "guide_tree, align = do_msa_from_fasta(\"exercise2_sequences.fasta\", method = \"NJ\", mode = \"GLOBAL\",\\\n",
    "                 matrix = MatrixInfo.blosum62, score_gap_ini = -10, \n",
    "                 score_gap_cont = -5, verbose = False)\n",
    "\n",
    "print()\n",
    "print(\"# Guide Tree NJ:\", guide_tree)\n",
    "print(\"# Alignment:\")\n",
    "for s in align:\n",
    "    print(s)\n",
    "    print()\n",
    "print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Obtain MSA mioglobines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 626,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['myoglobine_human', 'myoglobine_chimp', 'myoglobine_gorilla', 'myoglobine_orang-oetang', 'myoglobine_babboon', 'myoglobine_pig', 'myoglobine_beaver', 'myoglobine_rabbit', 'myoglobine_horse', 'myoglobine_chicken', 'myoglobine_alligator', 'myoglobine_lizard', 'myoglobine_carp', 'myoglobine_tuna']\n",
      "{(1, 0): 99, (2, 0): 99, (2, 1): 98, (3, 0): 98, (3, 1): 97, (3, 2): 97, (4, 0): 95, (4, 1): 95, (4, 2): 95, (4, 3): 97, (5, 0): 93, (5, 1): 92, (5, 2): 93, (5, 3): 92, (5, 4): 94, (6, 0): 90, (6, 1): 89, (6, 2): 90, (6, 3): 89, (6, 4): 89, (6, 5): 90, (7, 0): 89, (7, 1): 90, (7, 2): 89, (7, 3): 88, (7, 4): 89, (7, 5): 92, (7, 6): 88, (8, 0): 88, (8, 1): 88, (8, 2): 88, (8, 3): 87, (8, 4): 88, (8, 5): 90, (8, 6): 88, (8, 7): 89, (9, 0): 76, (9, 1): 76, (9, 2): 76, (9, 3): 75, (9, 4): 75, (9, 5): 76, (9, 6): 71, (9, 7): 76, (9, 8): 75, (10, 0): 65, (10, 1): 65, (10, 2): 65, (10, 3): 65, (10, 4): 65, (10, 5): 66, (10, 6): 62, (10, 7): 66, (10, 8): 63, (10, 9): 72, (11, 0): 72, (11, 1): 72, (11, 2): 71, (11, 3): 72, (11, 4): 72, (11, 5): 71, (11, 6): 68, (11, 7): 71, (11, 8): 70, (11, 9): 74, (11, 10): 71, (12, 0): 46, (12, 1): 45, (12, 2): 45, (12, 3): 44, (12, 4): 45, (12, 5): 45, (12, 6): 45, (12, 7): 45, (12, 8): 42, (12, 9): 47, (12, 10): 43, (12, 11): 46, (13, 0): 45, (13, 1): 44, (13, 2): 45, (13, 3): 45, (13, 4): 46, (13, 5): 44, (13, 6): 45, (13, 7): 43, (13, 8): 43, (13, 9): 47, (13, 10): 43, (13, 11): 42, (13, 12): 70}\n",
      "\n",
      "# Guide Tree UPGMA: [(2, 0, 14), (14, 1, 15), (15, 3, 16), (16, 4, 17), (17, 5, 18), (18, 7, 19), (19, 8, 20), (20, 6, 21), (11, 9, 22), (22, 10, 23), (13, 12, 24), (23, 21, 25), (25, 24, 26)]\n",
      "# Alignment:\n",
      "MGLSDEEWKKVVDIWGKVEPDLPSHGQEVIIRMFQNHPETQDRFAKFKNLKTLDEMKNSEDLKKHGTTVLTALGRILKQKGHHEAEIAPLAQTHANTHKIPIKYLEFICEVIVGVIAEKHSADFGADSQEAMRKALELFRNDMASRYKELGFQG\n",
      "\n",
      "MGLSDQEWQQVLTIWGKVEADIAGHGHEVLMRLFHDHPETLDRFDKFKGLKTPDQMKGSEDLKKHGATVLTQLGKILKQKGNHESELKPLAQTHATKHKIPVKYLEFISEVIIKVIAEKHAADFGADSQAAMKKALELFRNDMASKYKEFGFQG\n",
      "\n",
      "MELSDQEWKHVLDIWTKVESKLPEHGHEVIIRLLQEHPETQERFEKFKHMKTADEMKSSEKMKQHGNTVFTALGNILKQKGNHAEVLKPLAKSHALEHKIPVKYLEFISEIIVKVIAEKYPADFGADSQAAMRKALELFRNDMASKYKEFGYQG\n",
      "\n",
      "-GLSDGEWQLVLNVWGKVEADISGHGQEVLIRLFKGHPETLEKFDKFKHLKSEDEMKASEDLKKHGATVLTALGGILKKKGHHEAEIKPLAQSHATKHKIPVKYLEFISECIIQVLQSKHPGDFGADAQGAMNKALELFRKDMASNYKELGFQG\n",
      "\n",
      "MGLSDGEWQLVLNVWGKVEADIPGHGQEVLIRLFKGHPETLEKFDKFKHLKSEDEMKASEDLKKHGATVLTALGGILKKKGHHEAEIKPLAQSHATKHKIPVKYLEFISECIIQVLQSKHPGDFGADAQGAMNKALELFRKDMASNYKELGFQG\n",
      "\n",
      "-GLSDGEWQLVLNVWGKVEADIPGHGQEVLIRLFKGHPETLEKFDKFKHLKSEDEMKASEDLKKHGATVLTALGGILKKKGHHEAEIKPLAQSHATKHKIPVKYLEFISECIIQVLHSKHPGDFGADAQGAMNKALELFRKDMASNYKELGFQG\n",
      "\n",
      "-GLSDGEWQLVLNVWGKVEADIPSHGQEVLIRLFKGHPETLEKFDKFKHLKSEDEMKASEDLKKHGATVLTALGGILKKKGHHEAEIKPLAQSHATKHKIPVKYLEFISESIIQVLQSKHPGDFGADAQGAMNKALELFRKDMASNYKELGFQG\n",
      "\n",
      "-GLSDGEWQLVLNVWGKVEADIPSHGQEVLIRLFKGHPETLEKFDKFKHLKSEDEMKASEDLKKHGATVLTALGGILKKKGHHEAEIKPLAQSHATKHKIPVKYLELISESIIQVLQSKHPGDFGADAQGAMNKALELFRNDMAAKYKELGFQG\n",
      "\n",
      "MGLSDGEWQLVLNVWGKVEADVAGHGQEVLIRLFKGHPETLEKFDKFKHLKSEDEMKASEDLKKHGNTVLTALGGILKKKGHHEAELTPLAQSHATKHKIPVKYLEFISEAIIQVLQSKHPGDFGADAQGAMSKALELFRNDMAAKYKELGFQG\n",
      "\n",
      "-GLSDAEWQLVLNVWGKVEADLAGHGQEVLIRLFHTHPETLEKFDKFKHLKSEDEMKASEDLKKHGNTVLTALGAILKKKGHHEAEIKPLAQSHATKHKIPVKYLEFISEAIIHVLHSKHPGDFGADAQAAMSKALELFRNDIAAQYKELGFQG\n",
      "\n",
      "-GLSDGEWQQVLNVWGKVEADIAGHGQEVLIRLFTGHPETLEKFDKFKHLKTEAEMKASEDLKKHGTVVLTALGGILKKKGHHEAELKPLAQSHATKHKIPIKYLEFISDAIIHVLHSKHPGDFGADAQGAMTKALELFRNDIAAKYKELGFQG\n",
      "\n",
      "-GLSDGEWQLVLHVWGKVEADLAGHGQEVLIRLFKGHPETLEKFNKFKHIKSEDEMKASEDLKKHGVTVLTALGGVLKKKGHHEAEIKPLAQSHATKHKIPIKYLEFISEAIIHVLQSKHPGBFGADABGAMNKALELFRKDIAAKYKELGFQG\n",
      "\n",
      "--MADFD--AVLKCWGPVEADYTTIGGLVLTRLFKEHPDTQKLFPKFAGI-AQADLAGNAAISAHGATVLKKLGELLKAKGSHASILKPMANSHATKHKIPINNFKLISEVLVKVMQEK-AG-LDAGGQTALRNVMGIIIADLEANYKELGFTG\n",
      "\n",
      "--MHDAE--LVLKCWGGVEADFEGTGGEVLTRLFKQHPETQKLFPKFVGI-ASNELAGNAAVKAHGATVLKKLGELLKARGDHAAILKPLATTHANTHKIALNNFRLITEVLVKVMAEK-AG-LDAGGQSALRRVMDVVIGDIDTYYKEIGFAG\n",
      "\n",
      "\n",
      "['myoglobine_human', 'myoglobine_chimp', 'myoglobine_gorilla', 'myoglobine_orang-oetang', 'myoglobine_babboon', 'myoglobine_pig', 'myoglobine_beaver', 'myoglobine_rabbit', 'myoglobine_horse', 'myoglobine_chicken', 'myoglobine_alligator', 'myoglobine_lizard', 'myoglobine_carp', 'myoglobine_tuna']\n",
      "DMATRIX: {(26, 1): 1.042236328125, (1, 26): 1.042236328125}\n",
      "Rooting tuple: [1, 26, 27]\n",
      "\n",
      "# Guide Tree NJ: [(13, 12, 15), (15, 10, 16), (16, 11, 17), (17, 9, 18), (18, 7, 19), (19, 8, 20), (20, 5, 21), (21, 6, 22), (22, 4, 23), (23, 3, 24), (24, 0, 25), (25, 2, 26), [1, 26, 27]]\n",
      "# Alignment:\n",
      "-GLSDGEWQLVLNVWGKVEADIPGHGQEVLIRLFKGHPETLEKFDKFKHLKSEDEMKASEDLKKHGATVLTALGGILKKKGHHEAEIKPLAQSHATKHKIPVKYLEFISECIIQVLHSKHPGDFGADAQGAMNKALELFRKDMASNYKELGFQG\n",
      "\n",
      "M--ADFD-A-VLKCWGPVEADYTTIGGLVLTRLFKEHPDTQKLFPKFAGIAQAD-LAGNAAISAHGATVLKKLGELLKAKGSHASILKPMANSHATKHKIPINNFKLISEVLVKVMQEK--AGLDAGGQTALRNVMGIIIADLEANYKELGFTG\n",
      "\n",
      "M--HDAE-L-VLKCWGGVEADFEGTGGEVLTRLFKQHPETQKLFPKFVGIASNE-LAGNAAVKAHGATVLKKLGELLKARGDHAAILKPLATTHANTHKIALNNFRLITEVLVKVMAEK--AGLDAGGQSALRRVMDVVIGDIDTYYKEIGFAG\n",
      "\n",
      "MELSDQEWKHVLDIWTKVESKLPEHGHEVIIRLLQEHPETQERFEKFKHMKTADEMKSSEKMKQHGNTVFTALGNILKQKGNHAEVLKPLAKSHALEHKIPVKYLEFISEIIVKVIAEKYPADFGADSQAAMRKALELFRNDMASKYKEFGYQG\n",
      "\n",
      "MGLSDEEWKKVVDIWGKVEPDLPSHGQEVIIRMFQNHPETQDRFAKFKNLKTLDEMKNSEDLKKHGTTVLTALGRILKQKGHHEAEIAPLAQTHANTHKIPIKYLEFICEVIVGVIAEKHSADFGADSQEAMRKALELFRNDMASRYKELGFQG\n",
      "\n",
      "MGLSDQEWQQVLTIWGKVEADIAGHGHEVLMRLFHDHPETLDRFDKFKGLKTPDQMKGSEDLKKHGATVLTQLGKILKQKGNHESELKPLAQTHATKHKIPVKYLEFISEVIIKVIAEKHAADFGADSQAAMKKALELFRNDMASKYKEFGFQG\n",
      "\n",
      "-GLSDAEWQLVLNVWGKVEADLAGHGQEVLIRLFHTHPETLEKFDKFKHLKSEDEMKASEDLKKHGNTVLTALGAILKKKGHHEAEIKPLAQSHATKHKIPVKYLEFISEAIIHVLHSKHPGDFGADAQAAMSKALELFRNDIAAQYKELGFQG\n",
      "\n",
      "-GLSDGEWQQVLNVWGKVEADIAGHGQEVLIRLFTGHPETLEKFDKFKHLKTEAEMKASEDLKKHGTVVLTALGGILKKKGHHEAELKPLAQSHATKHKIPIKYLEFISDAIIHVLHSKHPGDFGADAQGAMTKALELFRNDIAAKYKELGFQG\n",
      "\n",
      "MGLSDGEWQLVLNVWGKVEADVAGHGQEVLIRLFKGHPETLEKFDKFKHLKSEDEMKASEDLKKHGNTVLTALGGILKKKGHHEAELTPLAQSHATKHKIPVKYLEFISEAIIQVLQSKHPGDFGADAQGAMSKALELFRNDMAAKYKELGFQG\n",
      "\n",
      "-GLSDGEWQLVLHVWGKVEADLAGHGQEVLIRLFKGHPETLEKFNKFKHIKSEDEMKASEDLKKHGVTVLTALGGVLKKKGHHEAEIKPLAQSHATKHKIPIKYLEFISEAIIHVLQSKHPGBFGADABGAMNKALELFRKDIAAKYKELGFQG\n",
      "\n",
      "-GLSDGEWQLVLNVWGKVEADIPSHGQEVLIRLFKGHPETLEKFDKFKHLKSEDEMKASEDLKKHGATVLTALGGILKKKGHHEAEIKPLAQSHATKHKIPVKYLELISESIIQVLQSKHPGDFGADAQGAMNKALELFRNDMAAKYKELGFQG\n",
      "\n",
      "-GLSDGEWQLVLNVWGKVEADIPSHGQEVLIRLFKGHPETLEKFDKFKHLKSEDEMKASEDLKKHGATVLTALGGILKKKGHHEAEIKPLAQSHATKHKIPVKYLEFISESIIQVLQSKHPGDFGADAQGAMNKALELFRKDMASNYKELGFQG\n",
      "\n",
      "MGLSDGEWQLVLNVWGKVEADIPGHGQEVLIRLFKGHPETLEKFDKFKHLKSEDEMKASEDLKKHGATVLTALGGILKKKGHHEAEIKPLAQSHATKHKIPVKYLEFISECIIQVLQSKHPGDFGADAQGAMNKALELFRKDMASNYKELGFQG\n",
      "\n",
      "-GLSDGEWQLVLNVWGKVEADISGHGQEVLIRLFKGHPETLEKFDKFKHLKSEDEMKASEDLKKHGATVLTALGGILKKKGHHEAEIKPLAQSHATKHKIPVKYLEFISECIIQVLQSKHPGDFGADAQGAMNKALELFRKDMASNYKELGFQG\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "guide_tree, align = do_msa_from_fasta(\"globin_sequences.fasta\", method = \"UPGMA\", mode = \"GLOBAL\",\\\n",
    "                 matrix = MatrixInfo.blosum95, score_gap_ini = -10, \n",
    "                 score_gap_cont = -5, verbose = False)\n",
    "print()\n",
    "print(\"# Guide Tree UPGMA:\", guide_tree)\n",
    "print(\"# Alignment:\")\n",
    "for s in align:\n",
    "    print(s)\n",
    "    print()\n",
    "print()\n",
    "\n",
    "guide_tree, align = do_msa_from_fasta(\"globin_sequences.fasta\", method = \"NJ\", mode = \"GLOBAL\",\\\n",
    "                 matrix = MatrixInfo.blosum95, score_gap_ini = -10, \n",
    "                 score_gap_cont = -5, verbose = False)\n",
    "print()\n",
    "print(\"# Guide Tree NJ:\", guide_tree)\n",
    "print(\"# Alignment:\")\n",
    "for s in align:\n",
    "    print(s)\n",
    "    print()\n",
    "print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Verification against CLUSTALW"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 618,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      " CLUSTAL 2.1 Multiple Sequence Alignments\n",
      "\n",
      "\n",
      "Sequence type explicitly set to Protein\n",
      "Sequence format is Pearson\n",
      "Sequence 1: s0            10 aa\n",
      "Sequence 2: s1            10 aa\n",
      "Sequence 3: s2             8 aa\n",
      "Sequence 4: s3             9 aa\n",
      "Sequence 5: s4             9 aa\n",
      "Start of Pairwise alignments\n",
      "Aligning...\n",
      "\n",
      "Sequences (1:2) Aligned. Score:  50\n",
      "Sequences (1:3) Aligned. Score:  62\n",
      "Sequences (1:4) Aligned. Score:  33\n",
      "Sequences (1:5) Aligned. Score:  44\n",
      "Sequences (2:3) Aligned. Score:  50\n",
      "Sequences (2:4) Aligned. Score:  44\n",
      "Sequences (2:5) Aligned. Score:  55\n",
      "Sequences (3:4) Aligned. Score:  37\n",
      "Sequences (3:5) Aligned. Score:  37\n",
      "Sequences (4:5) Aligned. Score:  88\n",
      "Guide tree file created:   [exercise2_sequences.dnd]\n",
      "\n",
      "Start of Multiple Alignment\n",
      "\n",
      "Aligning...\n",
      "Group 1: Sequences:   2      Score:64\n",
      "Group 2: Sequences:   2      Score:35\n",
      "Group 3: Sequences:   3      Score:34\n",
      "Group 4: Sequences:   5      Score:22\n",
      "Alignment Score 182\n",
      "firstres = 1 lastres = 10\n",
      "FASTA file created!\n",
      "\n",
      "Fasta-Alignment file created    [align.fasta]\n",
      "\n",
      ">s3\n",
      "GADGK-DCCS\n",
      ">s4\n",
      "GADGK-DCAS\n",
      ">s0\n",
      "PPGVKSDCAS\n",
      ">s2\n",
      "PPDGKSDS--\n",
      ">s1\n",
      "PADGVKDCAS\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "~/clustalw2  -OUTPUTTREE=phylip -NEGATIVE -INFILE=exercise2_sequences.fasta -OUTORDER=ALIGN -STATS=align.log -TREE -ALIGN -CLUSTERING=UPGMA -OUTFILE=align.fasta -OUTPUT=FASTA -MATRIX=BLOSUM -TYPE=PROTEIN -PWGAPOPEN=10 -PWGAPEXT=5 \n",
    "#cat globin_sequences.fasta\n",
    "cat align.fasta\n",
    "#cat align.log"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "latex_metadata": {
     "hidden": "true",
     "lexer": "bash"
    }
   },
   "source": [
    "## Outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 620,
   "metadata": {
    "code_folding": [],
    "latex_metadata": {
     "hidden": "true",
     "lexer": "bash"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[NbConvertApp] Converting notebook worksheet.ipynb to latex\n",
      "[NbConvertApp] Writing 83217 bytes to worksheet.tex\n",
      "[NbConvertApp] Converting notebook worksheet.ipynb to html_with_toclenvs\n",
      "[NbConvertApp] Writing 491128 bytes to worksheet.html\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "#cd /Users/nandoide/Desktop/uni/STRBI.practical\n",
    "jupyter nbconvert --to=latex --template=~/report.tplx worksheet.ipynb 1> /dev/null\n",
    "pdflatex -shell-escape worksheet 1> /dev/null\n",
    "jupyter nbconvert --to html_with_toclenvs worksheet.ipynb 1> /dev/null"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "nbTranslate": {
   "displayLangs": [
    "en",
    "sp"
   ],
   "hotkey": "alt-t",
   "langInMainMenu": true,
   "sourceLang": "es",
   "targetLang": "en",
   "useGoogleTranslate": true
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
