{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Worksheet PCALG"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "## Class AlignmentSequences\n",
    "This class implements recursively three alignment algorithms:\n",
    "1. Global alignment (Needleman-Wunsch inspired).\n",
    "2. Local alignment (Smith-Waterman inspired) \n",
    "3. Longest common substring.(the search for the longest common sequence can also be considered a type of alignment)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "\"\"\"This module shows alternative recursive implementations of global sequence alignments:\n",
    "    Global alignment (Needleman-Wunsch based)\n",
    "    Local (Smith-Waterman based)\n",
    "    Finding of the longest common substring. \n",
    "Todo:\n",
    "    * Return all the solutions of the alignments. Now it only returns one solution\n",
    "    * Control of errors\n",
    "    * Implement multi-alignments\n",
    "    * Implement heuristic algorithms\n",
    "\"\"\"\n",
    "import time\n",
    "import sys\n",
    "\n",
    "MIN = -sys.maxsize - 1\n",
    "COMPAC = 100000\n",
    "\"\"\"int: Constant to compact max score.\"\"\"\n",
    "SCORE_MATCH = 2\n",
    "\"\"\"int: Default match score.\"\"\"\n",
    "SCORE_NO_MATCH = -3\n",
    "\"\"\"int: Default no match score.\"\"\"\n",
    "SCORE_GAP_INI = -10\n",
    "\"\"\"int: Default gap ini in affine gap penalty.\"\"\"\n",
    "SCORE_GAP_CONT = -2\n",
    "\"\"\"int: Default gap continuation in affine gap penalty\"\"\"\n",
    "DEFAULT_SUBST_MATRIX = {('A', 'A'): 0, ('A', 'C'): 1, ('A', 'G'): 1, ('A', 'T'): 1, ('C', 'A'): 1, ('C', 'C'): 0, ('C', 'G'): 1, ('C', 'T'): 1, ('G', 'A'): 1, ('G', 'C'): 1, ('G', 'G'): 0, ('G', 'T'): 1, ('T', 'A'): 1, ('T', 'C'): 1, ('T', 'G'): 1, ('T', 'T'): 0}\n",
    "\"\"\"dict: Default substitution matrix (for \"ACGT\" common nucleotide alphabet)\"\"\"\n",
    "\n",
    "sys.setrecursionlimit(5000)\n",
    "\n",
    "class AlignSequences:\n",
    "    \"\"\"Recursive implementation of global, local and long substring alignments methods.\n",
    "\n",
    "    Attributes:\n",
    "        sequences (list of str): Contains the wo sequences to align. The first\n",
    "            one (index 0) is the query sequence (BLAST concept) or bottom sequence on alignment prints\n",
    "            or vertical sequence in the common graphical representation of score matrix.\n",
    "        len_seq0 (int): Sequence 0 length.\n",
    "        len_seq1 (int): Sequence 1 length.\n",
    "        mode (str): Computation mode:\n",
    "            'GLOBAL'           Global Alignment\n",
    "            'LOCAL'            Local Alignment\n",
    "            'LONG_SUBSTRING'   Obtain long common substring\n",
    "        score_match (int): Score of match characters.\n",
    "        score_no_match (int): Score of no match characters.\n",
    "        score_gap_ini (int): Score of gap init.\n",
    "        score_gap_cont (int): Score of gap continuation.\n",
    "        score (int): Score of last computed alignment.\n",
    "        gaps (int): Number of gaps of the last computed alignment.\n",
    "        align_seq0 (str): Sequence 0 with the gaps necessary for the alignment.\n",
    "        align_seq1 (str): Sequence 1 with the gaps necessary for the alignment.\n",
    "        matching (str): Printable line with the align relations ('|, '.', ' ') between\n",
    "            both align_seq, necessary for printing the alignment.\n",
    "        ini_time (int): Initial time of computation, for profiling purposes\n",
    "        finish_time (int): Final time of computation, for profiling purposes\n",
    "        score_store (dict of tuple int): Store of scores, for each calculated cell with tuple (i,j,g) \n",
    "            where i is the coordinate of the bottom sequence, j the coordinate of the top sequence \n",
    "            and g has the value 1 if the cell is a gap init cell and 0 if it's a gap continuation.\n",
    "            For a explanation of calculared cell see align method.\n",
    "        matches_store (dict of tuple int): Store of the number of matches in the calculated cell\n",
    "        gaps_store (dict of tuple int): Store of the number of gaps in the calculated cell\n",
    "        max_score_index (tuple of int): Cell coordinate tuple of the cell with the maximun score\n",
    "        max_score (int): maximum computed score\n",
    "        forward_arrow (dict of str): Store of the optimal displacements accomplished at a cell \n",
    "            to guarantee an optimal score: 'v' vertical (down), 'h' horizontal (rigth), \n",
    "            'd' diagonal.\n",
    "        \n",
    "    \"\"\"\n",
    "   \n",
    "    def __init__(self, sequences, mode=\"ALIGN\", score_match=SCORE_MATCH, score_no_match=SCORE_NO_MATCH,\\\n",
    "                 score_gap_ini=SCORE_GAP_INI, score_gap_cont=SCORE_GAP_CONT, subst_matrix={}):\n",
    "        \"\"\"Init parameters of alignment\"\"\"\n",
    "        self.set_sequences(sequences)\n",
    "        self.len_seq0 = len(self.sequences[0])\n",
    "        self.len_seq1 = len(self.sequences[1])\n",
    "        self.init_stores()\n",
    "        self.set_scores(score_match, score_no_match, score_gap_ini, score_gap_cont)\n",
    "        self.set_mode(mode)\n",
    "        self.score = 0\n",
    "        self.matches = 0\n",
    "        self.gaps = 0\n",
    "        self.align_seq0 = \"\"\n",
    "        self.align_seq1 = \"\"\n",
    "        self.matching = \"\"\n",
    "        self.ini_time = 0\n",
    "        self.finish_time = 0\n",
    "        self.set_subst_matrix(subst_matrix)\n",
    "       \n",
    "    def init_stores(self):\n",
    "        \"\"\"Init dictionary that store temp data of the alignment\"\"\"\n",
    "        self.score_store = {}\n",
    "        self.matches_store = {}\n",
    "        self.gaps_store = {}\n",
    "        self.max_score_index = (0, 0, 0)\n",
    "        self.max_score = 0\n",
    "        self.forward_arrow = {}\n",
    "    \n",
    "    def set_sequences(self, sequences):\n",
    "        \"\"\"Update the target sequences of the alignment\"\"\"\n",
    "        self.sequences = sequences\n",
    "        \n",
    "    def set_subst_matrix(self, subst_matrix={}):\n",
    "        \"\"\"Update the score matrix\"\"\"\n",
    "        self.subst_matrix = subst_matrix\n",
    "    \n",
    "    def set_scores(self, score_match=SCORE_MATCH, score_no_match=SCORE_NO_MATCH,\\\n",
    "                   score_gap_ini=SCORE_GAP_INI, score_gap_cont=SCORE_GAP_CONT):\n",
    "        \"\"\"Update the weigth scores of the alignment\"\"\"\n",
    "        self.score_match = score_match\n",
    "        self.score_no_match = score_no_match\n",
    "        self.score_gap_ini = score_gap_ini\n",
    "        self.score_gap_cont = score_gap_cont\n",
    "    \n",
    "    def set_mode(self, mode=\"ALIGN\"):\n",
    "        \"\"\"Set computation mode\"\"\"\n",
    "        self.mode = mode\n",
    "        \n",
    "    def forward_track(self, index):\n",
    "        \"\"\"Calc alignments in forward direction.\n",
    "        \n",
    "            The alignment strings are calculated from init cell (0,0) in global\n",
    "            alignments or maximum score cell in local alignments.\n",
    "            \n",
    "            In local mode it's necessary to extend the alignments (local) to the total length of \n",
    "            the sequences to show the location of the alignment, and in order to compare with \n",
    "            BioPython outputs.\n",
    "            \n",
    "            Args:\n",
    "                index (tuple of int): Cell coordinates of the starting cell\n",
    "\n",
    "            Returns:\n",
    "                string: align sequence 0 (bottom) for printing purposes\n",
    "                string: align sequence 1 (top) for printing purposes\n",
    "                tuple of int: Coordinates of the last cell\n",
    "        \"\"\"\n",
    "        ret_align_seq0, ret_align_seq1 = \"\", \"\"\n",
    "        (i, j, gap_ini) = index\n",
    "        ret_final_pos = (self.len_seq0, self.len_seq1)\n",
    "        while i < self.len_seq0 or j < self.len_seq1:\n",
    "            if self.mode == \"LOCAL\" and self.score_store[(i, j, gap_ini)] == 0: \n",
    "                ret_final_pos = (i, j)\n",
    "                break\n",
    "            arrow = self.forward_arrow[(i, j, gap_ini)]\n",
    "            if arrow == \"d\":\n",
    "                ret_align_seq0 += self.sequences[0][i]\n",
    "                ret_align_seq1 += self.sequences[1][j]\n",
    "                i, j, gap_ini = i + 1, j + 1, 1\n",
    "            elif arrow == \"h\":\n",
    "                ret_align_seq0 += \"-\"\n",
    "                ret_align_seq1 += self.sequences[1][j]\n",
    "                i, j, gap_ini = i , j + 1, 0\n",
    "            elif arrow == \"v\":\n",
    "                ret_align_seq0 += self.sequences[0][i]\n",
    "                ret_align_seq1 += \"-\"\n",
    "                i, j, gap_ini = i + 1, j, 0\n",
    "        #compute the complete align in local mode\n",
    "        if self.mode == \"LOCAL\":\n",
    "            ret_align_seq0 = self.sequences[0][0:index[0]] +\\\n",
    "                             ret_align_seq0 + self.sequences[0][ret_final_pos[0]:]\n",
    "            ret_align_seq1 = self.sequences[1][0:index[1]] +\\\n",
    "                             ret_align_seq1 + self.sequences[1][ret_final_pos[1]:]\n",
    "            diff_pos_ini = index[1] - index[0]\n",
    "            if diff_pos_ini > 0:\n",
    "                ret_align_seq0 = '-' * diff_pos_ini + ret_align_seq0\n",
    "            else:\n",
    "                ret_align_seq1 = '-' * -diff_pos_ini + ret_align_seq1\n",
    "            diff_len = len(ret_align_seq1) - len(ret_align_seq0)\n",
    "            if diff_len > 0:\n",
    "                ret_align_seq0 += '-' * diff_len\n",
    "            else:\n",
    "                ret_align_seq1 += '-' * -diff_len\n",
    "        return ret_align_seq0, ret_align_seq1, ret_final_pos\n",
    "    \n",
    "    def calc_matching(self, align_seq0, align_seq1, ini_pos=(), final_pos=()):\n",
    "        \"\"\"Calc matching string\n",
    "        \n",
    "            The matching string is the string line to print between the top and\n",
    "            bottom alignment strings. It contains the match (|), no match (.) and\n",
    "            gap ( ) indicators.\n",
    "            \n",
    "            Args:\n",
    "                align_seq0 (string): Bottom sequence\n",
    "                align_seq1 (string): Top sequence\n",
    "                ini_pos (tuple of int): Initial cell coordinates\n",
    "                final_pos (tuple of int): Final cell coordinates\n",
    "\n",
    "            Returns:\n",
    "                string: Matching string\n",
    "                \n",
    "        \"\"\"\n",
    "        count = 0\n",
    "        ret_matching = \"\"\n",
    "        diff_pos_ini = ini_pos[1] - ini_pos[0]\n",
    "        if diff_pos_ini > 0:\n",
    "            delta_pos = diff_pos_ini\n",
    "        else:\n",
    "            delta_pos = 0\n",
    "        for n, (i, j) in enumerate(zip(align_seq0, align_seq1)):\n",
    "            if self.mode == \"LOCAL\" and not (n >= ini_pos[0] + delta_pos and n < final_pos[0] + delta_pos):\n",
    "                ret_matching += ' '\n",
    "            else:\n",
    "                if i == j: ret_matching += '|'\n",
    "                elif i != j and i != '-' and j != '-': ret_matching += '.'\n",
    "                else: ret_matching += ' '\n",
    "            count += 1\n",
    "        return ret_matching\n",
    "        \n",
    "    def store(self, key, score, matches, gaps):\n",
    "        \"\"\"Store info related to a computed cell\n",
    "        The maximum score is computed having into account the number of matches, if there are\n",
    "        most than one solution. If the score are equal, the path with more matches is selected.\n",
    "             Args:\n",
    "                key (tuple of int): Cell coordinates\n",
    "                score (int): Cell score\n",
    "                matches (int): Cell matches\n",
    "                gaps (int): Cell gaps\n",
    "        \"\"\"\n",
    "        self.score_store[key] = score\n",
    "        super_score = score * COMPAC + 10 * matches\n",
    "        if super_score > self.max_score: \n",
    "            self.max_score_index = key\n",
    "            self.max_score = super_score\n",
    "        self.matches_store[key] = matches\n",
    "        self.gaps_store[key] = gaps\n",
    "        \n",
    "    def calc_score(self, i, j):\n",
    "        \"\"\"Compute alignment scores.\n",
    "        If there are a substitution matrix (actually dictionary) defined, \n",
    "        the scores are computed from the dictionary.\n",
    "            Args:\n",
    "                i (int): Sequence 0 index\n",
    "                j (int): Sequence 1 index\n",
    "        \"\"\"\n",
    "        if self.subst_matrix:\n",
    "            subst_matrix_index = (self.sequences[0][i], self.sequences[1][j])\n",
    "            subst_matrix_index_swap = (self.sequences[1][j], self.sequences[0][i])\n",
    "            if subst_matrix_index in self.subst_matrix:\n",
    "                matrix_score = self.subst_matrix[subst_matrix_index]\n",
    "            elif subst_matrix_index_swap in self.subst_matrix:\n",
    "                matrix_score = self.subst_matrix[subst_matrix_index_swap]\n",
    "        if self.sequences[0][i] == self.sequences[1][j]:\n",
    "            if self.subst_matrix:\n",
    "                inc_score = matrix_score\n",
    "            else:\n",
    "                inc_score = self.score_match\n",
    "            inc_matches = 1\n",
    "        else:\n",
    "            if self.subst_matrix:\n",
    "                inc_score = matrix_score\n",
    "            else:\n",
    "                inc_score = self.score_no_match\n",
    "            inc_matches = 0\n",
    "        return inc_score, inc_matches\n",
    "    \n",
    "    def align(self, i=0, j=0, ini_gap=1):\n",
    "        \"\"\"Recursive align of sequences\n",
    "        For each cell, which coordinates are (i, j, ini_gap), calc the maximum score path from\n",
    "        three alternative displacements:\n",
    "        \n",
    "        1) To (i + 1, j + 1, 1), that is, matching or no matching the seq0(i) and seq1(i) characters.\n",
    "        This is a diagonal displacement. \n",
    "        2) To (i, j + 1, 0), that is, seting a gap in seq0 and advance seq1. Horizontal displacement.\n",
    "        3) To (i + 1, j, 0), that is, seting a gap in seq1 and advance seq0. Vertical displacement.\n",
    "        \n",
    "        The scores of these displacements are calculated adding the score ot the target cells \n",
    "        (that are computed recursively) and the matrix, default of gap scores in each case.\n",
    "        \n",
    "        The score, matches, gaps and forward_arrow are stored at related dictionary entry based on \n",
    "        coordinates (i, j, ini_gap), all of them asociated to the maximum score of \n",
    "        the three possible paths starting from the cell, avoiding recomputation \n",
    "        of the cell if it's called from another recuersive path.\n",
    "        \n",
    "        Each cell has a third score coordinate, because a cell could be called from a cell with yet \n",
    "        has a gap (only from horizontal or vertical prior displacement) or from a cell with has a match/no match.\n",
    "        Then we need to store two scores, matches, gaps and forward_arrows related to the two possible \n",
    "        cell incarnations at coordinates (i, j, 0) and (i, j, 1).\n",
    "        \n",
    "        We store matches and gaps in order to have one aditional criterion to tiebreaker \n",
    "        if some of the scores are equal. We are using this aproach in local alignment computation. If two\n",
    "        scores are equal we choose the solucion with the greatest number of matches.\n",
    "        \n",
    "        We store the displacement directions in forward_arrow dict to compute the alignment. \n",
    "        It's posible to avoid this, using only the score information, but we have let this aproach \n",
    "        as proof of concept and for clarity in the algorithm.\n",
    "        \n",
    "        In this scenario we observe that the differences between the global, \n",
    "        local and long substring algorithms are minimal.\n",
    "        \n",
    "        Local algorithm:\n",
    "\n",
    "            Starting from the global algorithm, which would be the most general, \n",
    "            the local algorithm only changes two aspects:\n",
    "                1. Rejection of the roads with negative values of the score, equaling these values \n",
    "                to 0, that is, not letting previous alignments of poor quality affect the final result.\n",
    "                \n",
    "                2. Use as cell of beginning of the alignment the one with the highest scores. \n",
    "                In our implementation we also take into account the number of matches, \n",
    "                as we have already mentioned.\n",
    "\n",
    "        Finally, but outside the algorithm of alignment itself (at forward_track and matching methods)\n",
    "        it only remains to extend the alignment obtained to show its location within the chains to be aligned.\n",
    "\n",
    "        Search algorithm of the long common substring:\n",
    "\n",
    "            Modify the global algorithm in the following aspects:\n",
    "                1. Only computes matches between characters or gaps in one or another initial sequence.\n",
    "        \n",
    "         Args:\n",
    "                i (int): Sequence 0 index\n",
    "                j (int): Sequence 1 index\n",
    "                ini_gap (int): 1 if gap initiation, 0 if gap continuation\n",
    "        \"\"\"\n",
    "        score_diag, score_hor, score_ver = MIN, MIN, MIN\n",
    "        matches_diag, matches_hor, matches_ver = MIN, MIN, MIN\n",
    "        gaps_diag, gaps_hor, gaps_ver = MIN, MIN, MIN\n",
    "        #align and advance seq0 and seq1\n",
    "        #in long_substring mode only matches are processed\n",
    "        if i < self.len_seq0 and j < self.len_seq1 and\\\n",
    "        (self.mode != \"LONG_SUBSTRING\" or self.sequences[0][i] == self.sequences[1][j]):\n",
    "            inc_score, inc_matches = self.calc_score(i, j)\n",
    "            key = (i + 1, j + 1, 1)\n",
    "            if key in self.score_store:\n",
    "                score_diag, matches_diag, gaps_diag = \\\n",
    "                self.score_store[key] + inc_score, self.matches_store[key] + inc_matches, self.gaps_store[key]\n",
    "            else:\n",
    "                score, matches, gaps = self.align(i + 1, j + 1, 1)\n",
    "                self.store(key, score, matches, gaps)\n",
    "                score_diag, matches_diag, gaps_diag = score + inc_score, matches + inc_matches, gaps\n",
    "        #don't align and gap in seq0 (advance seq1)\n",
    "        if j < self.len_seq1:\n",
    "            gap_score = self.score_gap_cont + ini_gap * self.score_gap_ini\n",
    "            key = (i, j + 1, 0)\n",
    "            if key in self.score_store:\n",
    "                score_hor, matches_hor, gaps_hor = self.score_store[key] + gap_score,\\\n",
    "                self.matches_store[key], self.gaps_store[key] + 1\n",
    "            else:\n",
    "                score, matches, gaps = self.align(i, j + 1, 0)\n",
    "                self.store(key, score, matches, gaps)\n",
    "                score_hor, matches_hor, gaps_hor = score + gap_score, matches, gaps + 1\n",
    "        #don't align and gap in seq1 (advance seq0)\n",
    "        if i < self.len_seq0:\n",
    "            gap_score = self.score_gap_cont + ini_gap * self.score_gap_ini\n",
    "            key = (i + 1, j, 0)\n",
    "            if key in self.score_store:\n",
    "                score_ver, matches_ver, gaps_ver =\\\n",
    "                self.score_store[key] + gap_score, self.matches_store[key], self.gaps_store[key] + 1\n",
    "            else:\n",
    "                score, matches, gaps = self.align(i + 1, j, 0)\n",
    "                self.store(key, score, matches, gaps)\n",
    "                score_ver, matches_ver, gaps_ver = score + gap_score, matches, gaps + 1\n",
    "        #choose the high score path\n",
    "        matcher_diag, matcher_hor, matcher_ver = score_diag, score_hor, score_ver\n",
    "        if i < self.len_seq0 or j < self.len_seq1:\n",
    "            if self.mode == \"LOCAL\" and matcher_diag < 0 and matcher_hor < 0 and matcher_ver < 0:\n",
    "                score_diag, score_hor, score_ver = 0, 0, 0\n",
    "                #matcher_diag, matcher_hor, matcher_ver  = 0, 0, 0\n",
    "            if matcher_diag > matcher_hor and matcher_diag > matcher_ver:\n",
    "                ret_score, ret_matches, ret_gaps, ret_arrow =\\\n",
    "                score_diag, matches_diag, gaps_diag, \"d\"\n",
    "            elif matcher_hor > matcher_ver:\n",
    "                ret_score, ret_matches, ret_gaps, ret_arrow =\\\n",
    "                score_hor, matches_hor, gaps_hor, \"h\"\n",
    "            else:\n",
    "                ret_score, ret_matches, ret_gaps, ret_arrow =\\\n",
    "                score_ver, matches_ver, gaps_ver, \"v\"\n",
    "        else:\n",
    "            ret_score, ret_matches, ret_gaps, ret_arrow =\\\n",
    "            0, 0, 0, \"\"\n",
    "        self.forward_arrow[(i, j, ini_gap)] = ret_arrow\n",
    "        if i == 0 and j == 0:\n",
    "            self.store((0, 0, 1), ret_score, ret_matches, ret_gaps)\n",
    "            if self.mode in [\"GLOBAL\", \"LONG_SUBSTRING\"]: self.max_score_index = (0, 0, 1)\n",
    "            else: ret_score = self.max_score // COMPAC\n",
    "            ret_matches = self.matches_store[self.max_score_index]\n",
    "            ret_gaps = self.gaps_store[self.max_score_index]\n",
    "            \n",
    "        return ret_score, ret_matches, ret_gaps\n",
    "\n",
    "    def compute(self, mode=\"LOCAL\", silent=False):\n",
    "        \"\"\"Calc alignment\n",
    "            Args:\n",
    "                mode (str): Type of algorithm (local, global or long substring)\n",
    "                silent (bool): If true don't show alignment output\n",
    "        \"\"\"\n",
    "        self.ini_time = time.time()\n",
    "        self.init_stores()\n",
    "        self.set_mode(mode)\n",
    "        self.score, self.matches, self.gaps = self.align()\n",
    "        self.align_seq0, self.align_seq1, final_pos = self.forward_track(self.max_score_index)\n",
    "        self.matching = self.calc_matching(self.align_seq0, self.align_seq1, self.max_score_index, final_pos)\n",
    "        self.finish_time = time.time()\n",
    "        if not silent:\n",
    "            self.view()\n",
    "    \n",
    "    def get_len_long_common_substring(self):\n",
    "        \"\"\"Getter for the len of the common substring\n",
    "        That is equal to the number of matches of the alignment        \n",
    "        \"\"\"\n",
    "        return self.matches\n",
    "    \n",
    "    def  get_long_common_substring(self):\n",
    "        \"\"\"Returns the longest common substring\n",
    "        whitout alignment (positional) information\n",
    "        \"\"\"\n",
    "        long_common_substring = \"\"\n",
    "        for (char, match_char) in zip(self.align_seq1, self.matching):\n",
    "            if match_char == '|':\n",
    "                long_common_substring += char\n",
    "        return long_common_substring\n",
    "        \n",
    "    def view(self):\n",
    "        \"\"\"Prints the alignment data\"\"\"\n",
    "        unmatches = self.matching.count('.')\n",
    "        gaps = self.matching.count(' ')\n",
    "        if self.matching:\n",
    "            gap_groups = self.matching.count('| ') + self.matching.count('. ') + self.matching[0].count(' ')\n",
    "        else:\n",
    "            gap_groups = 0\n",
    "        print(\" \")\n",
    "        if self.mode == \"LOCAL\":\n",
    "            print(\"### AlignSequences. Local alignment (Smith-Waterman)\")\n",
    "        elif self.mode == \"LONG_SUBSTRING\":\n",
    "            print(\"### AlignSequences. Long substring finder\")\n",
    "        else:\n",
    "            print(\"### AlignSequences. Global alignment (Needleman-Wunsch)\")\n",
    "        if self.subst_matrix:\n",
    "            print(\"\\tUsing score matrix\")\n",
    "        print(self.align_seq1)\n",
    "        print(self.matching)\n",
    "        print(self.align_seq0)\n",
    "        print(\"\\tScore:\", self.score)\n",
    "        print(\"\\tInit index:\", self.max_score_index)\n",
    "        print(\"\\tMatches:\", self.matches, \" Unmatches:\", unmatches, \" Gaps:\", gaps, \" Gap groups:\", gap_groups)\n",
    "        #simple scoring verification todo: apply to matrix\n",
    "        if not self.subst_matrix:\n",
    "            print(\"\\tScore verified:\", self.matches * self.score_match + unmatches * self.score_no_match \\\n",
    "                  + gaps * self.score_gap_cont + gap_groups * self.score_gap_ini)\n",
    "        print(\"\\tFinish. Execution milliseconds:\", round((self.finish_time - self.ini_time) * 1000))\n",
    "        print(\"\\tScore Dictionary Size\", len(list(self.score_store.keys())))\n",
    "        \n",
    "    def edit_distance(self, score_match=0, score_no_match=-1, score_gap_ini=0, score_gap_cont=-1):\n",
    "        \"\"\"Calculates an edit distance as requested in questions 1 and 3\n",
    "        It's the same computation as a global alignment with -1 penalities applied to\n",
    "        score_gap_cont and score_nomatch and 0 in score_match and score_gap_ini\n",
    "            \n",
    "            Args:\n",
    "                score_match (int): Score of match characters.\n",
    "                score_no_match (int): Score of no match characters.\n",
    "                score_gap_ini (int): Score of gap init.\n",
    "                score_gap_cont (int): Score of gap continuation.\n",
    "        \"\"\"\n",
    "        self.set_scores(score_match, score_no_match, score_gap_ini, score_gap_cont)\n",
    "        self.compute(\"GLOBAL\", True)\n",
    "        return abs(self.score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 1\n",
    "In this exercise you will test the Needleman-Wunsch algorithm on a short\n",
    "sequence parts of hemoglobin (PDB code 1AOW) and myoglobin 1 (PDB code 1AZI). Here\n",
    "you will align the sequence **HGSAQVKGHG** to the sequence **KTEAEMKASEDLKKHGT**.\n",
    "The two sequences are arranged in a matrix in Table 1. \n",
    "\n",
    "The sequences start at the upper right corner, the initial gap penalties are listed at each offset starting position. The gap penalty is considered as -8.\n",
    " \n",
    "\n",
    "The similarity scores Si,j from looking up the matches comes from the **BLOSUM40** table. The\n",
    "table is shown here labeled with 1-letter amino acid codes:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Response\n",
    "As I have developed my version of global alignement, very similar to Needleman-Wunsch algorithm, I'm going to use this.\n",
    "Tne algorithm is in the class `AlignSequences` that I include on top."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " \n",
      "### AlignSequences. Global alignment (Needleman-Wunsch)\n",
      "\tUsing score matrix\n",
      "KTEAEMKASEDLKKHGT\n",
      "  ..  .| . .|.|| \n",
      "--HG--SA-Q-VKGHG-\n",
      "\tScore: -21\n",
      "\tInit index: (0, 0, 1)\n",
      "\tMatches: 4  Unmatches: 6  Gaps: 7  Gap groups: 5\n",
      "\tFinish. Execution milliseconds: 4\n",
      "\tScore Dictionary Size 368\n"
     ]
    }
   ],
   "source": [
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio.SubsMat import MatrixInfo\n",
    "\n",
    "s1 = \"HGSAQVKGHG\"\n",
    "s2 = \"KTEAEMKASEDLKKHGT\"\n",
    "mode = \"GLOBAL\"\n",
    "matrix = MatrixInfo.blosum40\n",
    "\n",
    "align = AlignSequences([s1, s2])\n",
    "align.set_scores(score_match = 0, score_no_match = 0, score_gap_ini = 0, score_gap_cont = -8)\n",
    "align.set_subst_matrix(matrix)\n",
    "align.compute(mode.upper(), silent = False)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It's OK?\n",
    "Whe could compare against **biopython**. By means of the function `test_alignment`. This function compare `AlignSequences` results with the alignments of Biopython. This software returns all the solutions of maximun score. My software not yet and so we test that my result is into the set of Biopython ones.\n",
    "I do the test with several BLOSUM matrices."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Passed test 1: scores are equal '-31'\n",
      "Passed test 1: alignments are equal '--HGS--A-Q-VKGHG-'\n",
      "Passed test 2: scores are equal '-32'\n",
      "Passed test 2: alignments are equal '--HG--SA-Q-VKGHG-'\n",
      " \n",
      "### AlignSequences. Global alignment (Needleman-Wunsch)\n",
      "\tUsing score matrix\n",
      "KTEAEMKASEDLKKHGT\n",
      "  ..  .| . .|.|| \n",
      "--HG--SA-Q-VKGHG-\n",
      "\tScore: -21\n",
      "\tInit index: (0, 0, 1)\n",
      "\tMatches: 4  Unmatches: 6  Gaps: 7  Gap groups: 5\n",
      "\tFinish. Execution milliseconds: 2\n",
      "\tScore Dictionary Size 368\n",
      "\n",
      "BioPython alignment:\n",
      "KTEAEMKASEDLKKHGT\n",
      "  ..  .| . .|.|| \n",
      "--HG--SA-Q-VKGHG-\n",
      "  Score=-21\n",
      "\n",
      "Passed test 3: scores are equal '-21'\n",
      "Passed test 3: alignments are equal '--HG--SA-Q-VKGHG-'\n",
      " \n",
      "Passed All Test\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "failed = 0\n",
    "passed = 0\n",
    "launched = 0\n",
    "\n",
    "def test_alignment(number, s1, s2, verbose=False, tipus=\"local\", matrix={},\\\n",
    "                          score_match=2, score_no_match=-3, score_gap_ini=-5, score_gap_cont=-2):\n",
    "    \"\"\"Comparisons of global and local alignments between Biopython and AlignSequences implementation.\n",
    "    \n",
    "    Args:\n",
    "        number (int): The number(identifier) of the test.\n",
    "        s1 (str): Query string to align.\n",
    "        s2 (str): Subject string to align\n",
    "        verbose (bool): If True print outputs, default False\n",
    "        tipus (str) : If 'local' the alignment is local (Smith), if 'global' Waterman.\n",
    "        matrix (dict of int) : Substitution matrix\n",
    "        score_match (int): Score of character match\n",
    "        score_no_match (int): Score of character no match\n",
    "        score_gap_ini (int): Score of gap initiation\n",
    "        score_gap_cont (int): Score of gap continuation\n",
    "        \n",
    "    \"\"\"\n",
    "    global failed, passed, launched\n",
    "    try:\n",
    "        launched += 1\n",
    "        align = AlignSequences([s1, s2])\n",
    "        align.set_scores(score_match, score_no_match, score_gap_ini, score_gap_cont)\n",
    "        if matrix != {}:\n",
    "            method = getattr(pairwise2.align, tipus + 'ds')\n",
    "            alignments = method(s2, s1, matrix,\\\n",
    "                                             score_gap_ini + score_gap_cont, score_gap_cont)\n",
    "            align.set_subst_matrix(matrix)\n",
    "        else:\n",
    "            method = getattr(pairwise2.align, tipus + 'ms')\n",
    "            alignments = method(s2, s1, score_match, score_no_match,\\\n",
    "                                              score_gap_ini + score_gap_cont, score_gap_cont)\n",
    "        \n",
    "        align.compute(tipus.upper(), silent=not verbose)\n",
    "        m = re.match(r\".*Score=([-1234567890]*)\", format_alignment(*alignments[0]).replace(\"\\n\", \"\"))\n",
    "        score = int(m.group(1))\n",
    "        \n",
    "        #search AlignSequences alignment in all possibles alignments fron Biopython\n",
    "        found = False\n",
    "        for a in alignments:\n",
    "            if verbose: \n",
    "                print()\n",
    "                print(\"BioPython alignment:\")\n",
    "                print(format_alignment(*a))\n",
    "            if (align.align_seq0 == a[1]):\n",
    "                    found = True\n",
    "                    break             \n",
    "        assert(align.score == score)\n",
    "        print (\"Passed test %s: scores are equal '%s'\" % (number, align.score ))\n",
    "        assert(found)\n",
    "        print (\"Passed test %s: alignments are equal '%s'\" % (number, align.align_seq0 ))\n",
    "        passed += 1\n",
    "    \n",
    "    except AssertionError:\n",
    "        print (\"Failed test %s: alignments differ: \\nBiopython:\\n'%s'\\nScore = %s \\\n",
    "        \\nAlignSequences\\n'%s'\\nScore = %s\"\\\n",
    "               % (number, alignments[0][1], score, align.align_seq0, align.score ))\n",
    "        failed += 1\n",
    "        exit(1)\n",
    "\n",
    "prot1 = \"HGSAQVKGHG\"\n",
    "prot2 = \"KTEAEMKASEDLKKHGT\"\n",
    "\n",
    "test_alignment(1, prot1, prot2, False, \"global\", MatrixInfo.blosum80, 0, 0, 0, -8)\n",
    "test_alignment(2, prot1, prot2, False, \"global\", MatrixInfo.blosum62, 0, 0, 0, -8)\n",
    "test_alignment(3, prot1, prot2, True, \"global\", MatrixInfo.blosum40, 0, 0, 0, -8)\n",
    "\n",
    "\n",
    "print(\" \")\n",
    "if launched == passed: print('Passed All Test')\n",
    "else: print(\"ERROR: There are failed tests\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exercise 2 \n",
    "Given the multiple sequence set:\n",
    "- S1: PPGVKSDCAS\n",
    "- S2: PADGVKDCAS\n",
    "- S3: PPDGKSDS\n",
    "- S4: GADGKDCCS\n",
    "- S5: GADGKDCAS\n",
    "\n",
    "Use the popular progressive alignment method to globally align the previous set of sequences.\n",
    "Generate the guide tree by neighbord-joining. Compare your outcome (alignment) with the\n",
    "one from Clustal-W."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Response"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#do the algorithm here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "latex_metadata": {
     "hidden": "true",
     "lexer": "bash"
    }
   },
   "source": [
    "## Outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "code_folding": [],
    "hidden": "true",
    "latex_metadata": {
     "hidden": "true",
     "lexer": "bash"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is pdfTeX, Version 3.14159265-2.6-1.40.19 (TeX Live 2018) (preloaded format=pdflatex)\n",
      " \\write18 enabled.\n",
      "entering extended mode\n",
      "(./worksheet.tex\n",
      "LaTeX2e <2018-04-01> patch level 2\n",
      "Babel <3.18> and hyphenation patterns for 84 language(s) loaded.\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/base/article.cls\n",
      "Document Class: article 2014/09/29 v1.4h Standard LaTeX document class\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/base/size11.clo))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/placeins/placeins.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/amsfonts/amssymb.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/amsfonts/amsfonts.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/amsmath/amsmath.sty\n",
      "For additional information on amsmath, use the `?' option.\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/amsmath/amstext.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/amsmath/amsgen.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/amsmath/amsbsy.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/amsmath/amsopn.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/amscls/amsthm.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/mathtools/mathtools.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/graphics/keyval.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/tools/calc.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/mathtools/mhsetup.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/base/fontenc.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/base/t1enc.def))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/psnfss/mathpazo.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/graphics/graphicx.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/graphics/graphics.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/graphics/trig.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/graphics-cfg/graphics.cfg)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/graphics-def/pdftex.def)))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/caption/caption.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/caption/caption3.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/adjustbox/adjustbox.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/xkeyval/xkeyval.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/xkeyval/xkeyval.tex\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/xkeyval/xkvutils.tex)))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/adjustbox/adjcalc.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/adjustbox/trimclip.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/collectbox/collectbox.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/adjustbox/tc-pdftex.def))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/ifoddpage/ifoddpage.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/varwidth/varwidth.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/xcolor/xcolor.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/graphics-cfg/color.cfg))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/tools/enumerate.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/geometry/geometry.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/ifpdf.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/ifvtex.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/ifxetex/ifxetex.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/base/textcomp.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/base/ts1enc.def\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/base/ts1enc.dfu)))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/upquote/upquote.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/eurosym/eurosym.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/ucs/ucs.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/ucs/data/uni-global.def))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/base/inputenc.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/ucs/utf8x.def))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/fancyvrb/fancyvrb.sty\n",
      "Style option: `fancyvrb' v2.7a, with DG/SPQR fixes, and firstline=lastline fix \n",
      "<2008/02/07> (tvz))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/oberdiek/grffile.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/oberdiek/kvoptions.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/ltxcmds.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/kvsetkeys.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/infwarerr.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/etexcmds.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/ifluatex.sty))))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/pdftexcmds.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/hyperref/hyperref.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/hobsub-hyperref.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/hobsub-generic.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/oberdiek/auxhook.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/hyperref/pd1enc.def)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/latexconfig/hyperref.cfg)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/url/url.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/hyperref/hpdftex.def\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/oberdiek/rerunfilecheck.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/tools/longtable.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/booktabs/booktabs.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/enumitem/enumitem.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/ulem/ulem.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/minted/minted.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/fvextra/fvextra.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/base/ifthen.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/etoolbox/etoolbox.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/lineno/lineno.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/tools/shellesc.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/ifplatform/ifplatform.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/catchfile.sty)\n",
      "(./worksheet.w18))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/xstring/xstring.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/xstring/xstring.tex))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/framed/framed.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/float/float.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/tcolorbox/tcolorbox.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/pgf/basiclayer/pgf.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/pgf/utilities/pgfrcs.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/utilities/pgfutil-common.te\n",
      "x\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/utilities/pgfutil-common-li\n",
      "sts.tex))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/utilities/pgfutil-latex.def\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/ms/everyshi.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/utilities/pgfrcs.code.tex))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/pgf/basiclayer/pgfcore.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/pgf/systemlayer/pgfsys.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/systemlayer/pgfsys.code.tex\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/utilities/pgfkeys.code.tex\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/utilities/pgfkeysfiltered.c\n",
      "ode.tex))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/systemlayer/pgf.cfg)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/systemlayer/pgfsys-pdftex.d\n",
      "ef\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/systemlayer/pgfsys-common-p\n",
      "df.def)))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/systemlayer/pgfsyssoftpath.\n",
      "code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/systemlayer/pgfsysprotocol.\n",
      "code.tex))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcore.code.tex\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmath.code.tex\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathcalc.code.tex\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathutil.code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathparser.code.tex\n",
      ")\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathfunctions.code.\n",
      "tex\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathfunctions.basic\n",
      ".code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathfunctions.trigo\n",
      "nometric.code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathfunctions.rando\n",
      "m.code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathfunctions.compa\n",
      "rison.code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathfunctions.base.\n",
      "code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathfunctions.round\n",
      ".code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathfunctions.misc.\n",
      "code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathfunctions.integ\n",
      "erarithmetics.code.tex)))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmathfloat.code.tex)\n",
      ")\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcorepoints.co\n",
      "de.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcorepathconst\n",
      "ruct.code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcorepathusage\n",
      ".code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcorescopes.co\n",
      "de.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcoregraphicst\n",
      "ate.code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcoretransform\n",
      "ations.code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcorequick.cod\n",
      "e.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcoreobjects.c\n",
      "ode.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcorepathproce\n",
      "ssing.code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcorearrows.co\n",
      "de.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcoreshade.cod\n",
      "e.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcoreimage.cod\n",
      "e.tex\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcoreexternal.\n",
      "code.tex))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcorelayers.co\n",
      "de.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcoretranspare\n",
      "ncy.code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/basiclayer/pgfcorepatterns.\n",
      "code.tex)))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/modules/pgfmoduleshapes.cod\n",
      "e.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/modules/pgfmoduleplot.code.\n",
      "tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/pgf/compatibility/pgfcomp-version\n",
      "-0-65.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/pgf/compatibility/pgfcomp-version\n",
      "-1-18.sty)) (/usr/local/texlive/2018/texmf-dist/tex/latex/tools/verbatim.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/environ/environ.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/trimspaces/trimspaces.sty)))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/tcolorbox/tcbminted.code.tex\n",
      "Library (tcolorbox): 'tcbminted.code.tex' version '4.13'\n",
      "\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/tcolorbox/tcblistingscore.code.te\n",
      "x\n",
      "Library (tcolorbox): 'tcblistingscore.code.tex' version '4.13'\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/tcolorbox/tcbprocessing.code.tex\n",
      "Library (tcolorbox): 'tcbprocessing.code.tex' version '4.13'\n",
      ")))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/tcolorbox/tcbbreakable.code.tex\n",
      "Library (tcolorbox): 'tcbbreakable.code.tex' version '4.13'\n",
      ") (/usr/local/texlive/2018/texmf-dist/tex/latex/tcolorbox/tcbskins.code.tex\n",
      "Library (tcolorbox): 'tcbskins.code.tex' version '4.13'\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/pgf/frontendlayer/tikz.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/pgf/utilities/pgffor.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/pgf/utilities/pgfkeys.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/utilities/pgfkeys.code.tex)\n",
      ") (/usr/local/texlive/2018/texmf-dist/tex/latex/pgf/math/pgfmath.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmath.code.tex))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/utilities/pgffor.code.tex\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/math/pgfmath.code.tex)))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/frontendlayer/tikz/tikz.cod\n",
      "e.tex\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/libraries/pgflibraryplothan\n",
      "dlers.code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/modules/pgfmodulematrix.cod\n",
      "e.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/pgf/frontendlayer/tikz/librarie\n",
      "s/tikzlibrarytopaths.code.tex)))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/tcolorbox/tcbskinsjigsaw.code.tex\n",
      "Library (tcolorbox): 'tcbskinsjigsaw.code.tex' version '4.13'\n",
      ")) (/usr/local/texlive/2018/texmf-dist/tex/latex/mdframed/mdframed.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/l3packages/xparse/xparse.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/l3kernel/expl3.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/l3kernel/expl3-code.tex)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/l3kernel/l3pdfmode.def)))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/oberdiek/zref-abspage.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/oberdiek/zref-base.sty))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/needspace/needspace.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/mdframed/md-frame-1.mdf))/anaconda2/envs/python3/bin/pygmentize\n",
      "\n",
      "(./worksheet.aux) (/usr/local/texlive/2018/texmf-dist/tex/latex/base/ts1cmr.fd)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/psnfss/t1ppl.fd)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/context/base/mkii/supp-pdf.mkii\n",
      "[Loading MPS to PDF converter (version 2006.09.02).]\n",
      ") (/usr/local/texlive/2018/texmf-dist/tex/latex/oberdiek/epstopdf-base.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/oberdiek/grfext.sty)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/latexconfig/epstopdf-sys.cfg))\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/caption/ltcaption.sty)\n",
      "*geometry* driver: auto-detecting\n",
      "*geometry* detected driver: pdftex\n",
      "\n",
      "Package geometry Warning: The marginal notes overrun the paper.\n",
      "     Add 25.63757pt and more to the right margin.\n",
      "\n",
      "*geometry* verbose mode - [ preamble ] result:\n",
      "* driver: pdftex\n",
      "* paper: <default>\n",
      "* layout: <same size as paper>\n",
      "* layoutoffset:(h,v)=(0.0pt,0.0pt)\n",
      "* modes: \n",
      "* h-part:(L,W,R)=(43.36243pt, 527.57014pt, 43.36243pt)\n",
      "* v-part:(T,H,B)=(72.26999pt, 650.43001pt, 72.26999pt)\n",
      "* \\paperwidth=614.295pt\n",
      "* \\paperheight=794.96999pt\n",
      "* \\textwidth=527.57014pt\n",
      "* \\textheight=650.43001pt\n",
      "* \\oddsidemargin=-28.90756pt\n",
      "* \\evensidemargin=-28.90756pt\n",
      "* \\topmargin=-37.0pt\n",
      "* \\headheight=12.0pt\n",
      "* \\headsep=25.0pt\n",
      "* \\topskip=11.0pt\n",
      "* \\footskip=30.0pt\n",
      "* \\marginparwidth=59.0pt\n",
      "* \\marginparsep=10.0pt\n",
      "* \\columnsep=10.0pt\n",
      "* \\skip\\footins=10.0pt plus 4.0pt minus 2.0pt\n",
      "* \\hoffset=0.0pt\n",
      "* \\voffset=0.0pt\n",
      "* \\mag=1000\n",
      "* \\@twocolumnfalse\n",
      "* \\@twosidefalse\n",
      "* \\@mparswitchfalse\n",
      "* \\@reversemarginfalse\n",
      "* (1in=72.27pt=25.4mm, 1cm=28.453pt)\n",
      "\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/ucs/ucsencs.def)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/hyperref/nameref.sty\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/generic/oberdiek/gettitlestring.sty))\n",
      "(./worksheet.out) (./worksheet.out) ABD: EveryShipout initializing macros\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/psnfss/ot1ppl.fd)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/psnfss/omlzplm.fd)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/psnfss/omszplm.fd)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/psnfss/omxzplm.fd)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/psnfss/ot1zplm.fd)\n",
      "(./worksheet.toc) [1{/usr/local/texlive/2018/texmf-var/fonts/map/pdftex/updmap/\n",
      "pdftex.map}] (./_minted-worksheet/default-pyg-prefix.pygstyle)\n",
      "(./_minted-worksheet/default.pygstyle)\n",
      "(./_minted-worksheet/2CF3CA77632ABA7E23D0CC693686383BB20D91E1B3CCA495EF4356D4A7\n",
      "633E11.pygtex (/usr/local/texlive/2018/texmf-dist/tex/latex/base/t1cmtt.fd)\n",
      "(/usr/local/texlive/2018/texmf-dist/tex/latex/base/ts1cmtt.fd)\n",
      "Underfull \\hbox (badness 10000) in paragraph at lines 71--71\n",
      "|                 []\\T1/cmtt/m/n/10 s[]c[]o[]r[]e[]_[]g[]a[]p[]_[]i[]n[]i[][]=[\n",
      "][]S[]C[]O[]R[]E[]_[]G[]A[]P[]_[]I[]N[]I[], []s[]c[]o[]r[]e[]_[]g[]a[]p[]_[]c[]\n",
      "o[]n[]t[][]=[][]S[]C[]O[]R[]E[]_[]G[]A[]P[]_[]C[]O[]N[]T[], []s[]u[]b[]s[]t[]_[\n",
      "]m[]a[]t[]r[]i[]x[][]=[][]{[]}[]\n",
      "\n",
      "Underfull \\hbox (badness 10000) in paragraph at lines 389--389\n",
      "|        [][]\\T1/cmtt/m/n/10 s[]e[]l[]f[][][].[][]a[]l[]i[]g[]n[]_[]s[]e[]q[]0[\n",
      "], [][]s[]e[]l[]f[][][].[][]a[]l[]i[]g[]n[]_[]s[]e[]q[]1[], []f[]i[]n[]a[]l[]_[\n",
      "]p[]o[]s [][]=[] [][]s[]e[]l[]f[][][].[][]f[]o[]r[]w[]a[]r[]d[]_[]t[]r[]a[]c[]k\n",
      "[]([][]s[]e[]l[]f[][][].[][]m[]a[]x[]_[]s[]c[]o[]r[]e[]_[]i[]n[]d[]e[]\n",
      ") [2] [3] [4] [5] [6] [7] [8] [9] [10]\n",
      "(./_minted-worksheet/03511521863BD1480BF8649541F9BA3EB20D91E1B3CCA495EF4356D4A7\n",
      "633E11.pygtex) [11]\n",
      "(./_minted-worksheet/8A67FBACAAB838D02C6B1F96242211047901E88107CAC51870C6F4AB91\n",
      "10AE2B.pygtex)\n",
      "(./_minted-worksheet/27C90F97824E9084029F32FBECE366DFB20D91E1B3CCA495EF4356D4A7\n",
      "633E11.pygtex) [12] [13]\n",
      "(./_minted-worksheet/444142DCBA21C71D5BDFCE726ED2D8C77901E88107CAC51870C6F4AB91\n",
      "10AE2B.pygtex)\n",
      "(./_minted-worksheet/0B860F9A9F481FF9651521F96C4CB6A7B20D91E1B3CCA495EF4356D4A7\n",
      "633E11.pygtex) [14] [15] (./worksheet.aux)\n",
      "\n",
      "Package rerunfilecheck Warning: File `worksheet.out' has changed.\n",
      "(rerunfilecheck)                Rerun to get outlines right\n",
      "(rerunfilecheck)                or use package `bookmark'.\n",
      "\n",
      "\n",
      "LaTeX Warning: Label(s) may have changed. Rerun to get cross-references right.\n",
      "\n",
      " )\n",
      "(see the transcript file for additional information){/usr/local/texlive/2018/te\n",
      "xmf-dist/fonts/enc/dvips/cm-super/cm-super-t1.enc}{/usr/local/texlive/2018/texm\n",
      "f-dist/fonts/enc/dvips/cm-super/cm-super-ts1.enc}{/usr/local/texlive/2018/texmf\n",
      "-dist/fonts/enc/dvips/base/8r.enc}</usr/local/texlive/2018/texmf-dist/fonts/typ\n",
      "e1/public/amsfonts/cm/cmmi10.pfb></usr/local/texlive/2018/texmf-dist/fonts/type\n",
      "1/public/amsfonts/cm/cmsy10.pfb></usr/local/texlive/2018/texmf-dist/fonts/type1\n",
      "/public/cm-super/sfit1000.pfb></usr/local/texlive/2018/texmf-dist/fonts/type1/p\n",
      "ublic/cm-super/sftt1000.pfb></usr/local/texlive/2018/texmf-dist/fonts/type1/pub\n",
      "lic/cm-super/sftt1095.pfb></usr/local/texlive/2018/texmf-dist/fonts/type1/urw/p\n",
      "alatino/uplb8a.pfb></usr/local/texlive/2018/texmf-dist/fonts/type1/urw/palatino\n",
      "/uplr8a.pfb>\n",
      "Output written on worksheet.pdf (15 pages, 178209 bytes).\n",
      "Transcript written on worksheet.log.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[NbConvertApp] Converting notebook worksheet.ipynb to latex\n",
      "[NbConvertApp] Writing 54427 bytes to worksheet.tex\n",
      "[NbConvertApp] Converting notebook worksheet.ipynb to html_with_toclenvs\n",
      "[NbConvertApp] Writing 397122 bytes to worksheet.html\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "#cd /Users/nandoide/Desktop/uni/STRBI.practical\n",
    "jupyter nbconvert --to=latex --template=~/report.tplx worksheet.ipynb\n",
    "pdflatex -shell-escape worksheet\n",
    "jupyter nbconvert --to html_with_toclenvs worksheet.ipynb"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "nbTranslate": {
   "displayLangs": [
    "en",
    "sp"
   ],
   "hotkey": "alt-t",
   "langInMainMenu": true,
   "sourceLang": "sp",
   "targetLang": "en",
   "useGoogleTranslate": false
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
