{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "latex_metadata": {
     "hidden": "true"
    }
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# The next line indicates to matplotlib to show the plots inline in the notebook\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Programming assignment \n",
    "\n",
    "You have program (in couples) a series of functions in order to generate a protein translation table. A protein is created from a sequence of nucleic acids codified in the messenger RNA (mRNA). This mRNA is read by the ribosome three bases at a time (called codons) to add one amino acid to the protein. In total there are $4^3=64$ different codons that correspond to the possible combinations of three bases.  Each codon codifies a single amino acid. The mRNA has one start codon that always codifies M and an end codon that signals the end of the protein.\n",
    "\n",
    "The objective of this assignment is to generate this translation table. To do so you will have to:\n",
    "\n",
    "+ Read the genome of a species from a FASTA file\n",
    "+ Read the annotation file for that genome to extract the location in the genome of the Coding Sequences (CDS) for each protein\n",
    "+ Extract the mRNAs from the genome that codifies each protein using its location \n",
    "+ Read the proteins from a FASTA file\n",
    "+ Generate the translation table using pairs of mRNA and protein\n",
    "\n",
    "\n",
    "In the following you will find the template of the functions to use and some other interesting execises "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## File parsing \n",
    "\n",
    "Implement the function below to parse a fasta file. Find an example of a fasta format file here:\n",
    "```\n",
    ">NC_000907.1:1073118-1073456 Haemophilus influenzae Rd KW20 chromoso\n",
    "ATGAAAACATTATTCACTTCGGTTGTATTGTGTGGTGCGCTGGTTGTTTCTTCGTCTTTTGCTGAGGAAA\n",
    "AAGCGACAGANCAAACCGCTCAATCTGTTGTGACAACTCAAGCTGAAGCTCAAGTAGCACCAGCCGTAGT\n",
    "GAGCGATAAATTGAATATCAACACAGCAACTGCCAGTGAAATTCAAAAATCCTTAACTGGCATTGGTGCG\n",
    "AAAAAAGCGGAAGCTATTGTGCAATATCGTGAAAAACACGGTAATTTTKCTAATGCAGAACAGCTTTTAG\n",
    "AAGTACAAGGAATTGGCAAAGCAACACTAGAGAAAAATCGTGATCGTATAATCTTTTAA\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readFasta(file):\n",
    "    \"\"\" Reads all sequences of a FASTA file \n",
    "        returns a dictionary \"\"\"  \n",
    "    ret_seqs = {}\n",
    "    seq = \"\"\n",
    "    key_found = False\n",
    "    with open(file, 'r') as f:\n",
    "        key = \"\"\n",
    "        for line in f:\n",
    "            line = line.replace('\\n', '')\n",
    "            if len(line) > 0:\n",
    "                if line[0] == \">\":\n",
    "                    if key_found:\n",
    "                        ret_seqs[key] = seq\n",
    "                    key_found = True\n",
    "                    key = line[1:].split(\" \")[0]\n",
    "                    seq = \"\"\n",
    "                elif key_found:\n",
    "                    seq += line\n",
    "    if key_found:\n",
    "        ret_seqs[key] = seq\n",
    "    return ret_seqs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "latex_metadata": {
     "lexer": "text"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting test.fasta\n"
     ]
    }
   ],
   "source": [
    "%%writefile test.fasta\n",
    ">NC_000907.1\n",
    "ATGAAAACATTATTCACTTCGGTTGTATTGTGTGGTGCGCTGGTTGTTTCTTCGTCTTTTGCTGAGGAAA\n",
    "AAGCGACAGANCAAACCGCTCAATCTGTTGTGACAACTCAAGCTGAAGCTCAAGTAGCACCAGCCGTAGT\n",
    "GAGCGATAAATTGAATATCAACACAGCAACTGCCAGTGAAATTCAAAAATCCTTAACTGGCATTGGTGCG\n",
    "AAAAAAGCGGAAGCTATTGTGCAATATCGTGAAAAACACGGTAATTTTKCTAATGCAGAACAGCTTTTAG\n",
    "AAGTACAAGGAATTGGCAAAGCAACACTAGAGAAAAATCGTGATCGTATAATCTTTTAACCCCC\n",
    ">NC_000907.2\n",
    "AAA\n",
    ">NC_000907.3\n",
    "AAACCCCC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "latex_metadata": {
     "lexer": "python"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'NC_000907.1': 'ATGAAAACATTATTCACTTCGGTTGTATTGTGTGGTGCGCTGGTTGTTTCTTCGTCTTTTGCTGAGGAAAAAGCGACAGANCAAACCGCTCAATCTGTTGTGACAACTCAAGCTGAAGCTCAAGTAGCACCAGCCGTAGTGAGCGATAAATTGAATATCAACACAGCAACTGCCAGTGAAATTCAAAAATCCTTAACTGGCATTGGTGCGAAAAAAGCGGAAGCTATTGTGCAATATCGTGAAAAACACGGTAATTTTKCTAATGCAGAACAGCTTTTAGAAGTACAAGGAATTGGCAAAGCAACACTAGAGAAAAATCGTGATCGTATAATCTTTTAACCCCC', 'NC_000907.2': 'AAA', 'NC_000907.3': 'AAACCCCC'}\n"
     ]
    }
   ],
   "source": [
    "seqs = readFasta(\"test.fasta\")\n",
    "print(seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.6505926449899562, 0.5848455819941591]\n"
     ]
    }
   ],
   "source": [
    "# Verify timings: initial problems with concatenation method that unable to process this file\n",
    "timings = %timeit -o -n 1 -r 2 -q readFasta(\"homo_sapiens/homo_sapiens_chr21.fasta\")\n",
    "print(timings.all_runs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All tests Ok\n"
     ]
    }
   ],
   "source": [
    "# Cell for checking if readFasta is working\n",
    "seqs = readFasta(\"phix174/phix.fa\")\n",
    "#print(seqs['gi|216019|gb|J02482.1|PX1CG'])\n",
    "# Check that the number of sequences is one\n",
    "assert(len(seqs)==1)\n",
    "# Check that the length of the sequences is correct\n",
    "assert(len(seqs['gi|216019|gb|J02482.1|PX1CG'])==5386)\n",
    "\n",
    "print(\"All tests Ok\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Count bases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([], [])"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def countBases(seq, bases = ['A', 'C', 'G', 'T']):\n",
    "    \"\"\" Counts the number of occurrences of each base in the sequence seq\n",
    "        There are two return values: the list of bases and the list of counts\n",
    "    \"\"\"\n",
    "    counts = []\n",
    "    for base in bases:\n",
    "        counts.append(seq.count(base) + seq.count(base.lower()))\n",
    "    return bases, counts\n",
    "countBases(\"AADFF\", [])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAY4AAAEKCAYAAAAFJbKyAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4wLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvqOYd8AAAFThJREFUeJzt3X+w5XV93/Hny90AKpWfV4K7i7vG1QSoSXAlKG2iUPmhGRdbTKFp3RjGnRo0P2yjqNOhk9QJNpliTQx2K1sgsRBK/LGGrYQAajKVHwsiuiJyiwhX0L0KYiwTcfHdP85nh+Pdu7v3e3fP+d7LPh8zd873+/5+zjnvPcPwms/3Z6oKSZLm6hl9NyBJWlwMDklSJwaHJKkTg0OS1InBIUnqxOCQJHVicEiSOjE4JEmdGBySpE6W9t3AKBx55JG1cuXKvtuQpEXl9ttv/3ZVTexp3NMyOFauXMmWLVv6bkOSFpUkX5/LOHdVSZI6MTgkSZ0YHJKkTgwOSVInBockqRODQ5LUicEhSerE4JAkdWJwSJI6eVpeOS5p/7Xygmv7bqFX91/02pF/hzMOSVInBockqRODQ5LUicEhSerE4JAkdTKy4EiyMcm2JF+aUX9bknuSbE3yn4fq70oy2badPlQ/o9Umk1wwqn4lSXMzytNxLwP+BLhiRyHJq4C1wEuq6gdJntvqxwLnAMcBzwP+JsmL2ts+CLwamAJuS7Kpqr48wr4lSbsxsuCoqs8mWTmj/Bbgoqr6QRuzrdXXAle1+teSTAIntm2TVXUfQJKr2liDQ5J6Mu5jHC8C/mmSW5J8JsnLWn0Z8ODQuKlW21VdktSTcV85vhQ4DDgJeBlwdZIXAJllbDF7sNVsH5xkPbAe4JhjjtknzUqSdjbuGccU8NEauBX4EXBkq68YGrcceGg39Z1U1YaqWlNVayYmJkbSvCRp/MHxceAUgHbw+wDg28Am4JwkByZZBawGbgVuA1YnWZXkAAYH0DeNuWdJ0pCR7apKciXwSuDIJFPAhcBGYGM7RfcJYF1VFbA1ydUMDnpvB86vqifb57wVuA5YAmysqq2j6lmStGejPKvq3F1s+te7GP9e4L2z1DcDm/dha5KkveCV45KkTgwOSVInBockqRODQ5LUicEhSerE4JAkdWJwSJI6MTgkSZ0YHJKkTgwOSVInBockqRODQ5LUicEhSerE4JAkdWJwSJI6MTgkSZ2MLDiSbEyyrT3tb+a2f5+kkhzZ1pPkA0kmk9yV5IShseuS3Nv+1o2qX0nS3IxyxnEZcMbMYpIVwKuBB4bKZzJ4zvhqYD1wSRt7OINHzv4CcCJwYZLDRtizJGkPRhYcVfVZ4JFZNl0MvAOoodpa4IoauBk4NMnRwOnA9VX1SFU9ClzPLGEkSRqfsR7jSPI64BtV9YUZm5YBDw6tT7XaruqSpJ4sHdcXJXkW8B7gtNk2z1Kr3dRn+/z1DHZzccwxx8yzS0nSnoxzxvFTwCrgC0nuB5YDdyT5SQYziRVDY5cDD+2mvpOq2lBVa6pqzcTExAjalyTBGIOjqr5YVc+tqpVVtZJBKJxQVd8ENgFvbGdXnQQ8VlUPA9cBpyU5rB0UP63VJEk9GeXpuFcCnwNenGQqyXm7Gb4ZuA+YBP478BsAVfUI8PvAbe3v91pNktSTkR3jqKpz97B95dByAefvYtxGYOM+bU6SNG9eOS5J6sTgkCR1YnBIkjoxOCRJnRgckqRODA5JUicGhySpE4NDktTJ2G5yuJisvODavlvo1f0XvbbvFiQtYM44JEmdGBySpE4MDklSJwaHJKkTD45LC4wnZ3hyxkLnjEOS1InBIUnqZJRPANyYZFuSLw3V/jDJV5LcleRjSQ4d2vauJJNJ7kly+lD9jFabTHLBqPqVJM3NKGcclwFnzKhdDxxfVS8Bvgq8CyDJscA5wHHtPX+aZEmSJcAHgTOBY4Fz21hJUk9GFhxV9VngkRm1v66q7W31ZmB5W14LXFVVP6iqrzF49viJ7W+yqu6rqieAq9pYSVJP+jzG8evA/27Ly4AHh7ZNtdqu6jtJsj7JliRbpqenR9CuJAl6Co4k7wG2Ax/ZUZplWO2mvnOxakNVramqNRMTE/umUUnSTsZ+HUeSdcAvA6dW1Y4QmAJWDA1bDjzUlndVlyT1YKwzjiRnAO8EXldVjw9t2gSck+TAJKuA1cCtwG3A6iSrkhzA4AD6pnH2LEn6cSObcSS5EnglcGSSKeBCBmdRHQhcnwTg5qr6t1W1NcnVwJcZ7MI6v6qebJ/zVuA6YAmwsaq2jqpnSdKejSw4qurcWcqX7mb8e4H3zlLfDGzeh61JkvaC96rSPue9lrzXkp7evOWIJKkTg0OS1InBIUnqxOCQJHVicEiSOjE4JEmdGBySpE4MDklSJwaHJKkTg0OS1InBIUnqxOCQJHVicEiSOjE4JEmdjCw4kmxMsi3Jl4Zqhye5Psm97fWwVk+SDySZTHJXkhOG3rOujb+3PXZWktSjUc44LgPOmFG7ALihqlYDN7R1gDMZPC52NbAeuAQGQcPgyYG/AJwIXLgjbCRJ/RhZcFTVZ4FHZpTXApe35cuBs4bqV9TAzcChSY4GTgeur6pHqupR4Hp2DiNJ0hiN+xjHUVX1MEB7fW6rLwMeHBo31Wq7qkuSerJQDo5nllrtpr7zByTrk2xJsmV6enqfNidJesqcgiPJyXOpzcG32i4o2uu2Vp8CVgyNWw48tJv6TqpqQ1Wtqao1ExMT82hNkjQXc51x/PEca3uyCdhxZtQ64BND9Te2s6tOAh5ru7KuA05Lclg7KH5aq0mSerJ0dxuTvBx4BTCR5O1Dm54DLNnDe68EXgkcmWSKwdlRFwFXJzkPeAB4Qxu+GXgNMAk8DrwJoKoeSfL7wG1t3O9V1cwD7pKkMdptcAAHAAe3cf9oqP494OzdvbGqzt3FplNnGVvA+bv4nI3Axj30KUkak90GR1V9BvhMksuq6utj6kmStIDtacaxw4FJNgArh99TVaeMoilJ0sI11+D4X8CHgA8DT46uHUnSQjfX4NheVZeMtBNJ0qIw19NxP5nkN5Ic3W5UeHi7j5QkaT8z1xnHjmsvfneoVsAL9m07kqSFbk7BUVWrRt2IJGlxmFNwJHnjbPWqumLftiNJWujmuqvqZUPLBzG4iO8OwOCQpP3MXHdVvW14PckhwJ+NpCNJ0oI239uqP87gaX2SpP3MXI9xfJKnnoOxBPgZ4OpRNSVJWrjmeozjj4aWtwNfr6qpEfQjSVrg5rSrqt3s8CsM7pB7GPDEKJuSJC1cc30C4K8AtzJ4fsavALck2e1t1SVJT09z3VX1HuBlVbUNIMkE8DfANaNqTJK0MM31rKpn7AiN5jsd3ruTJL+TZGuSLyW5MslBSVYluSXJvUn+IskBbeyBbX2ybV853++VJO29uf7P/1NJrkvya0l+DbiWweNeO0uyDPhNYE1VHc/gLK1zgPcBF1fVauBR4Lz2lvOAR6vqhcDFbZwkqSe7DY4kL0xyclX9LvDfgJcAPwt8DtiwF9+7FHhmkqXAs4CHgVN4atfX5cBZbXltW6dtPzVJ9uK7JUl7YU8zjvcDfw9QVR+tqrdX1e8wmG28fz5fWFXfYHB67wMMAuMx4Hbgu1W1vQ2bApa15WXAg+2929v4I2Z+bpL1SbYk2TI9PT2f1iRJc7Cn4FhZVXfNLFbVFgaPke0syWEMZhGrgOcBzwbOnGXojgsOZ5td1E6Fqg1Vtaaq1kxMTMynNUnSHOwpOA7azbZnzvM7/xnwtaqarqofAh8FXgEc2nZdASwHHmrLU8AKgLb9EOCReX63JGkv7Sk4bkvy5pnFJOcx2L00Hw8AJyV5VjtWcSrwZeAmYMe1IeuAT7TlTTz1IKmzgRuraqcZhyRpPPZ0HcdvAx9L8qs8FRRrgAOA18/nC6vqliTXMLgt+3bg8wwOtF8LXJXkP7Xape0tlwJ/lmSSwUzjnPl8ryRp39htcFTVt4BXJHkVcHwrX1tVN+7Nl1bVhcCFM8r3ASfOMvYfGFyxLklaAOb6PI6bGOxKkiTt5+Z99bckaf9kcEiSOjE4JEmdGBySpE4MDklSJwaHJKkTg0OS1InBIUnqxOCQJHVicEiSOjE4JEmdGBySpE4MDklSJwaHJKmTXoIjyaFJrknylSR3J3l5ksOTXJ/k3vZ6WBubJB9IMpnkriQn9NGzJGmgrxnHfwU+VVU/DfwscDdwAXBDVa0GbmjrAGcCq9vfeuCS8bcrSdph7MGR5DnAL9IeDVtVT1TVd4G1wOVt2OXAWW15LXBFDdwMHJrk6DG3LUlq+phxvACYBv5Hks8n+XCSZwNHVdXDAO31uW38MuDBofdPtZokqQd9BMdS4ATgkqr6eeD/8dRuqdlkllrtNChZn2RLki3T09P7plNJ0k76CI4pYKqqbmnr1zAIkm/t2AXVXrcNjV8x9P7lwEMzP7SqNlTVmqpaMzExMbLmJWl/N/bgqKpvAg8meXErnQp8GdgErGu1dcAn2vIm4I3t7KqTgMd27NKSJI3f0p6+923AR5IcANwHvIlBiF2d5DzgAeANbexm4DXAJPB4GytJ6kkvwVFVdwJrZtl06ixjCzh/5E1JkubEK8clSZ0YHJKkTgwOSVInBockqRODQ5LUicEhSerE4JAkdWJwSJI6MTgkSZ0YHJKkTgwOSVInBockqRODQ5LUicEhSerE4JAkdWJwSJI66S04kixJ8vkkf9XWVyW5Jcm9Sf6iPR2QJAe29cm2fWVfPUuS+p1x/BZw99D6+4CLq2o18ChwXqufBzxaVS8ELm7jJEk96SU4kiwHXgt8uK0HOAW4pg25HDirLa9t67Ttp7bxkqQe9DXjeD/wDuBHbf0I4LtVtb2tTwHL2vIy4EGAtv2xNl6S1IOxB0eSXwa2VdXtw+VZhtYctg1/7vokW5JsmZ6e3gedSpJm08eM42TgdUnuB65isIvq/cChSZa2McuBh9ryFLACoG0/BHhk5odW1YaqWlNVayYmJkb7L5Ck/djYg6Oq3lVVy6tqJXAOcGNV/SpwE3B2G7YO+ERb3tTWadtvrKqdZhySpPFYSNdxvBN4e5JJBscwLm31S4EjWv3twAU99SdJApbuecjoVNWngU+35fuAE2cZ8w/AG8bamCRplxbSjEOStAgYHJKkTgwOSVInBockqRODQ5LUicEhSerE4JAkdWJwSJI6MTgkSZ0YHJKkTgwOSVInBockqRODQ5LUicEhSerE4JAkdWJwSJI6GXtwJFmR5KYkdyfZmuS3Wv3wJNcnube9HtbqSfKBJJNJ7kpywrh7liQ9pY8Zx3bg31XVzwAnAecnOZbBI2FvqKrVwA089YjYM4HV7W89cMn4W5Yk7TD24Kiqh6vqjrb898DdwDJgLXB5G3Y5cFZbXgtcUQM3A4cmOXrMbUuSml6PcSRZCfw8cAtwVFU9DINwAZ7bhi0DHhx621Srzfys9Um2JNkyPT09yrYlab/WW3AkORj4S+C3q+p7uxs6S612KlRtqKo1VbVmYmJiX7UpSZqhl+BI8hMMQuMjVfXRVv7Wjl1Q7XVbq08BK4bevhx4aFy9SpJ+XB9nVQW4FLi7qv7L0KZNwLq2vA74xFD9je3sqpOAx3bs0pIkjd/SHr7zZODfAF9McmervRu4CLg6yXnAA8Ab2rbNwGuASeBx4E3jbVeSNGzswVFVf8fsxy0ATp1lfAHnj7QpSdKceeW4JKkTg0OS1InBIUnqxOCQJHVicEiSOjE4JEmdGBySpE4MDklSJwaHJKkTg0OS1InBIUnqxOCQJHVicEiSOjE4JEmdGBySpE4WTXAkOSPJPUkmk1zQdz+StL9aFMGRZAnwQeBM4Fjg3CTH9tuVJO2fFkVwACcCk1V1X1U9AVwFrO25J0naLy2W4FgGPDi0PtVqkqQxG/szx+dptmeU148NSNYD69vq95PcM/KuRudI4Nt9fXne19c37zP+fnvH32/vLObf7/lzGbRYgmMKWDG0vhx4aHhAVW0ANoyzqVFJsqWq1vTdx2Ll77d3/P32zv7w+y2WXVW3AauTrEpyAHAOsKnnniRpv7QoZhxVtT3JW4HrgCXAxqra2nNbkrRfWhTBAVBVm4HNffcxJk+LXW498vfbO/5+e+dp//ulqvY8SpKkZrEc45AkLRAGxwKS5PVJKslP993LYpTkJ5NcleT/Jvlyks1JXtR3X4tFkqOS/M8k9yW5Pcnnkry+774WgyRHJLmz/X0zyTeG1g/ou799zeBYWM4F/o7BWWPqIEmAjwGfrqqfqqpjgXcDR/Xb2eLQfr+PA5+tqhdU1UsZ/He4vN/OFoeq+k5V/VxV/RzwIeDiHevtbhdPKwbHApHkYOBk4DwMjvl4FfDDqvrQjkJV3VlVf9tjT4vJKcATM36/r1fVH/fYkxYog2PhOAv4VFV9FXgkyQl9N7TIHA/c3ncTi9hxwB19N6HFweBYOM5lcPNG2uu5Pfai/VySDyb5QpLb+u5FC8+iuY7j6SzJEQx2FRyfpBhc5FhJ3lGeLz1XW4Gz+25iEdsK/IsdK1V1fpIjgS39taSFyhnHwnA2cEVVPb+qVlbVCuBrwD/pua/F5EbgwCRv3lFI8rIkv9RjT4vJjcBBSd4yVHtWX81oYTM4FoZzGZwRNOwvgX/VQy+LUpuZvR54dTsddyvwH5lxM0zNrv1+ZwG/lORrSW4FLgfe2W9nWoi8clyS1IkzDklSJwaHJKkTg0OS1InBIUnqxOCQJHVicEjzlOTJdvfTLyS5I8kr+u5JGgdPx5XmKcn3q+rgtnw68O6q8oJDPe0545D2jecAj8LgTsdJbmizkC8mWdvqz05ybZuhfCnJv2z1lyb5THsGxnVJjm7132zPFbkryVW7/GZpzJxxSPOU5Engi8BBwNHAKVV1e5KlwLOq6nvtfk83A6uBfw6cUVVvbu8/BHgc+AywtqqmW5icXlW/nuQhYFVV/SDJoVX13fH/K6WdGRzSPM3YVfVy4MMMbu++FLgY+EXgR8CLgVUMZiXXAVcDf1VVf5vkeOD/APe1j10CPFxVpyX5FPB9Bg9Y+nhVfX9s/zhpN7w7rrQPVNXn2uxiAnhNe31pVf0wyf3AQVX11SQvbdv/IMlfM7hH2daqevksH/taBuHzOuA/JDmuqraP498j7Y7HOKR9oD0nfgnwHeAQYFsLjVcBz29jngc8XlV/DvwRcAJwDzDRZiwk+YkkxyV5BrCiqm4C3gEcChw87n+XNBtnHNL8PTPJnW05wLqqejLJR4BPJtkC3Al8pY35x8AfJvkR8EPgLVX1RJKzgQ+0Yx5LgfcDXwX+vNXC4BnWHuPQguAxDklSJ+6qkiR1YnBIkjoxOCRJnRgckqRODA5JUicGhySpE4NDktSJwSFJ6uT/A+SPxuNAUw2dAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Use matplotlib to show the histogram of bases for phiX174\n",
    "bases, counts = countBases(seqs['gi|216019|gb|J02482.1|PX1CG']) \n",
    "plt.bar(bases, counts)\n",
    "plt.xlabel('Bases')\n",
    "_ = plt.ylabel('Count')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GC Content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def GCContent(seq):\n",
    "    \"\"\" Computes the percentage of GC content of a sequence. \n",
    "        Note that N should be ignored \"\"\"\n",
    "    count_GC = 0\n",
    "    count_all = 0\n",
    "    for base in seq:\n",
    "        if base in 'GC': \n",
    "            count_GC += 1\n",
    "            count_all += 1\n",
    "        elif base in 'AT':\n",
    "            count_all += 1\n",
    "    return round(count_GC * 1.0 / count_all, 10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All tests Ok\n"
     ]
    }
   ],
   "source": [
    "# Check that the GC content is correct\n",
    "seqs = readFasta(\"phix174/phix.fa\")\n",
    "assert(GCContent(seqs['gi|216019|gb|J02482.1|PX1CG'])==0.4476420349)\n",
    "# Check that the GC content is correct\n",
    "seqs = readFasta(\"plamodium_falciparum/Plasmodium_falciparumchromosome_1.fa\")\n",
    "assert(GCContent(seqs['NC_004325.1'])==0.2054572418)\n",
    "\n",
    "print(\"All tests Ok\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GC content by species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing file: haemophilus_influenzae/GCF_000027305.1_ASM2730v1_genomic.fna\n",
      "Processing file: phix174/phix.fa\n",
      "Processing file: plamodium_falciparum/Plasmodium_falciparumchromosome_1.fa\n",
      "Processing file: arabidopsis_thaliana/GCF_000001735.3_TAIR10_genomic.fna\n",
      "Processing file: homo_sapiens/homo_sapiens_chr21.fasta\n",
      "[0.3815028554, 0.4476420349, 0.2054572418, 0.3907552942, 0.4093836458]\n"
     ]
    }
   ],
   "source": [
    "files = [\"haemophilus_influenzae/GCF_000027305.1_ASM2730v1_genomic.fna\",\n",
    "         \"phix174/phix.fa\",\n",
    "         \"plamodium_falciparum/Plasmodium_falciparumchromosome_1.fa\",\n",
    "         \"arabidopsis_thaliana/GCF_000001735.3_TAIR10_genomic.fna\",\n",
    "         \"homo_sapiens/homo_sapiens_chr21.fasta\"]\n",
    "\n",
    "gcs = []\n",
    "for file in files:\n",
    "    print(\"Processing file:\", file)\n",
    "    seqs = readFasta(file)\n",
    "    # Count only the first sequence\n",
    "    keys = list(seqs.keys())\n",
    "    gcs.append(GCContent(seqs[keys[0]])) \n",
    "    \n",
    "print(gcs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAEKCAYAAAD9xUlFAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4wLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvqOYd8AAAFs1JREFUeJzt3Xu4JHV95/H3h+GqqBEZb4AMItFMFFEHiNewBjewRtAIEbKuugGJKJo1asI+uiySzSbKEzURvKBRzKohGI0ZZRSyCCt4QWZg5OYS5xkRZskqGsSwGnTgu39UnR/Noc85PcOpc+DM+/U8/ZyqX1dXf6tPdX/q0vXrVBWSJAFst9gFSJLuPwwFSVJjKEiSGkNBktQYCpKkxlCQJDWGgiSpMRQkSY2hIElqtl/sArbU7rvvXitWrFjsMiTpAWXdunU/qKrlc033gAuFFStWsHbt2sUuQ5IeUJJ8d5LpPHwkSWoMBUlSYyhIkhpDQZLUGAqSpMZQkCQ1hoIkqTEUJEmNoSBJah5wVzRr6604+bzFLmFe3PCnL1zsEqQlyz0FSVJjKEiSGkNBktQYCpKkxlCQJDWGgiSpMRQkSY2hIElqDAVJUmMoSJIaQ0GS1BgKkqTGUJAkNYaCJKkxFCRJjaEgSWr8kR1JS9pS+XEpWJgfmBp0TyHJYUmuT7IhycmzTHdUkkqyash6JEmzGywUkiwDzgQOB1YCxyZZOWa6hwBvAC4bqhZJ0mSG3FM4CNhQVRur6mfAOcCRY6b7I+CdwL8OWIskaQJDhsIewE0j45v6tibJ04C9qurzA9YhSZrQkKGQMW3V7ky2A94NvGnOGSUnJFmbZO0tt9wyjyVKkkYNGQqbgL1GxvcEbh4ZfwjwZODiJDcAvwKsHneyuarOqqpVVbVq+fLlA5YsSdu2IUPhcmC/JPsk2RE4Blg9dWdV3VZVu1fViqpaAXwdOKKq1g5YkyRpFoOFQlVtBk4Czge+BZxbVdcmOS3JEUM9ryRp6w168VpVrQHWTGs7ZYZpDxmyFmlbtlQu4FqIi7e2dXZzIUlqDAVJUmMoSJKabapDvKVyXBU8tippGO4pSJIaQ0GS1BgKkqTGUJAkNYaCJKkxFCRJjaEgSWoMBUlSYyhIkhpDQZLUGAqSpMZQkCQ1hoIkqTEUJEmNoSBJagwFSVJjKEiSGkNBktQYCpKkxlCQJDWGgiSpMRQkSY2hIElqDAVJUmMoSJIaQ0GS1BgKkqTGUJAkNYaCJKkxFCRJjaEgSWoMBUlSYyhIkppBQyHJYUmuT7Ihyclj7n9NkquTrE9yaZKVQ9YjSZrdYKGQZBlwJnA4sBI4dsyH/ier6ilVdQDwTuBdQ9UjSZrbkHsKBwEbqmpjVf0MOAc4cnSCqvrxyOiDgRqwHknSHLYfcN57ADeNjG8CDp4+UZLXAb8P7Ag8f8B6JElzGHJPIWPa7rUnUFVnVtW+wB8Cbxs7o+SEJGuTrL3lllvmuUxJ0pQhQ2ETsNfI+J7AzbNMfw7w4nF3VNVZVbWqqlYtX758HkuUJI0aMhQuB/ZLsk+SHYFjgNWjEyTZb2T0hcC3B6xHkjSHwc4pVNXmJCcB5wPLgI9U1bVJTgPWVtVq4KQkhwI/B24FXjlUPZKkuQ15opmqWgOsmdZ2ysjw7w35/JKkLeMVzZKkxlCQJDWGgiSpmTMUkrxjkjZJ0gPfJHsKLxjTdvh8FyJJWnwzfvsoyYnAa4HHJ7lq5K6HAF8ZujBJ0sKb7SupnwS+APwJMNrt9b9U1T8PWpUkaVHMGApVdRtwG12X18uAR/XT75pk16q6cYFqlCQtkDkvXuuvSj4V+B5wV99cwP7DlSVJWgyTXNH8n4AnVtUPhy5GkrS4Jvn20U10h5EkSUvcJHsKG4GLk5wH3DHVWFX+dKYkLTGThMKN/W3H/iZJWqLmDIWqevtCFCJJWnyTfPvoIsb/jKa/pyxJS8wkh4/ePDK8M/BSYPMw5UiSFtMkh4/WTWv6SpL/NVA9kqRFNMnho91GRrcDngE8erCKJEmLZpLDR+vozimE7rDRd4DjhixKkrQ4Jjl8tM9CFCJJWnyTHD7aATgReF7fdDHwwar6+YB1SZIWwSSHj94P7AC8rx//D33b8UMVJUlaHJOEwoFV9dSR8S8l+eZQBUmSFs8kHeLdmWTfqZEkjwfuHK4kSdJimWRP4S3ARUk20n0DaW/gPw5alSRpUUzy7aMLk+wHPJEuFP53Vd0xx8MkSQ9Acx4+SvI6YJequqqqvgk8KMlrhy9NkrTQJjmn8Oqq+tHUSFXdCrx6uJIkSYtlklDYLkmmRpIsw99VkKQlaZITzecD5yb5AF13F68BvjhoVZKkRTFJKPwhcALdVc0BLgA+PGRRkqTFMcm3j+4CPtDfJElL2CTnFCRJ2whDQZLUzBgKSXZOsnxM+yOT7DxsWZKkxTDbnsJfAM8d0/4C4N3DlCNJWkyzhcJzquoz0xur6hPc/dsKkqQlZLZQyCz3eS5Ckpag2T7cv5/koOmNSQ4Ebplk5kkOS3J9kg1JTh5z/+8nuS7JVUkuTLL35KVLkubbbNcpvIXuSuazgXV92yrgFcAxc8247w7jTLpzEJuAy5OsrqrrRia7ElhVVT9JciLwTuBlW7wU0hxWnHzeYpcwb2740xcudglawmbcU6iqbwAH0x1GelV/C3BwVV02wbwPAjZU1caq+hlwDnDktOe4qKp+0o9+HdhzSxdAkjR/Zr2iuaq+B/zXrZz3HsBNI+Ob6EJmJscBXxh3R5IT6Lra4HGPe9xWliNJmsts1ykc2f+WwtT4ZUk29rejJ5j3uBPVNcNzvZzu0NTp4+6vqrOqalVVrVq+/F6XTkiS5slsJ5r/AFg9Mr4TcCBwCF1PqXPZBOw1Mr4ncPP0iZIcCrwVOMJfdJOkxTVbKOxYVaOHfy6tqh9W1Y3AgyeY9+XAfkn2SbIj3cnp0ZAhydOAD9IFwve3sHZJ0jybLRQePjpSVSeNjM55DKeqNgMn0f0ew7eAc6vq2iSnJTmin+x0YFfgU0nWJ1k9w+wkSQtgthPNlyV5dVV9aLQxye8C35hk5lW1Blgzre2UkeFDt6BWSdLAZguFNwKfTfLbwBV92zPozi28eOjCJEkLb8ZQ6I/xPyvJ84Ff7pvPq6ovLUhlkqQFN8kvr30JMAgkaRtgx3aSpMZQkCQ1hoIkqTEUJEmNoSBJagwFSVJjKEiSGkNBktQYCpKkxlCQJDWGgiSpMRQkSY2hIElqDAVJUmMoSJIaQ0GS1BgKkqTGUJAkNYaCJKkxFCRJjaEgSWoMBUlSYyhIkhpDQZLUGAqSpMZQkCQ1hoIkqTEUJEmNoSBJagwFSVJjKEiSGkNBktQYCpKkxlCQJDWDhkKSw5Jcn2RDkpPH3P+8JFck2ZzkqCFrkSTNbbBQSLIMOBM4HFgJHJtk5bTJbgReBXxyqDokSZPbfsB5HwRsqKqNAEnOAY4ErpuaoKpu6O+7a8A6JEkTGvLw0R7ATSPjm/o2SdL91JChkDFttVUzSk5IsjbJ2ltuueU+liVJmsmQobAJ2GtkfE/g5q2ZUVWdVVWrqmrV8uXL56U4SdK9DRkKlwP7JdknyY7AMcDqAZ9PknQfDRYKVbUZOAk4H/gWcG5VXZvktCRHACQ5MMkm4Gjgg0muHaoeSdLchvz2EVW1Blgzre2UkeHL6Q4rSZLuB7yiWZLUGAqSpMZQkCQ1hoIkqTEUJEmNoSBJagwFSVJjKEiSGkNBktQYCpKkxlCQJDWGgiSpMRQkSY2hIElqDAVJUmMoSJIaQ0GS1BgKkqTGUJAkNYaCJKkxFCRJjaEgSWoMBUlSYyhIkhpDQZLUGAqSpMZQkCQ1hoIkqTEUJEmNoSBJagwFSVJjKEiSGkNBktQYCpKkxlCQJDWGgiSpMRQkSc2goZDksCTXJ9mQ5OQx9++U5G/6+y9LsmLIeiRJsxssFJIsA84EDgdWAscmWTltsuOAW6vqCcC7gXcMVY8kaW5D7ikcBGyoqo1V9TPgHODIadMcCXysH/5b4NeSZMCaJEmzGDIU9gBuGhnf1LeNnaaqNgO3AY8YsCZJ0iy2H3De47b4ayumIckJwAn96O1Jrr+PtQ1td+AHQz5B7r8H2lz2gW3Ly78tLzvc5+Xfe5KJhgyFTcBeI+N7AjfPMM2mJNsDDwP+efqMquos4KyB6px3SdZW1arFrmMxuOzb5rLDtr38S2nZhzx8dDmwX5J9kuwIHAOsnjbNauCV/fBRwJeq6l57CpKkhTHYnkJVbU5yEnA+sAz4SFVdm+Q0YG1VrQb+EvgfSTbQ7SEcM1Q9kqS5DXn4iKpaA6yZ1nbKyPC/AkcPWcMiecAc6hqAy77t2paXf8ksezxaI0maYjcXkqRmSYZCktunjb8qyRlzPGZFkmsmmPfRSb6V5KIkhyT5/H2t94EiyZ1J1ie5JsmnkjxottctyWlJDl2s2uaY/uNJvtM/Zn2S180x/aVJDpjl/pckecvW1j+kvrZK8qQJpv27/vXYkOS2kdfnWUluSLL7Fjxve38kOWJcVzcLYSs/D7ZL8hf9+nR1ksuT7LOVz/+aJK/YmscuhkHPKSxRxwGvraqLkhyy2MUssJ9W1QEAST4BvAb4zEwTj54/WgDjanvXHI95Y1V9dj6evKr+bj7mk2T7/kLO+XQscCndFzlOnW3CqnpJX8chwJur6jdGatvqAvovlkz/9uH92cuAxwL7V9VdSfYE/t/WzKiqPjCvlQ1sSe4p3Ff9lsRnknwxybeTvLNvPwV4DvCBJKdPe8ypSd48Mn7NVAd/SV6e5Bv9FtcH+36hSHJ7kj9O8s0kX0/yqL59/cjtp0l+NclBSb6a5Mr+7xP7aZclOb3fkrkqye8uxGsEXAI8oR9eluRDSa5NckGSXfrazk5yVJKHpesYcarmv07y6gWqbYskOSvJ2n5ZxoZakhcmuaL/v13Qtx2f5D398MeTvD/JJUn+Mcnhffu+fduVSdYlObhvPzTJ/0xyDnBlkickWT/yfCcneVs/fGmSd/XzuS7Jqn7r/ttJTh1T667As+k2Zubj232v75f96qk9j5nWzWl1tK3zJC9K1wHmlf1yT633pyb5SJKLk2xM8oaRx3+2f82uTXcx69AeA/xTVd0FUFWbqurWvpb3j6wjbx+p8YYk7+jf699I8oSR5XpzP7xv/7myrv8fTr2GZ6fbM/lqv+xH9e2PSfLl3L0X/NzBl7yqltwNuBNYP3K7EThjjsesAK7ph18FbKS7mG5n4LvAXv19FwOr+uFDgM/3w6fSbVlNze+afp6/BHwO2KFvfx/win64gBf1w+8E3jatphfRfcDtADwU2L5vPxT4dD98wtTjgJ2AtcA+A72ut/d/twf+HjixX8bNwAH9fecCL++HzwaO6odfAHyN7oPpiwtR2xzTfxz4zsg6srJv321kPpeMtF8KHAA8ul+f9p42/fHAe0bm/Xm6ja4n0nXlshPwIGDnfponAZeN/D9vBx7Xjz8BWD9S68kj/+NLgT/uh99EdwHoo/r19GbgF6Yt58uBv+yHvwo8fcLX8xD6dXuk7Qbg9f3wa4EP98MzrZttHnTvqTP64Ydz95dcjgf+bOQ99NX+tdod+CF3v2+mXudd6N5bjxj482DPfnnXA38GPG3kvqlaltF9Huw/8vq8tR9+BWM+G4ALgf364YPprs2C7r3yqX6dWUnXb9zU//itI8/3kCHe26O3pXr4qB1KgG4rBdjSqw0vrKrb+sdfR3eJ+E2zP2SsXwOeAVyebvd7F+D7/X0/o/vwAFhH98E5VfN+wOnA86vq50keDXysby+6oAD4t8D+U1sWdEG2H90H3nzbZWQL9hK660weC3ynqqba19EFxT1U1T8kOZqu59ynLlBtcxl3+OjYJMfRhcJj6d6g143c/0zgoqr6LkBV3esK/N651W1lXp/kJrr/yf8BzkjyVLog3Xdk+q9V1Y0T1Ax3H4a5Gri6qr4H3ZYq3YfZj0aXB3hPP3xOP37FhM8zztThwnXAb/bDD2P8ujmTPYG/SfIYYEfuua6eV1V3AHck+T5d4G0C3pDkJf00e9G9nj+csOYt/jyoqk39Hs/z+9uFSY6uqguB3+r3Vran26NYCVzVP/SvR/6+e3Se/V7bs4BP5e5DcTuNTPLZfp25bmrvie4i4I8k2aG/fz0DW6qhMB/uGBm+k7lfq83c83Dczv3fAB+rqv885jE/r34TYPQ5kjyYbov71VU11TXIH9F9GL0k3WGpi0fm//qqOn+uBZoH93hz9bXCvV+rXaY/MMl2dHtNPwV2o3ujD1rbluo/1H4POKiqfpTk49z9f2yTMaZ/rjGmT1N0W3030W2970C3dzBl9Hj1uHVp9DzD1Ot9F/d87e9iZD1N8gi6D7QnJym6Lc1K8gcj692Wmnq+0ffETOvmTN4LvKuqVqc7d3HqmPm35+inORR4ZlX9JMnF3Pv/Mu/6cPoC8IUk3wNenGQj8GbgwKq6NcnZ02qpGYah+5/+aJb1dHTZ09fw5STPA15Id6Hv6VX1V1u9UBPwnML8uQF4OkCSpwNT31S4EDgqySP7+3ZLMlfHVB8FPlpVl4y0PYxuSxO6XfEp5wMn9lsSJPnFPlTub94IfItuS3Vqy+f+5qHAvwA/7rdif33MNF8Bnj/1P0yy2wzzOjqdX6Tbsv023f/wn/oP5FcyvkNIgP8LPDbJw5PsTPeBsDWOAv6qqvauqhVVtRfdVvlztnJ+M5lp3Zxk+lfONuHI9Lf2gfAk4Fe2uMItlOTpSR7bD28H7E93GPmhdAF+W781f/i0h75s5O/XRu+oqh8D3+n3mOnXj1n3mvv17PtV9SG6vd+n36cFm8A2Fwrpvhp32gCz/jSwW38I40TgHwGq6jrgbcAFSa4C/oFul3Om+vamezP/Tu4+2byK7pzDnyT5Ct0W35QP0x3euCLdV0M/yP1sD7D/YDweeFMfdF+me00W4rnPnwrkCVxB91peA3yILgDuoT9UcyLw90m+CXxihnltoFvOzwEnVPebImcAxyf5Ot3hyDvGPbC6K/3/O92hg9Xc8/DVljgWmP6tqE8Dvw3dFxq2cr7TzbRuzuRUukMolzBZz6JfpNtjuIpur+TrW1vodLN8HjwS+Fz/nrqKbk/tjKr6JnAlcC3wEe69juyU5DK6Pc43jpnvvweO69eda7n3b8xMdwiwPsmVwEuBP59owe4Dr2iW5ll/2Olvx5yv0BLWn9NZVVWDd6E9pG1uT0GSNDP3FCRJjXsKkqTGUJAkNYaCJKkxFCQgyVv7vmyu6r8GfPA8zntNkl+Yr/lJQ7pffZ9dWgxJngn8Bl2/QHek6x56x/maf1X9u/malzQ09xSk7mLCH/TdGlBVP6iqmzNzr5fLk3w6Xc+0lyd5dt++a5KPputB9KokL+3b2+8QZEyPuf3t7Nzdd/+4i56kBWEoSHABsFe6bq7fl+RXR+77cVUdRHc18lTHcn8OvLuqDqS7yvTDfft/AW6rqqdU1f7Al0afJMkv0XV/8Oy+/5s76a5wPQDYo6qeXFVPoevmRFoUHj7SNq+qbk/yDOC5wL+h68Fz6lfCxvV6eSiwcqSny4cmeUjf3n6zoPr+90fM1GPu54DHJ3kvcB5dSEmLwlCQgKq6k653z4uTXM3dHbWN6/VyO7oeO386Oo90n/SzXQ06Y4+5fcdovw68Dvgt4He2YjGk+8zDR9rmJXli3232lAPoesSE8b1eXgCcNPL4A2Zof/i0pxrbY25/vmG7qvo03SGowXvClGbinoIEuwLv7b82upmuh9MT6L6RNNXr5XZ0vY4CvAE4s++1c3u63lBfA/y3vv0auvMFb2fkN6yr6rp0P6t5Qd8d88/p9gx+Cny0bwMY99sb0oKw7yNpBkul10tpS3j4SJLUuKcgSWrcU5AkNYaCJKkxFCRJjaEgSWoMBUlSYyhIkpr/D6ysaD+xP5Z1AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Use matplotlib to show the GC content of the different species with a bar plot\n",
    "plt.bar(['H. Influenzae', 'Phix', 'P. Falciparum', 'A. Thaliana', 'H. Sapiens'], gcs)\n",
    "plt.xlabel('Species')\n",
    "_ = plt.ylabel('GC count')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read Fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readFastq(file):\n",
    "    \"\"\" Read a FASTQ format file\n",
    "    The read adheres specified FASTQ format in initial Sanger papers,\n",
    "    that is, we expect that the sequence and scores could spread over several lines.\n",
    "    It's more complicated that the simplified case more common nowadays, which considers \n",
    "    one line for scores and one line for sequence.\n",
    "    Args:\n",
    "        file (str): path of the file to read\n",
    "    \n",
    "    Returns:\n",
    "        list of str: sequences\n",
    "        list of str: related qualities\n",
    "    \"\"\"\n",
    "    reads = []\n",
    "    quals = []\n",
    "    readsMode = False\n",
    "    with open(file, 'r') as f:\n",
    "        for line in f:\n",
    "            line = line.replace('\\n', '')\n",
    "            if len(line) > 0:\n",
    "                if line[0] == \"+\":\n",
    "                    if readsMode: #this is a reads line end\n",
    "                        readsMode = False\n",
    "                    else: #this is a quals line\n",
    "                        quals[-1] += line \n",
    "                elif line[0] == \"@\": \n",
    "                    #this is a quals line\n",
    "                    if len(reads) > 0 and (len(quals[-1]) + len(line) <= len(reads[-1])):\n",
    "                        quals[-1] += line\n",
    "                    #this is a comment\n",
    "                    else:  \n",
    "                        reads.append(\"\")\n",
    "                        quals.append(\"\")\n",
    "                        readsMode = True\n",
    "                elif readsMode:\n",
    "                    reads[-1] += line\n",
    "                else:\n",
    "                    quals[-1] += line\n",
    " \n",
    "    return reads, quals"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test\n",
    "Test with a file that spread sequence and quality scores over several lines. Also with empty lines in the middle."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "latex_metadata": {
     "lexer": "text"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting ugly.fastq\n"
     ]
    }
   ],
   "source": [
    "%%writefile ugly.fastq\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "@SRR835775.1 1/1\n",
    "TAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTCACC\n",
    "CTAACCCTAACCCTAACCGTATCCGTC\n",
    "ACCCTAACCCTAAC\n",
    "+\n",
    "???B1ADDD8??BB+C?B+:AA883CEE8?C3@DDD3)?D2;DC?8\n",
    "?=BAD=@C@(.6.6=A?=?@#######################\n",
    "###########\n",
    "@SRR835775.2 2/1\n",
    "TAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACC\n",
    "+\n",
    "CCCFFFFFGHHGHJJJJJIJGIIJJJJJJJIJIJJJJJFJJFGIIIIH=CBFCF=CCEG)=>EHB2@@DEC>;;?=;(=?BBD?59?BA###########\n",
    "\n",
    "\n",
    "\n",
    "@SRR835775.3 3/1\n",
    "TAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGG\n",
    "+\n",
    "@@<DD?DDHHH<CBHII:CFGIGAGHIIG?CCGGE0BDHIIHIGICH8=FHGI=@EHGGGEEHH>);?CC@.;(=;?59,5<A599?CB>ABBCB><88A\n",
    "@SRR835775.4 4/1\n",
    "TAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAAGGGTTGGGGGTTAGGGGTAGGGGTAGGGTTA\n",
    "+\n",
    "\n",
    "\n",
    "@CCFFDDF\n",
    "H\n",
    "HHDFHIJJCGGIJJHIIHJC?DHHIJ0?DGHI9BBFHICGGIGI=CDEGI=?AAEF7@?################################\n",
    "@SRR835775.5 5/1\n",
    "CTAACCCTAACCCTAACCCTAACCCTACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCAACCCTAACCCTAACCCTCACC\n",
    "+\n",
    "@<@FDFDDBBFHH@EBGGGEH@FFHE;EHIEGGBHG?)9?8BDGGBGGBCDGI=93=C6==C;CCD(?@>@#############################\n",
    "@SRR835775.6 6/1\n",
    "AACCCTAACCCTAACCCTAACCCTAACCCTAACCCCTAACCCTAACCCTACCCCTAACCCCCAACCCTCACACCAACCCTAACCCTACCCCCAACCCCAC\n",
    "+\n",
    "CCCFFDFDHGHHDHHGIIIHGIJ<GBBBFBGGHDIH@B)?)0?F>DB#####################################################\n",
    "@SRR835775.7 7/1\n",
    "TAGGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGCTGGGTTAGGGGTAGGGTTAGGGTTAGGGTTAGGGGTAGGAGTTCGGGAGAGCACACG\n",
    "+\n",
    "???=DD0@3ADB<<CE;+3<;EE?1:)1?C1?3BDCB@;)99)0;@;)-(.7A'55A1?.6;(6;.5;;??#############################\n",
    "@SRR835775.8 8/1\n",
    "TAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCCAACCCCAACCCCAACCCCAACCCTAACCCTAAACCCAAACCTAAA\n",
    "+\n",
    "@@@DDF?EFFHA;DDHGI<FDFGECCHF:C9DH9F3):??GHB)BFHEAHBG;C';C@EE45?#####################################\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8 8 800 800\n",
      "TAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTCACCCTAACCCTAACCCTAACCGTATCCGTCACCCTAACCCTAAC\n",
      "???B1ADDD8??BB+C?B+:AA883CEE8?C3@DDD3)?D2;DC?8?=BAD=@C@(.6.6=A?=?@##################################\n",
      "\n",
      "TAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAAGGGTTGGGGGTTAGGGGTAGGGGTAGGGTTA\n",
      "@CCFFDDFHHHDFHIJJCGGIJJHIIHJC?DHHIJ0?DGHI9BBFHICGGIGI=CDEGI=?AAEF7@?################################\n",
      "\n",
      "TAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCCAACCCCAACCCCAACCCCAACCCTAACCCTAAACCCAAACCTAAA\n",
      "@@@DDF?EFFHA;DDHGI<FDFGECCHF:C9DH9F3):??GHB)BFHEAHBG;C';C@EE45?#####################################\n"
     ]
    }
   ],
   "source": [
    "r, q = readFastq('ugly.fastq')\n",
    "print(len(r), len(q), len(''.join(r)),  len(''.join(q)))\n",
    "# Length of all secuences and scores needs to be equal\n",
    "assert len(''.join(r)) == len(''.join(q))\n",
    "print(r[0])\n",
    "print(q[0])\n",
    "print()\n",
    "print(r[3])\n",
    "print(q[3])\n",
    "print()\n",
    "# Check the last one\n",
    "print(r[-1])\n",
    "print(q[-1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Histogram of Q"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZUAAAEKCAYAAADaa8itAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4wLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvqOYd8AAAFt5JREFUeJzt3X/wXXV95/Hnq0Ep1h8BiQ6boMGacUXHnymyteNYsBDAMbgLXZyuZJVOZh380Z3O1tDOLFuVXWyrqNPqDCvRaB2RpXbJKEozAcfZmRUJQpEf0mSBSoQ10YCwo8UG3/vH/Xzx5pv7Dd/ke7733Euej5k733M+53Pufd8z5L44n/O556aqkCSpC7/SdwGSpKcOQ0WS1BlDRZLUGUNFktQZQ0WS1BlDRZLUGUNFktQZQ0WS1BlDRZLUmSP6LmDcjj322Fq5cmXfZUjSVLn55pt/VFXLnqzfYRcqK1euZNu2bX2XIUlTJck/zqefw1+SpM4YKpKkzixaqCTZmGRXktuH2v48yfeS3Jbkb5MsHdp2UZIdSe5OcvpQ+5rWtiPJhqH2E5LcmGR7ki8lefpivRdJ0vws5pnKZ4E1s9q2AC+vqlcA/wBcBJDkROA84GVtn08mWZJkCfBXwBnAicDbWl+ADwOXVdUq4CHggkV8L5KkeVi0UKmqbwJ7ZrX9XVXtbavfAla05bXAlVX1WFXdC+wATmqPHVV1T1X9HLgSWJskwCnA1W3/TcDZi/VeJEnz0+c1lXcCX2vLy4H7h7btbG1ztT8XeHgooGbaR0qyPsm2JNt2797dUfmSpNl6CZUkfwLsBb4w0zSiWx1C+0hVdXlVra6q1cuWPek0a0nSIRr791SSrAPeDJxav/wt453A8UPdVgAPtOVR7T8CliY5op2tDPeXJPVkrGcqSdYA7wfeUlU/Hdq0GTgvyZFJTgBWAd8GbgJWtZleT2dwMX9zC6MbgHPa/uuAa8b1PiRJoy3amUqSLwJvBI5NshO4mMFsryOBLYNr7Xyrqv5DVd2R5CrgTgbDYhdW1ePted4NXAcsATZW1R3tJd4PXJnkQ8AtwBWL9V4kHX5WbvjqyPb7Lj1rzJVMl0ULlap624jmOT/4q+oS4JIR7dcC145ov4fB7DBJ0oTwG/WSpM4YKpKkzhgqkqTOGCqSpM4YKpKkzhgqkqTOGCqSpM4YKpKkzhgqkqTOGCqSpM4YKpKkzhgqkqTOGCqSpM4YKpKkzhgqkqTOGCqSpM4YKpKkzhgqkqTOGCqSpM4YKpKkzhgqkqTOGCqSpM4YKpKkzhgqkqTOGCqSpM4YKpKkzixaqCTZmGRXktuH2o5JsiXJ9vb36NaeJJ9IsiPJbUleM7TPutZ/e5J1Q+2vTfLdts8nkmSx3oskaX4W80zls8CaWW0bgK1VtQrY2tYBzgBWtcd64FMwCCHgYuB1wEnAxTNB1PqsH9pv9mtJksZs0UKlqr4J7JnVvBbY1JY3AWcPtX+uBr4FLE1yHHA6sKWq9lTVQ8AWYE3b9uyq+t9VVcDnhp5LktSTcV9TeX5VPQjQ/j6vtS8H7h/qt7O1Hah954j2kZKsT7Itybbdu3cv+E1IkkablAv1o66H1CG0j1RVl1fV6qpavWzZskMsUZL0ZMYdKj9sQ1e0v7ta+07g+KF+K4AHnqR9xYh2SVKPxh0qm4GZGVzrgGuG2s9vs8BOBn7ShseuA05LcnS7QH8acF3b9miSk9usr/OHnkuS1JMjFuuJk3wReCNwbJKdDGZxXQpcleQC4PvAua37tcCZwA7gp8A7AKpqT5IPAje1fh+oqpmL/+9iMMPsKOBr7SFJ6tGihUpVvW2OTaeO6FvAhXM8z0Zg44j2bcDLF1KjJKlbk3KhXpL0FGCoSJI6Y6hIkjpjqEiSOmOoSJI6Y6hIkjqzaFOKJUmLY+WGr45sv+/Ss8Zcyf48U5EkdcZQkSR1xlCRJHXGUJEkdcZQkSR1xlCRJHXGUJEkdcZQkSR1xlCRJHXGUJEkdcZQkSR1xlCRJHXGUJEkdcZQkSR1xlCRJHXGUJEkdcYf6ZJ02JrkH7uaVp6pSJI6Y6hIkjrTy/BXkv8I/D5QwHeBdwDHAVcCxwDfAd5eVT9PciTwOeC1wI+Bf1tV97XnuQi4AHgceG9VXTfmtyJJnVvIsFzfQ3pjP1NJshx4L7C6ql4OLAHOAz4MXFZVq4CHGIQF7e9DVfVi4LLWjyQntv1eBqwBPplkyTjfiyRpX30Nfx0BHJXkCOAZwIPAKcDVbfsm4Oy2vLat07afmiSt/cqqeqyq7gV2ACeNqX5J0ghjD5Wq+gHwF8D3GYTJT4CbgYeram/rthNY3paXA/e3ffe2/s8dbh+xjySpB30Mfx3N4CzjBOBfAL8GnDGia83sMse2udpHveb6JNuSbNu9e/fBFy1Jmpc+hr/eBNxbVbur6p+BLwO/CSxtw2EAK4AH2vJO4HiAtv05wJ7h9hH77KOqLq+q1VW1etmyZV2/H0lS00eofB84Ockz2rWRU4E7gRuAc1qfdcA1bXlzW6dtv76qqrWfl+TIJCcAq4Bvj+k9SJJGGPuU4qq6McnVDKYN7wVuAS4HvgpcmeRDre2KtssVwOeT7GBwhnJee547klzFIJD2AhdW1eNjfTOSNIe+p/b2pZfvqVTVxcDFs5rvYcTsrar6J+DcOZ7nEuCSzguUJB0Sv1EvSeqMoSJJ6oyhIknqjKEiSeqMoSJJ6oyhIknqjKEiSeqMoSJJ6syThkqSz8+nTZKk+ZypvGx4pf0Q1msXpxxJ0jSbM1SSXJTkUeAVSR5pj0eBXfzyZo+SJD1hzlCpqv9WVc8C/ryqnt0ez6qq51bVRWOsUZI0JZ70hpJVdVH7XfkXDvevqm8uZmGSpOnzpKGS5FIGt5u/E5i5tXwBhookaR/zufX9W4GXVNVji12MJGm6zWf21z3A0xa7EEnS9JvPmcpPgVuTbAWeOFupqvcuWlWSpKk0n1DZ3B6SJB3QfGZ/bRpHIZKk6Tef2V/3MpjttY+qetGiVCRJT3ErN3x1ZPt9l5415kq6N5/hr9VDy78KnAscszjlSJKm2ZPO/qqqHw89flBVHwNOGUNtkqQpM5/hr9cMrf4KgzOXZy1aRZKkqTWf4a+PDC3vBe4DfndRqpEkTbX5zP767XEUIkmafvP5ka7nJPlokm3t8ZEkzxlHcZKk6TKf27RsBB5lMOT1u8AjwGcW8qJJlia5Osn3ktyV5F8lOSbJliTb29+jW98k+USSHUluG77Gk2Rd6789ybqF1CRJWrj5XFP59ar6N0Prf5rk1gW+7seBr1fVOUmeDjwD+GNga1VdmmQDsAF4P3AGsKo9Xgd8CnhdkmOAixlMHCjg5iSbq+qhBdY2p6fy3HJJ6sJ8zlR+luS3ZlaSvB742aG+YJJnA28ArgCoqp9X1cPAWmDm2/ubgLPb8lrgczXwLWBpkuOA04EtVbWnBckWYM2h1iVJWrj5nKm8C9g0dB3lIeDfL+A1XwTsBj6T5JXAzcD7gOdX1YMAVfVgkue1/suB+4f239na5mqXJPVkPrO/bgVe2c4wqKpHOnjN1wDvqaobk3ycwVDXXDKqrAO07/8EyXpgPcALXvCCg6tWkjRv85n99V+TLK2qR6rqkSRHJ/nQAl5zJ7Czqm5s61czCJkftmEt2t9dQ/2PH9p/BfDAAdr3U1WXV9Xqqlq9bNmyBZQuSTqQ+VxTOaNd8wCgXb8481BfsKr+L3B/kpe0plMZ/FTxZmBmBtc64Jq2vBk4v80COxn4SRsmuw44rYXc0cBprU2S1JP5XFNZkuTImZ8TTnIUcOQCX/c9wBfazK97gHcwCLirklwAfJ/BjSsBrmUQYjsY/GDYOwCqak+SDwI3tX4fqKo9C6xLkrQA8wmVvwa2JvkMg2sW7+SXs7QOSbtOs3rEplNH9C3gwjmeZyOD79FIkibAfC7U/1mS24A3Mbg4/sGqcphJkrSf+ZypUFVfB76+yLVIkqbcfC7US5I0L4aKJKkz8/mRrmcAL26rd8/MApOkSef9+sZvzjOVJE9L8jEGXzL8DIMZX/e0mz2S5NXjKVGSNC0OdKbyEQZ3D35hVT0KT9wM8i+SfIrBzRtPWPwSJUnT4kChciawqn1PBBjc9yvJu4AfMbglvSRJTzjQhfpfDAfKjKp6HNjdbkMvSdITDhQqdyY5f3Zjkn8H3LV4JUmSptWBhr8uBL6c5J0MfvOkgN8AjgLeOobaJElTZs5QqaofMPjZ3lOAlzG4RcvXqmrruIqTJE2X+dz763rg+jHUIkmacn6jXpLUGUNFktQZQ0WS1BlDRZLUGUNFktQZQ0WS1BlDRZLUGUNFktSZef1GvST1xR/ami6eqUiSOmOoSJI6Y6hIkjrjNRVJOgRe6xmttzOVJEuS3JLkK239hCQ3Jtme5EtJnt7aj2zrO9r2lUPPcVFrvzvJ6f28E0nSjD6Hv97Hvr8g+WHgsqpaBTwEXNDaLwAeqqoXA5e1fiQ5ETiPwW+9rAE+mWTJmGqXJI3QS6gkWQGcBXy6rQc4Bbi6ddkEnN2W17Z12vZTW/+1wJVV9VhV3QvsAE4azzuQJI3S15nKx4A/An7R1p8LPFxVe9v6TmB5W14O3A/Qtv+k9X+ifcQ+kqQejD1UkrwZ2FVVNw83j+haT7LtQPvMfs31SbYl2bZ79+6DqleSNH99nKm8HnhLkvuAKxkMe30MWJpkZjbaCuCBtrwTOB6gbX8OsGe4fcQ++6iqy6tqdVWtXrZsWbfvRpL0hLGHSlVdVFUrqmolgwvt11fV7wE3AOe0buuAa9ry5rZO2359VVVrP6/NDjsBWAV8e0xvQ5I0wiR9T+X9wJVJPgTcAlzR2q8APp9kB4MzlPMAquqOJFcBdwJ7gQur6vHxly1JmtFrqFTVN4BvtOV7GDF7q6r+CTh3jv0vAS5ZvAolSQfD27RIkjpjqEiSOmOoSJI6Y6hIkjpjqEiSOjNJU4olHYa8hfxTi2cqkqTOGCqSpM44/CVpajl0Nnk8U5EkdcZQkSR1xlCRJHXGUJEkdcZQkSR1xlCRJHXGUJEkdcZQkSR1xlCRJHXGUJEkdcZQkSR1xlCRJHXGUJEkdcZQkSR1xlCRJHXGUJEkdcZQkSR1ZuyhkuT4JDckuSvJHUne19qPSbIlyfb29+jWniSfSLIjyW1JXjP0XOta/+1J1o37vUiS9tXHmcpe4A+r6qXAycCFSU4ENgBbq2oVsLWtA5wBrGqP9cCnYBBCwMXA64CTgItngkiS1I+xh0pVPVhV32nLjwJ3AcuBtcCm1m0TcHZbXgt8rga+BSxNchxwOrClqvZU1UPAFmDNGN+KJGmWXq+pJFkJvBq4EXh+VT0Ig+ABnte6LQfuH9ptZ2ubq12S1JPeQiXJM4G/Af6gqh45UNcRbXWA9lGvtT7JtiTbdu/effDFSpLmpZdQSfI0BoHyhar6cmv+YRvWov3d1dp3AscP7b4CeOAA7fupqsuranVVrV62bFl3b0SStI8+Zn8FuAK4q6o+OrRpMzAzg2sdcM1Q+/ltFtjJwE/a8Nh1wGlJjm4X6E9rbZKknhzRw2u+Hng78N0kt7a2PwYuBa5KcgHwfeDctu1a4ExgB/BT4B0AVbUnyQeBm1q/D1TVnvG8BUnSKGMPlar6X4y+HgJw6oj+BVw4x3NtBDZ2V52krq3c8NU5t9136VljrETj4DfqJUmdMVQkSZ0xVCRJnTFUJEmd6WP2l6QpM9fFdi+0azZDRZIztNQZh78kSZ0xVCRJnTFUJEmdMVQkSZ0xVCRJnTFUJEmdcUqx9BTR53dJ/B6LZnimIknqjGcq0mHCswmNg2cqkqTOeKYiTRHPNjTpPFORJHXGUJEkdcZQkSR1xmsqTwGOs0uaFIaKNMtCQ9qQ1+HM4S9JUmcMFUlSZxz+Ogw4HLO/ST0mk1qXNF+GijRGhoae6qY+VJKsAT4OLAE+XVWX9lzSU8qkfgh6R15pMk11qCRZAvwV8DvATuCmJJur6s5+K9N8LOaHsx/8Uj+mOlSAk4AdVXUPQJIrgbXAUy5UFutDcq7nne9zH6iuhT63pOkz7aGyHLh/aH0n8LqealkQ/896fx4Tafqkqvqu4ZAlORc4vap+v62/HTipqt4zq996YH1bfQlwd1s+FvjRmMo9GNZ1cCa1Lpjc2qzr4ExqXTC+2l5YVcuerNO0n6nsBI4fWl8BPDC7U1VdDlw+uz3JtqpavXjlHRrrOjiTWhdMbm3WdXAmtS6YvNqm/cuPNwGrkpyQ5OnAecDmnmuSpMPWVJ+pVNXeJO8GrmMwpXhjVd3Rc1mSdNia6lABqKprgWsPcff9hsQmhHUdnEmtCya3Nus6OJNaF0xYbVN9oV6SNFmm/ZqKJGmCHJahkmRNkruT7Eiyoe96hiW5L8l3k9yaZFuPdWxMsivJ7UNtxyTZkmR7+3v0hNT1X5L8oB2zW5Oc2UNdxye5IcldSe5I8r7W3usxO0Bdk3DMfjXJt5P8favtT1v7CUlubMfsS20SziTU9dkk9w4ds1eNs66h+pYkuSXJV9p6r8drtsMuVIZu7XIGcCLwtiQn9lvVfn67ql7V8zTBzwJrZrVtALZW1Spga1sft8+yf10Al7Vj9qp2nW3c9gJ/WFUvBU4GLmz/XfV9zOaqC/o/Zo8Bp1TVK4FXAWuSnAx8uNW2CngIuGBC6gL4T0PH7NYx1zXjfcBdQ+t9H699HHahwtCtXarq58DMrV00pKq+CeyZ1bwW2NSWNwFnj7Uo5qyrd1X1YFV9py0/yuAf/XJ6PmYHqKt3NfD/2urT2qOAU4CrW3sfx2yuunqXZAVwFvDpth56Pl6zHY6hMurWLhPxj6wp4O+S3NzuBDBJnl9VD8Lgwwp4Xs/1DHt3ktva8NjYh+WGJVkJvBq4kQk6ZrPqggk4Zm0o51ZgF7AF+D/Aw1W1t3Xp5d/n7LqqauaYXdKO2WVJjhx3XcDHgD8CftHWn8sEHK9hh2OoZETbRPxfSPP6qnoNg+G5C5O8oe+CpsCngF9nMFTxIPCRvgpJ8kzgb4A/qKpH+qpjthF1TcQxq6rHq+pVDO6GcRLw0lHdxlvV/nUleTlwEfAvgd8AjgHeP86akrwZ2FVVNw83j+ja6+fZ4Rgq87q1S1+q6oH2dxfwtwz+oU2KHyY5DqD93dVzPQBU1Q/bh8AvgP9OT8csydMYfHB/oaq+3Jp7P2aj6pqUYzajqh4GvsHgus/SJDPfoev13+dQXWvaUGJV1WPAZxj/MXs98JYk9zEYtj+FwZnLxBwvODxDZWJv7ZLk15I8a2YZOA24/cB7jdVmYF1bXgdc02MtT5j50G7eSg/HrI1tXwHcVVUfHdrU6zGbq64JOWbLkixty0cBb2JwzecG4JzWrY9jNqqu7w39z0EYXLcY6zGrqouqakVVrWTwuXV9Vf0ePR+v/VTVYfcAzgT+gcH47Z/0Xc9QXS8C/r497uizNuCLDIZF/pnB2d0FDMZvtwLb299jJqSuzwPfBW5j8CF+XA91/RaDYYfbgFvb48y+j9kB6pqEY/YK4JZWw+3Af27tLwK+DewA/gdw5ITUdX07ZrcDfw08c9zHbKjGNwJfmYTjNfvhN+olSZ05HIe/JEmLxFCRJHXGUJEkdcZQkSR1xlCRJHXGUJF6lmRFkmvaXWbvSfKXPd0CRFowQ0XqUfsi3ZeB/1mDu8yuAo4C/qzXwqRD5PdUpB4lORW4uKreMNT2bOAfgePrl3fLlaaCZypSv14GDN8gkBrc8PE+4MV9FCQthKEi9SuMvqvsqLvPShPPUJH6dQewzy98tuGv5wN391KRtACGitSvrcAzkpwPT/zc9UeAv6yqn/VamXQIDBWpRzWYKfNW4Jwk24EfA7+oqkv6rUw6NM7+kiZIkt9kcHv/f137/sKfNBUMFUlSZxz+kiR1xlCRJHXGUJEkdcZQkSR1xlCRJHXGUJEkdcZQkSR15v8DZLFFG4ZBfd0AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Use matplotlib to show the a histogram of Qs\n",
    "import math\n",
    "rr, qq = readFastq('phix174/SRR835775_1.first1000.fastq')\n",
    "#%time rr, qq = readFastq('sintetico/sample_01.fasta.fastq')\n",
    "q = ''.join(qq)\n",
    "quals = {}\n",
    "i = 0\n",
    "TOP = 10000000\n",
    "#math.pow(10,-(ord('@')-33)/10)\n",
    "for e in q:\n",
    "    m = ord(e) - 33\n",
    "    if m in quals:\n",
    "        quals[m] += 1\n",
    "    else:\n",
    "        quals[m] = 0\n",
    "    i += 1\n",
    "    if i > TOP: break\n",
    "\n",
    "plt.bar(list(quals.keys()), list(quals.values()))\n",
    "plt.xlabel('Q')\n",
    "_ = plt.ylabel('Q count')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Count bases (spark)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['A', 'C', 'G', 'T'], [0, 0, 0, 0])"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Implement a sequence of spark transformations to read a fasta file \n",
    "#   and to compute the number of occurrences of each base. \n",
    "# \n",
    "# You can use count countBases inside a tranformation instead of a \n",
    "# lambda\n",
    "\n",
    "def count_bases_spark_from_fasta(file, baselist=['A', 'C', 'G', 'T']):\n",
    "    \"\"\" Counts the number of occurrences of each base in a fasta file.\n",
    "        There are two return values: the list of bases and the list of counts\n",
    "    \"\"\"\n",
    "    fasta_lines = sc.textFile(file)\n",
    "    data_lines = fasta_lines.filter(lambda line: line[0] != \">\")\n",
    "    bases = data_lines.flatMap(lambda line: list(line))\n",
    "    base_duples = bases.map(lambda base: (base, 1))\n",
    "    base_counts = base_duples.reduceByKey(lambda v1, v2 : v1 + v2)\n",
    "    # Format response\n",
    "    bases = []\n",
    "    counts = []\n",
    "    for (base, count) in base_counts.collect():\n",
    "        if base in baselist:\n",
    "            bases.append(base)\n",
    "            counts.append(count)\n",
    "    for base in baselist:\n",
    "        if base not in bases:\n",
    "            bases.append(base)\n",
    "            counts.append(0)\n",
    "    return bases, counts\n",
    "\n",
    "count_bases_spark_from_fasta(\"wide.fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tests\n",
    "Files for test:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting p1.fasta\n"
     ]
    }
   ],
   "source": [
    "%%writefile p1.fasta\n",
    ">NC_000907.1\n",
    "ATGAAAACATTATTCACTTCGGTTGTATTGTGTGGTGCGCTGGTTGTTTCTTCGTCTTTTGCTGAGGAAA\n",
    "AAGCGACAGANCAAACCGCTCAATCTGTTGTGACAACTCAAGCTGAAGCTCAAGTAGCACCAGCCGTAGT\n",
    "GAGCGATAAATTGAATATCAACACAGCAACTGCCAGTGAAATTCAAAAATCCTTAACTGGCATTGGTGCG\n",
    "AAAAAAGCGGAAGCTATTGTGCAATATCGTGAAAAACACGGTAATTTTKCTAATGCAGAACAGCTTTTAG\n",
    "AAGTACAAGGAATTGGCAAAGCAACACTAGAGAAAAATCGTGATCGTATAATCTTTTAACCCCC\n",
    ">NC_000907.2\n",
    "AAA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting p2.fasta\n"
     ]
    }
   ],
   "source": [
    "%%writefile p2.fasta\n",
    ">NC_000907.1\n",
    "ATCG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting wide.fasta\n"
     ]
    }
   ],
   "source": [
    "%%writefile wide.fasta\n",
    ">NC_000907.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['C', 'A', 'T', 'G'] [1, 1, 1, 1]\n",
      "['A', 'C', 'G', 'T'] [1, 1, 1, 1]\n",
      "Test passed for file p2.fasta\n",
      "\n",
      "['C', 'A', 'T', 'G'] [66, 118, 89, 72]\n",
      "['A', 'C', 'G', 'T'] [118, 66, 72, 89]\n",
      "Test passed for file p1.fasta\n",
      "\n",
      "['A', 'C', 'G', 'T'] [0, 0, 0, 0]\n",
      "['A', 'C', 'G', 'T'] [0, 0, 0, 0]\n",
      "Test passed for file wide.fasta\n",
      "\n"
     ]
    }
   ],
   "source": [
    "def test(file, base_list=['A', 'C', 'G', 'T']):\n",
    "    \"\"\"\n",
    "    Verification versus result of count bases not spark\n",
    "    \"\"\"\n",
    "    bases_spark, counts_spark = count_bases_spark_from_fasta(file, base_list)\n",
    "    print(bases_spark, counts_spark)\n",
    "\n",
    "    #Calc no spark\n",
    "    seqs = readFasta(file)\n",
    "    sequence = \"\".join(list(seqs.values()))\n",
    "    bases_no_spark, counts_no_spark = countBases(sequence, base_list)\n",
    "    print(bases_no_spark, counts_no_spark)\n",
    "\n",
    "    #Verification\n",
    "    compares = {}\n",
    "    for (i, base) in enumerate(bases_spark):\n",
    "        compares[base] = counts_spark[i]\n",
    "\n",
    "    for (i, base) in enumerate(bases_no_spark):\n",
    "        assert compares[base] == counts_no_spark[i]\n",
    "    print(\"Test passed for file \" + file)\n",
    "    print()\n",
    "    \n",
    "test(\"p2.fasta\")\n",
    "test(\"p1.fasta\")\n",
    "test(\"wide.fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute inverse complementary(spark)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def complementary(gen):\n",
    "    \"\"\" From a sequence of nucleotides return the inverse complementary\n",
    "          e.g. GAACCT should return AGGTTC\n",
    "          Hint: Use a dictionary for the translations\n",
    "        Args:\n",
    "            gen (str): nucleotide sequence\n",
    "        \n",
    "        Returns:\n",
    "            str: inverse complementary of input sequence\n",
    "        \"\"\"\n",
    "    traslation = { 'A' : 'T', 'C' : 'G', 'T' : 'A', 'G' : 'C',\n",
    "                   'R' : 'Y',  'Y' : 'R',\n",
    "                   'S' : 'S',\n",
    "                   'W' : 'W',\n",
    "                   'K' : 'M', 'M' : 'K',\n",
    "                   'B' : 'V', 'V' : 'B',\n",
    "                   'D' : 'H',  'H' : 'D',\n",
    "                   'N' : 'N', 'X' : 'X'}\n",
    "                  \n",
    "    bases = sc.parallelize(gen)\n",
    "    bases_complementary = bases.map(lambda base: traslation[base])\n",
    "    bases_complementary_reverse_join = bases_complementary.reduce(lambda v1,v2 : v2 + v1)\n",
    "    return bases_complementary_reverse_join"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All tests OK!\n"
     ]
    }
   ],
   "source": [
    "# Tests\n",
    "assert(complementary('G')=='C')\n",
    "assert(complementary('AT')=='AT')\n",
    "assert(complementary('GC')=='GC')\n",
    "assert(complementary('GATACA')=='TGTATC')\n",
    "# If you added all possible combination of bases uncomment the \n",
    "# following line to check\n",
    "assert(complementary('GATACASWKMYRHDVBN')=='NVBHDYRKMWSTGTATC')\n",
    "\n",
    "print('All tests OK!')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reading CDS from GFF (spark)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extractProteinId(attribs):\n",
    "    \"\"\" Extract the protein_id from the complete string of the attributes \n",
    "        field of the gff \"\"\"\n",
    "    protein_id = attribs[attribs.find(\"protein_id=\")+len(\"protein_id=\"):]\n",
    "    i = protein_id.find(\";\")\n",
    "    if i>=0:\n",
    "        protein_id = protein_id[:i]\n",
    "        \n",
    "    return protein_id\n",
    "    \n",
    "def extractCDSFromGFF(gff_file):\n",
    "    \"\"\" This function extract the regions information from a gff file for all CDSs.\n",
    "        \n",
    "        This function returns a dictionary with the protein_id being the keys and \n",
    "        the values in a list with the sequence_id and tuples with (start, end, strand, frame)\n",
    "        \n",
    "        Note that a CDS might spread over multiple lines. For each line you should include \n",
    "        one tuple.\n",
    "        \n",
    "        Check the tests in the next cell for reference        \n",
    "    \"\"\"\n",
    "    cds_regions = {}\n",
    "    gff_cds_regions = sc.textFile(gff_file).\\\n",
    "        filter(lambda line: len(line) > 0 and line[0] != '#').\\\n",
    "        map(lambda line: line.split(\"\\t\")).\\\n",
    "        filter(lambda line: line[2] == 'CDS').\\\n",
    "        map(lambda line: [extractProteinId(line[8]), line[0],\\\n",
    "                          (int(line[3]), int(line[4]), line[6], int(line[7]))])\n",
    "    # reformat the list as dict \n",
    "    for region in gff_cds_regions.collect():\n",
    "        if region[0] in cds_regions:\n",
    "            cds_regions[region[0]].append(region[2])\n",
    "        else:\n",
    "            cds_regions[region[0]] = [region[1], region[2]]\n",
    "    return (cds_regions)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tests\n",
    "This is a file for tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Overwriting test.gff\n"
     ]
    }
   ],
   "source": [
    "%%writefile test.gff\n",
    "##gff-version 3\n",
    "#!gff-spec-version 1.21\n",
    "#!processor NCBI annotwriter\n",
    "#!genome-build ASM2730v1\n",
    "#!genome-build-accession NCBI_Assembly:GCF_000027305.1\n",
    "##sequence-region NC_000907.1 1 1830138\n",
    "##species https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=71421\n",
    "NC_000907.1\tRefSeq\tregion\t1\t1830138\t.\t+\t.\tID=id0;Dbxref=taxon:71421;Is_circular=true;Name=ANONYMOUS;gbkey=Src;genome=chromosome;mol_type=genomic DNA;old-name=Haemophilus influenzae Rd\n",
    "NC_000907.1\tRefSeq\tremark\t1\t1830138\t.\t+\t.\tID=id1;Note=REFSEQ gene predictions performed by GeneMark 2.4/GeneMark.hmm 2.0 with comparison to original submitter provided annotation%3B tRNA sequences predicted using tRNAscan-SE 1.21;gbkey=Comment\n",
    "NC_000907.1\tRefSeq\tgene\t2\t1021\t.\t+\t.\tID=gene0;Dbxref=GeneID:950899;Name=gapdH;gbkey=Gene;gene=gapdH;gene_biotype=protein_coding;locus_tag=HI0001\n",
    "NC_000907.1\tRefSeq\tCDS\t2\t1021\t.\t+\t0\tID=cds0;Parent=gene0;Dbxref=Genbank:NP_438174.1,GeneID:950899;Name=NP_438174.1;Note=catalyzes the formation of 3-phospho-D-glyceroyl phosphate from D-glyceraldehyde 3-phosphate;gbkey=CDS;gene=gapdH;product=glyceraldehyde-3-phosphate dehydrogenase;protein_id=NP_438174.1;transl_table=11\n",
    "NC_000907.1\tRefSeq\tgene\t104277\t105274\t.\t+\t.\tID=gene101;Dbxref=GeneID:950998;Name=hitA;gbkey=Gene;gene=hitA;gene_biotype=protein_coding;locus_tag=HI0097\n",
    "NC_000907.1\tRefSeq\tCDS\t104277\t104318\t.\t+\t0\tID=cds94;Parent=gene101;Dbxref=Genbank:NP_438271.2,GeneID:950998;Name=NP_438271.2;Note=periplasmic ferric iron binding protein%3B contains a frameshift with respect to purified protein%3B similar to PID:2340838;gbkey=CDS;gene=hitA;product=iron-utilization periplasmic protein hFbpA;protein_id=NP_438271.2;transl_table=11\n",
    "NC_000907.1\tRefSeq\tCDS\t104318\t105274\t.\t+\t0\tID=cds94;Parent=gene101;Dbxref=Genbank:NP_438271.2,GeneID:950998;Name=NP_438271.2;Note=periplasmic ferric iron binding protein%3B contains a frameshift with respect to purified protein%3B similar to PID:2340838;gbkey=CDS;gene=hitA;product=iron-utilization periplasmic protein hFbpA;protein_id=NP_438271.2;transl_table=11\n",
    "\n",
    "NC_000907.1\tRefSeq\tCDS\t1044343\t11\t.\t+\t0\tID=cds94;Parent=gene101;Dbxref=Genbank:NP_438271.2,GeneID:950998;Name=NP_438271.2;Note=periplasmic ferric iron binding protein%3B contains a frameshift with respect to purified protein%3B similar to PID:2340838;gbkey=CDS;gene=hitA;product=iron-utilization periplasmic protein hFbpA;protein_id=NP_438271.2;transl_table=11\n",
    "###\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'NP_438174.1': ['NC_000907.1', (2, 1021, '+', 0)],\n",
       " 'NP_438271.2': ['NC_000907.1',\n",
       "  (104277, 104318, '+', 0),\n",
       "  (104318, 105274, '+', 0),\n",
       "  (1044343, 11, '+', 0)]}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All test are OK!\n"
     ]
    }
   ],
   "source": [
    "display(extractCDSFromGFF('test.gff'))\n",
    "\n",
    "# Tests\n",
    "assert(extractCDSFromGFF('haemophilus_influenzae/test1.gff')=={'NP_438174.1': ['NC_000907.1', (2, 1021, '+', 0)]})\n",
    "assert(extractCDSFromGFF('haemophilus_influenzae/test2.gff')== {'NP_438174.1': ['NC_000907.1',  (2, 1021, '+', 0)],\n",
    "                                                  'NP_438175.1': ['NC_000907.1',  (1190, 3013, '+', 0)],\n",
    "                                                  'NP_438176.1': ['NC_000907.1', (3050, 3838, '-', 0)]})\n",
    "assert(extractCDSFromGFF('haemophilus_influenzae/test3.gff')=={'NP_438174.1': ['NC_000907.1', (2, 1021, '+', 0)],\n",
    "                 'NP_438271.2': ['NC_000907.1', (104277, 104318, '+', 0), (104318, 105274, '+',  0)]})\n",
    "print('All test are OK!')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Obtain translation table from cds and protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def obtainTranslationTable(cds, protein):\n",
    "    \"\"\" Creates a translation dictionary from the protein and DNA sequence.\n",
    "        The function returns:\n",
    "            1- A dictionary with the translation table where the codons are the\n",
    "               keys of the dictionary and the amino acids the values.\n",
    "            2- The start codon as a string\n",
    "            \n",
    "        Note 1: The sequence cds includes start and end codons. \n",
    "        Note 2: First codon always codifies M that is also included in the protein.\n",
    "        Note 3: End codon do not codify an a aminoacid and should be included in the\n",
    "                dictionary as *.\n",
    "                \n",
    "        Example: (Spaces added in seqeunces for readability) \n",
    "          cds     = GTG TTT TTG GTG TGG TGA\n",
    "          protein = M   F   L   V   W\n",
    "          \n",
    "          should return\n",
    "            tranlation dictionary = {'TTT':'F', 'TTG':'L', 'GTG':'V', 'TGG':'W', 'TGA':'*'}\n",
    "            start codon = 'GTG'\n",
    "        \"\"\"\n",
    "    trans_table = {}\n",
    "    start_codon = \"\"\n",
    "    for i in range(0, len(cds), 3):\n",
    "        codon = cds[i:i+3]\n",
    "        if i == 0:\n",
    "            start_codon = codon\n",
    "        if len(codon) == 3:\n",
    "            i_protein = i // 3\n",
    "            if i_protein >= len(protein):\n",
    "                trans_table[codon] = '*'\n",
    "            else:\n",
    "                trans_table[codon] = protein[i_protein]\n",
    "    \n",
    "    return trans_table, start_codon"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All test OK!\n"
     ]
    }
   ],
   "source": [
    "translation_dictionary, start_codon = obtainTranslationTable('GTGTTTTTGGTGTGGTGA','MFLVW')\n",
    "assert(translation_dictionary=={'TTT':'F', 'TTG':'L', 'GTG':'V', 'TGG':'W', 'TGA':'*'})\n",
    "assert(start_codon == 'GTG')\n",
    "\n",
    "print('All test OK!')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compose mRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compose_mRNA(cromosome, cds):\n",
    "    \"\"\"Returns the mRNA from the complete sequence of DNA and the locations of the CDS. \n",
    "        Parameters:\n",
    "            -cromosome: a string with the whole cromosome\n",
    "            -cds: a list containing tuples of (start, end, strand, phase) \"\"\"\n",
    "    mrna = ''\n",
    "    #print(cds)\n",
    "    for (start, end, strand, phase) in cds:\n",
    "        if strand == '+':\n",
    "            mrna += cromosome[start + phase - 1: end]\n",
    "        else:\n",
    "            mrna += complementary(cromosome[start - 1: end - phase])\n",
    "        \n",
    "    return mrna\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All tests are OK!\n"
     ]
    }
   ],
   "source": [
    "cod = 'GTGTTTTTGGTGTGGTGA'\n",
    "\n",
    "assert(compose_mRNA(cod,[(1,len(cod),'+',0)])==cod)\n",
    "assert(compose_mRNA(cod,[(1,len(cod),'-',0)])==complementary(cod))\n",
    "assert(compose_mRNA(\"GGGGGGGGGG\"+cod+\"GGGGGG\",[(11,10+len(cod),'+',0)])==cod)\n",
    "assert(compose_mRNA(\"GGGGGGGGGG\"+cod+\"GGGGGG\",[(11,10+len(cod),'-',0)])==complementary(cod))\n",
    "assert(compose_mRNA(\"GGGGGGGGGG\"+cod[:9]+\"GGGGGG\"+cod[9:],[(11,19,'+',0),(26,34,'+',0)])==cod)\n",
    "assert(compose_mRNA(\"GGGGGGGGGG\"+cod[:9]+\"GGGGGG\"+cod[9:],[(26,34,'-',0),(11,19,'-',0)])==complementary(cod))\n",
    "\n",
    "print(\"All tests are OK!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute the translation tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List of start codons:  {'ATG', 'TTG'}\n",
      "There are 64 entries in the tranlation table.\n",
      "Entries:\n",
      "\n",
      "AAA->K    AAC->N    AAG->K    AAT->N    ACA->T    ACC->T    ACG->T    ACT->T    \n",
      "AGA->R    AGC->S    AGG->R    AGT->S    ATA->I    ATC->I    ATG->M    ATT->I    \n",
      "CAA->Q    CAC->H    CAG->Q    CAT->H    CCA->P    CCC->P    CCG->P    CCT->P    \n",
      "CGA->R    CGC->R    CGG->R    CGT->R    CTA->L    CTC->L    CTG->L    CTT->L    \n",
      "GAA->E    GAC->D    GAG->E    GAT->D    GCA->A    GCC->A    GCG->A    GCT->A    \n",
      "GGA->G    GGC->G    GGG->G    GGT->G    GTA->V    GTC->V    GTG->V    GTT->V    \n",
      "TAA->*    TAC->Y    TAG->*    TAT->Y    TCA->S    TCC->S    TCG->S    TCT->S    \n",
      "TGA->*    TGC->C    TGG->W    TGT->C    TTA->L    TTC->F    TTG->L    TTT->F    "
     ]
    }
   ],
   "source": [
    "def cleanTransTable(tt):\n",
    "    # Remove entries that codify 'X' (unknown)\n",
    "    tt = {k:tt[k] for k in tt if tt[k]!='X'}\n",
    "    # Remove entries that are not GCAT only\n",
    "    tt = {k:tt[k] for k in tt if k.count('G')+k.count('C')+k.count('A')+k.count('T') == 3}\n",
    "    \n",
    "    return tt\n",
    "\n",
    "# Read data\n",
    "genome   = readFasta('haemophilus_influenzae/GCF_000027305.1_ASM2730v1_genomic.fna')\n",
    "cdss     = extractCDSFromGFF('haemophilus_influenzae/test4.gff')\n",
    "proteins = readFasta('haemophilus_influenzae/GCF_000027305.1_ASM2730v1_protein.faa')\n",
    "# Init trans table and start codons\n",
    "translation_table = {}\n",
    "start_codons = set()\n",
    "\n",
    "# Loop over all mRNA found\n",
    "for protein_id in cdss:\n",
    "    cds    = cdss[protein_id]\n",
    "    seq_id = cds[0]\n",
    "    \n",
    "    mRNA   = compose_mRNA(genome[seq_id], cds[1:])\n",
    "    #print(mRNA)\n",
    "    tt, start_codon = obtainTranslationTable(mRNA, proteins[protein_id])\n",
    "    \n",
    "    translation_table.update(tt)\n",
    "    start_codons.add(start_codon)\n",
    "\n",
    "# Remove spurious entries from the result\n",
    "translation_table = cleanTransTable(translation_table)\n",
    "\n",
    "#Print result\n",
    "print(\"List of start codons: \", start_codons)\n",
    "print(\"There are {0} entries in the tranlation table.\".format(len(translation_table)))\n",
    "print(\"Entries:\")\n",
    "\n",
    "for i,k in enumerate(sorted(translation_table.keys())):\n",
    "    if i%8==0:\n",
    "        print()\n",
    "    print(k + \"->\" + translation_table[k], end='    ')\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test against Biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Biopython Verification OK\n"
     ]
    }
   ],
   "source": [
    "from Bio.Data import CodonTable\n",
    "standard_table = CodonTable.unambiguous_dna_by_name[\"Standard\"]\n",
    "biopython_table = standard_table.forward_table.copy()\n",
    "# Add stop codons\n",
    "for codon in standard_table.stop_codons:\n",
    "    biopython_table[codon] = \"*\"\n",
    "assert biopython_table == translation_table\n",
    "print(\"Biopython Verification OK\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "latex_metadata": {
     "hidden": "true"
    }
   },
   "source": [
    "# Exports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "latex_metadata": {
     "hidden": "true"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[NbConvertApp] Converting notebook PRMDM_assignment2.ipynb to latex\n",
      "[NbConvertApp] Support files will be in PRMDM_assignment2_files/\n",
      "[NbConvertApp] Making directory PRMDM_assignment2_files\n",
      "[NbConvertApp] Making directory PRMDM_assignment2_files\n",
      "[NbConvertApp] Making directory PRMDM_assignment2_files\n",
      "[NbConvertApp] Writing 54912 bytes to PRMDM_assignment2.tex\n",
      "[NbConvertApp] Converting notebook PRMDM_assignment2.ipynb to html_with_toclenvs\n",
      "[NbConvertApp] Writing 404126 bytes to PRMDM_assignment2.html\n"
     ]
    }
   ],
   "source": [
    "# Exports\n",
    "!jupyter nbconvert --to latex --template ~/report.tplx PRMDM_assignment2.ipynb 1> /dev/null\n",
    "!pdflatex -shell-escape PRMDM_assignment2.tex 1> /dev/null\n",
    "!jupyter nbconvert --to html_with_toclenvs PRMDM_assignment2.ipynb 1> /dev/null"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "latex_metadata": {
   "author": "Fernando Freire",
   "title": "PRMDM Assignment2"
  },
  "nbTranslate": {
   "displayLangs": [
    "en",
    "sp"
   ],
   "hotkey": "alt-t",
   "langInMainMenu": true,
   "sourceLang": "sp",
   "targetLang": "en",
   "useGoogleTranslate": false
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "256px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
